---
title: "Two Sample Mendelian Randomization on Longevity GWAS"
author: "Christine Le"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document: default
---

```{r setup, include=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(VennDiagram)
library(reactable)
library(tidyverse)
library(kableExtra)
#data



Pilling_dat <- readRDS("data/MR/Pilling_multtest_gwas_ieu.rds")
Timmers_dat <- readRDS("data/MR/Timmers_multtest_gwas_ieu.rds")
Cases90_dat <- readRDS("data/MR/Cases90_multtest_gwas_ieu.rds")
Cases99_dat <- readRDS("data/MR/Cases99_multtest_gwas_ieu.rds")



```


# Background

```{r agingfig}
knitr::include_graphics("images/An_Aging_Nation.png")
```

# Motivation

```{r twomr, out.width= "50%"}
knitr::include_graphics("images/General_DAG.png")
```

**Longevity is complex**
  
  - Confounding makes finding true associations difficult
  
**Solution** : Two Sample Mendelian Randomization

 - use two datasets from similar populations
 
 - provide strong evidence of the true relationship between longevity and certain exposures

# Methods

```{r twomrfig, out.width= "50%"}
knitr::include_graphics("images/2MR.png")
```


**Two Sample Mendelian Randomization**

- Exposure: SNPs as a proxy for Exposure from one dataset
- Outcome: outcome from another dataset using SNPs from Exposure to assess the causal relationship between Exposure and Outcome, bypassing confounders that influence the E-O relationship.

**Advantages**

  - readily available data
  
  - simple analyses
  
  - cost-efficient


# Assumptions 

```{r assumptions, out.width= "50%"}
knitr::include_graphics("images/2MR.png")
```

**Assumption 1** : The first assumption is that SNPs are truly associated with Exposure

**Assumption 2** : No Horizontal Pleiotropy

**Assumption 3** : SNPs not associated with E-O confounders



# TWoSampleMR Package

**Software- TwoSampleMR Package**

R package developed to utilize 2-sample Mendelian Randomization on GWAS studies with curated databases. 
Allows researchers to more easily perform 2-sample Mendelian Randomization on GWASs to discover and eludicate possible causal relationships between phenotypes and evaluate assumptions of this statistical strategy. 

source: Gibran Hemani, Jie Zheng, Kaitlin H Wade, Charles Laurin, Benjamin Elsworth, Stephen Burgess, Jack Bowden, Ryan Langdon, Vanessa Tan, James Yarmolinsky, Hashem A. $ The MR-Base platform supports systematic causal inference across the human phenome. eLife 2018. doi: https://doi.org/10.7554/eLife.34408 



# Data Description - Exposures
Data obtained using IEU GWAS database

**GWAS **
Curated from the NHGRI-EBI GWAS Catalog of significant associations found in different GWAS studies to include only data that has the required information to perform MR analyis. 



# Data Description - Outcomes

**Pilling**
Results from GWAS study of parental longevity (combined monther's and father's age) on UK Biobank participants. Participants whose parental longevity was in the top 10 percent were compared with all other UK Biobank participants who met the inclusion criteria. 

**Timmers**
GWAS results from 1 million parental lifespans of 500,000 particpants with data on mortalilty risk factors. 

**Deleen**

- two defintions of longevity 
  
  - 90th percentile : individuals who lived to upper 90th percentile in age 
  - 99th percentile : individuals who lived to upper 99th percentile in age






# Flow Chart

```{r}
knitr::include_graphics("images/Flow_chart.png")
```


#  Table 1.Common Exposures  among any of the four Outcomes with at least one significant p-value (adjusted p-value $\leq$ 0.05) using the Inverse Variance Weighted Method*

```{r, results='hide',message= FALSE}

#Create a table containing adjusted p-values and information on exposure dataset, and outcome dataset sources
PillingA <- Pilling_dat %>% 
  filter(method == "Inverse variance weighted", exposure_dat != "methylation") %>%
  mutate(Pilling = "YES") %>%
  rename(Exposure.Dataset.Pilling = exposure_dat, Adjusted.p.Pilling = pval_BH, 
         b.pilling = b, se.pilling = se, nsnp.pilling = nsnp, pval.pilling = pval ) %>% 
  select(-id.exposure, -id.outcome, -outcome, -method) 
TimmersA <- Timmers_dat %>% 
  filter(method == "Inverse variance weighted") %>%
  mutate(Timmers = "YES") %>%
  rename(Exposure.Dataset.Timmers = exposure_dat, Adjusted.p.Timmers = pval_BH, 
         b.timmers = b, se.timmers = se, nsnp.timmers = nsnp, pval.timmers = pval) %>% 
  select(-id.exposure, -id.outcome, -outcome, -method) 
Cases90A <- Cases90_dat %>%
  filter(method == "Inverse variance weighted") %>%
  mutate(Cases90 = "YES") %>%
  rename(Exposure.Dataset.Cases90 = exposure_dat, Adjusted.p.Cases90 = pval_BH, 
         b.90 = b, se.90 = se, nsnp.90 = nsnp, pval.90 = pval) %>% 
  select(-id.exposure, -id.outcome, -outcome, -method)
Cases99A <- Cases99_dat %>%
  filter(method == "Inverse variance weighted") %>%
  mutate(Cases99 = "YES") %>%
  rename(Exposure.Dataset.Cases99 = exposure_dat, Adjusted.p.Cases99 = pval_BH,
         b.99 = b, se.99 = se, nsnp.99 = nsnp, pval.99 = pval) %>% 
  select(-id.exposure, -id.outcome, -outcome, -method)

dat_all <- merge(TimmersA, PillingA, by = "exposure", all = TRUE)
dat_all <- merge(dat_all, Cases90A, by = "exposure", all = TRUE)
dat_all <- merge(dat_all, Cases99A, by = "exposure", all = TRUE)


#replace NAs in Dataset columns with "NO"

dat_all$Pilling <- ifelse(is.na(dat_all$Pilling), "NO", dat_all$Pilling)
dat_all$Timmers <- ifelse(is.na(dat_all$Timmers), "NO", dat_all$Timmers)
dat_all$Cases90 <- ifelse(is.na(dat_all$Cases90), "NO", dat_all$Cases90)
dat_all$Cases99 <- ifelse(is.na(dat_all$Cases99), "NO", dat_all$Cases99)



#Make one Exposure Dataset column

dat_all$Exposure.Dataset <- ifelse(dat_all$Pilling == "YES", dat_all$Exposure.Dataset.Pilling,
                               ifelse(dat_all$Timmers == "YES", dat_all$Exposure.Dataset.Timmers,
                                      ifelse(dat_all$Cases90 == "YES", dat_all$Exposure.Dataset.Cases90,
                                             ifelse(dat_all$Cases99 == "YES", dat_all$Exposure.Dataset.Cases99, NA)
                                             )
                                      )
                               )
#remove extra exposure dataset columns
dat <- dat_all
dat <- dat %>%
 select(exposure, Pilling, Timmers, Cases90, Cases99, Adjusted.p.Pilling, Adjusted.p.Timmers, Adjusted.p.Cases90, Adjusted.p.Cases99, Exposure.Dataset)




# Add groups!
```


```{r, results='hide', message= FALSE}
# Table 5 Common Exposures Among All Four Outcomes
table5 <-dat_all %>% filter(Pilling == "YES" & Timmers == "YES" & Cases90 == "YES" & Cases99 == "YES") %>% 
  filter(Adjusted.p.Timmers <= 0.05 | Adjusted.p.Cases90 <= 0.05 | Adjusted.p.Cases99 <= 0.05 | Adjusted.p.Pilling <=0.05) %>%
  # select(exposure,  b.pilling, se.pilling, pval.pilling, Adjusted.p.Pilling,  b.timmers, se.timmers, pval.timmers, Adjusted.p.Timmers, b.90, se.90, pval.90, Adjusted.p.Cases90, b.99, se.99, pval.99, Adjusted.p.Cases99, Exposure.Dataset)
  select(exposure, nsnp.pilling, b.pilling, se.pilling, pval.pilling, Adjusted.p.Pilling, nsnp.timmers, b.timmers, se.timmers, pval.timmers, Adjusted.p.Timmers, nsnp.90, b.90, se.90, pval.90, Adjusted.p.Cases90, nsnp.99, b.99, se.99, pval.99, Adjusted.p.Cases99, Exposure.Dataset) %>%
  mutate_if(is.numeric, funs(as.character(signif(.,3))))

kbl(table5, booktabs = T, escape = F, row.names = FALSE) %>%
    add_header_above(c(" " = 1, "Beta"= 4, "SE"= 4,"No. SNPs" =4, "p-value" = 4, "Adjusted p-value"= 4, " " = 1))%>%
  kable_paper("striped", full_width = T)
```


```{r message=FALSE}
knitr::include_graphics("images/Results.png")
```





# Basal Metabolic Rate


```{r}
# Table 5 Common Exposures Among All Four Outcomes
table <-dat_all %>%
  # select(exposure,  b.pilling, se.pilling, pval.pilling, Adjusted.p.Pilling,  b.timmers, se.timmers, pval.timmers, Adjusted.p.Timmers, b.90, se.90, pval.90, Adjusted.p.Cases90, b.99, se.99, pval.99, Adjusted.p.Cases99, Exposure.Dataset)
  select(exposure,  b.pilling, b.timmers,  b.90, b.99, se.pilling, se.timmers, se.90, se.99, nsnp.pilling, nsnp.timmers, nsnp.90, nsnp.99,pval.pilling, pval.timmers,  pval.90,  pval.99, Adjusted.p.Pilling,  Adjusted.p.Timmers,  Adjusted.p.Cases90,  Adjusted.p.Cases99,) %>%
    filter(grepl("ukb-b-16446", exposure))%>%
  mutate_if(is.numeric, funs(as.character(signif(.,3))))

table$Adjusted.p.Pilling <-   cell_spec(table$Adjusted.p.Pilling, color = 'green')

table$Adjusted.p.Timmers <-   cell_spec(table$Adjusted.p.Timmers, color = "green")

table$Adjusted.p.Cases90<-   cell_spec(table$Adjusted.p.Cases90, color = "green")

table$Adjusted.p.Cases99 <-   cell_spec(table$Adjusted.p.Cases99, color = "red")

colnames(table) <- c( "Exposure", "Pilling", "Timmers", "90th Percentile", "99th Percentile","Pilling", "Timmers", "90th Percentile", "99th Percentile","Pilling", "Timmers", "90th Percentile", "99th Percentile","Pilling", "Timmers", "90th Percentile", "99th Percentile", "Pilling", "Timmers", "90th Percentile", "99th Percentile")

kbl(table,booktabs = T, escape = F, row.names = FALSE, caption = "Table 1. MR Results for Selected Exposure: Basal Metabolic Rate") %>%
    add_header_above(c(" " = 1, "Beta"= 4, "SE"= 4,"No. SNPs" =4, "p-value" = 4, "Adjusted p-value"= 4))%>%
  kable_paper("striped", full_width = T)
```

# Basal Metabolic Rate Results


```{r, results = 'hide'}
table2 <-dat_all %>%
  # select(exposure,  b.pilling, se.pilling, pval.pilling, Adjusted.p.Pilling,  b.timmers, se.timmers, pval.timmers, Adjusted.p.Timmers, b.90, se.90, pval.90, Adjusted.p.Cases90, b.99, se.99, pval.99, Adjusted.p.Cases99, Exposure.Dataset)
  select(exposure,  b.pilling, b.timmers,  b.90, b.99, se.pilling, se.timmers, se.90, se.99, nsnp.pilling, nsnp.timmers, nsnp.90, nsnp.99,pval.pilling, pval.timmers,  pval.90,  pval.99, Adjusted.p.Pilling,  Adjusted.p.Timmers,  Adjusted.p.Cases90,  Adjusted.p.Cases99,) %>%
    filter(grepl("ukb-b-16446", exposure))%>%
  mutate_if(is.numeric, funs(as.character(signif(.,3))))
betas <- table2 %>%
  dplyr::select( b.pilling, b.timmers, b.90, b.99) %>% 
  unlist()
se <- table2 %>%
  dplyr::select(se.pilling, se.timmers, se.90, se.99)%>%
  unlist()
nsnp <- table2 %>%
  dplyr::select(nsnp.pilling, nsnp.timmers, nsnp.90, nsnp.99)%>%
  unlist()

Outcome <- c("Pilling", "Timmers",  "Deleen 90", "Deleen 99")
exposure <- table2$exposure
graph.data <- as.data.frame(cbind(Outcome,exposure, betas, se, nsnp))%>%
  mutate(lower.CI = as.numeric(betas) - (1.96*as.numeric(se)),
         upper.CI = as.numeric(betas) + (1.96*as.numeric(se)),
         OR = signif(exp(as.numeric(betas)),3),
         lower.OR.CI = signif(exp(lower.CI), 3),
         upper.OR.CI = signif(exp(upper.CI),3))

OR_exp <- graph.data %>%
  mutate(CI.95 = paste("(", lower.OR.CI, ", ", upper.OR.CI, ")"))%>%
  select(Outcome, exposure, OR, CI.95, nsnp)
```


```{r}
kbl(OR_exp, booktabs = T, caption = "Table 2. Selected Exposure for Further Analysis: Basal Metabolic Rate.  ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

# Sensitivity Analysis for violations

```{r, results='hide'}
## Function for scatter plot of three methods 2x2

### Legend

# Modify mr_scatter_plot function for our purposes

mrscatter_legend <- function(mr_results, dat, title)
{
	# dat <- subset(dat, paste(id.outcome, id.exposure) %in% paste(mr_results$id.outcome, mr_results$id.exposure))
	requireNamespace("ggplot2", quietly=TRUE)
	requireNamespace("plyr", quietly=TRUE)
	mrres <- plyr::dlply(dat, c("id.exposure", "id.outcome"), function(d)
	{
		d <- plyr::mutate(d)
		if(nrow(d) < 2 | sum(d$mr_keep) == 0)
		{
			return(blank_plot("Insufficient number of SNPs"))
		}
		d <- subset(d, mr_keep)
		index <- d$beta.exposure < 0
		d$beta.exposure[index] <- d$beta.exposure[index] * -1
		d$beta.outcome[index] <- d$beta.outcome[index] * -1
		mrres <- subset(mr_results, id.exposure == d$id.exposure[1] & id.outcome == d$id.outcome[1])
		mrres$a <- 0
		if("MR Egger" %in% mrres$method)
		{
			temp <- mr_egger_regression(d$beta.exposure, d$beta.outcome, d$se.exposure, d$se.outcome, default_parameters())
			mrres$a[mrres$method == "MR Egger"] <- temp$b_i
		}

		if("MR Egger (bootstrap)" %in% mrres$method)
		{
			temp <- mr_egger_regression_bootstrap(d$beta.exposure, d$beta.outcome, d$se.exposure, d$se.outcome, default_parameters())
			mrres$a[mrres$method == "MR Egger (bootstrap)"] <- temp$b_i
		}

		ggplot2::ggplot(data=d, ggplot2::aes(x=beta.exposure, y=beta.outcome)) +
			ggplot2::geom_errorbar(ggplot2::aes(ymin=beta.outcome-se.outcome, ymax=beta.outcome+se.outcome), colour="grey", width=0) +
			ggplot2::geom_errorbarh(ggplot2::aes(xmin=beta.exposure-se.exposure, xmax=beta.exposure+se.exposure), colour="grey", height=0) +
			ggplot2::geom_point(ggplot2::aes(text=paste("SNP:", SNP))) +
			ggplot2::geom_abline(data=mrres, ggplot2::aes(intercept=a, slope=b, colour=method), show.legend=TRUE) +
			ggplot2::scale_colour_manual(values=c("#a6cee3", "#1f78b4", "#b2df8a", "#33a02c", "#fb9a99", "#e31a1c", "#fdbf6f", "#ff7f00", "#cab2d6", "#6a3d9a", "#ffff99", "#b15928")) +
			ggplot2::labs(colour="MR Test", x = "", y=paste("SNP effect on", d$outcome[1]), title = title) +
			ggplot2::theme(legend.position= c(-.1,-.3), legend.key = element_rect(fill = alpha("white" , 0.1)), legend.background = element_rect(fill = alpha("white" , 0.1)), legend.title = element_text(size = 4), legend.text = element_text(size =4)) +
			ggplot2::guides(colour=ggplot2::guide_legend(ncol=3))
	})
	mrres
}


blank_plot <- function(message)
{
	requireNamespace("ggplot2", quietly=TRUE)
	ggplot2::ggplot(data.frame(a=0,b=0,n=message)) + ggplot2::geom_text(ggplot2::aes(x=a,y=b,label=n)) + ggplot2::labs(x=NULL,y=NULL) + ggplot2::theme(axis.text=ggplot2::element_blank(), axis.ticks=ggplot2::element_blank())
}




### No legend

mrscatter_nolegend <- function(mr_results, dat, title)
{
	# dat <- subset(dat, paste(id.outcome, id.exposure) %in% paste(mr_results$id.outcome, mr_results$id.exposure))
	requireNamespace("ggplot2", quietly=TRUE)
	requireNamespace("plyr", quietly=TRUE)
	mrres <- plyr::dlply(dat, c("id.exposure", "id.outcome"), function(d)
	{
		d <- plyr::mutate(d)
		if(nrow(d) < 2 | sum(d$mr_keep) == 0)
		{
			return(blank_plot("Insufficient number of SNPs"))
		}
		d <- subset(d, mr_keep)
		index <- d$beta.exposure < 0
		d$beta.exposure[index] <- d$beta.exposure[index] * -1
		d$beta.outcome[index] <- d$beta.outcome[index] * -1
		mrres <- subset(mr_results, id.exposure == d$id.exposure[1] & id.outcome == d$id.outcome[1])
		mrres$a <- 0
		if("MR Egger" %in% mrres$method)
		{
			temp <- mr_egger_regression(d$beta.exposure, d$beta.outcome, d$se.exposure, d$se.outcome, default_parameters())
			mrres$a[mrres$method == "MR Egger"] <- temp$b_i
		}

		if("MR Egger (bootstrap)" %in% mrres$method)
		{
			temp <- mr_egger_regression_bootstrap(d$beta.exposure, d$beta.outcome, d$se.exposure, d$se.outcome, default_parameters())
			mrres$a[mrres$method == "MR Egger (bootstrap)"] <- temp$b_i
		}

		ggplot2::ggplot(data=d, ggplot2::aes(x=beta.exposure, y=beta.outcome)) +
			ggplot2::geom_errorbar(ggplot2::aes(ymin=beta.outcome-se.outcome, ymax=beta.outcome+se.outcome), colour="grey", width=0) +
			ggplot2::geom_errorbarh(ggplot2::aes(xmin=beta.exposure-se.exposure, xmax=beta.exposure+se.exposure), colour="grey", height=0) +
			ggplot2::geom_point(ggplot2::aes(text=paste("SNP:", SNP))) +
			ggplot2::geom_abline(data=mrres, ggplot2::aes(intercept=a, slope=b, colour=method), show.legend=TRUE) +
			ggplot2::scale_colour_manual(values=c("#a6cee3", "#1f78b4", "#b2df8a", "#33a02c", "#fb9a99", "#e31a1c", "#fdbf6f", "#ff7f00", "#cab2d6", "#6a3d9a", "#ffff99", "#b15928")) +
			ggplot2::labs(colour="MR Test", x=paste(""), y=paste("SNP effect on", d$outcome[1]), title = title) +
			ggplot2::theme(legend.position="none")
	})
	mrres
}


blank_plot <- function(message)
{
	requireNamespace("ggplot2", quietly=TRUE)
	ggplot2::ggplot(data.frame(a=0,b=0,n=message)) + ggplot2::geom_text(ggplot2::aes(x=a,y=b,label=n)) + ggplot2::labs(x=NULL,y=NULL) + ggplot2::theme(axis.text=ggplot2::element_blank(), axis.ticks=ggplot2::element_blank())
}



Pilling_harm <- readRDS("data/harmonized/pilling_df_gwas_ieu.rds")
Timmers_harm <- timmers_gwas_df <- readRDS("data/harmonized/timmers_gwas_ieu_df.rds")
Cases90_harm <-  readRDS("data/harmonized/cases_90_gwas_ieu_df.rds")
Cases99_harm <- readRDS("data/harmonized/cases_99_gwas_ieu_df.rds")

Pilling_harm_dat <- Pilling_harm %>%
  filter(id.exposure == "ukb-b-16446")

Timmers_harm_dat <- Timmers_harm %>%
  filter(id.exposure == "ukb-b-16446")

Cases90_harm_dat <- Cases90_harm %>%
  filter(id.exposure == "ukb-b-16446")

Cases99_harm_dat <- Cases99_harm %>%
  filter(id.exposure == "ukb-b-16446")

### Heterogeneity Test


het.pilling <- mr_heterogeneity(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
het.timmers <- mr_heterogeneity(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
het.cases90 <- mr_heterogeneity(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
het.cases99 <- mr_heterogeneity(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

het_test <- rbind(het.pilling,het.timmers,het.cases90,het.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, method, Q, Q_df, Q_pval) %>%
  rename(`Longevity Definition` = Outcome.dat)

### Check for Pleiotropy
plei.pilling <- mr_pleiotropy_test(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
plei.timmers <- mr_pleiotropy_test(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
plei.cases90 <- mr_pleiotropy_test(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
plei.cases99 <- mr_pleiotropy_test(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

pli_test <- rbind(plei.pilling, plei.timmers,plei.cases90,plei.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, egger_intercept, se, pval) %>%
  rename(`Longevity Definition` = Outcome.dat,
         `Egger Intercept` = egger_intercept)

## Leave on out Analysis - Check to see if there is one SNP influencing the results

loo.pilling <- mr_leaveoneout(Pilling_harm_dat)%>%
  mutate(outcome.data = "Pilling")
loo.timmers <-mr_leaveoneout(Timmers_harm_dat)%>%
  mutate(outcome.data = "Timmers")
loo.cases90 <-mr_leaveoneout(Cases90_harm_dat)%>%
  mutate(outcome.data = "Cases90")
loo.cases99 <- mr_leaveoneout(Cases99_harm_dat)%>%
  mutate(outcome.data = "Cases99")

p1 <- mr_leaveoneout_plot(loo.pilling)
p2 <- mr_leaveoneout_plot(loo.timmers)
p3 <- mr_leaveoneout_plot(loo.cases90)
p4 <- mr_leaveoneout_plot(loo.cases99)

p1 <- grid.arrange(grobs = p1)
p2 <- grid.arrange(grobs = p2)
p3 <- grid.arrange(grobs = p3)
p4 <- grid.arrange(grobs = p4)


## Scatter plot of Three Methods

mr.pilling <- mr(Pilling_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.timmers <- mr(Timmers_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases90 <- mr(Cases90_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases99 <- mr(Cases99_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))

sc.pilling <- mrscatter_nolegend(mr.pilling, Pilling_harm_dat, title = "(a) Pilling")
sc.timmers <- mrscatter_nolegend(mr.timmers, Timmers_harm_dat, title ="(b) Timmers")
sc.cases90 <- mrscatter_nolegend(mr.cases90, Cases90_harm_dat, title ="(c) Deleen 90th Percentile")
sc.cases99 <- mrscatter_legend(mr.cases99, Cases99_harm_dat, title ="(d) Deleen 99th Percentile")


# Forest plot
pilling.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
timmers.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases90.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases99.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))


## Funnel plot of Single SNP analysis compared to the estimates from all SNPs
pilling.funnel <- grid.arrange(grobs = mr_funnel_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
timmers.funnel <- grid.arrange(grobs = mr_funnel_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases90.funnel <- grid.arrange(grobs = mr_funnel_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases99.funnel <- grid.arrange(grobs = mr_funnel_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
```

```{r}

# Test for heterogeneity
kbl(het_test, booktabs = T, caption = "Test for Heterogeneity using Cochran's Q-test")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")

### Check for Pleiotropy

kbl(pli_test, booktabs = T, caption = "Test for Pleiotropy ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")


# Forest plot

plot_grid(pilling.forest, timmers.forest,labels =  c("Pilling", "Timmers"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
plot_grid(cases90.forest, cases99.forest,labels =  c("Deleen 90", "Deleen 99"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")


## Funnel plot of Single SNP analysis compared to the estimates from all SNPs

plot_grid(pilling.funnel, timmers.funnel,labels =  c("Pilling", "Timmers"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
plot_grid(cases90.funnel, cases99.funnel,labels =  c("Deleen 90", "Deleen 99"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
```

# Conclusion

- 90 sample size is much larger than 99 so p -value might be reduced because of small sample size, stronger effect as phenotype becomes more extreme

- lots of significant exposures





# Limitations 

- low power due to weak instruments

- the evidence in and of itself does not indicate causation, but provides evidence that could lead to investments for further study

- further examination of significant associations for strong violations of assumptions and need for further analyses 

- consistency of the direction of the beta coefficients across the different outcomes

- do loo analysis on all significant results and check if one of the non-european SNPs has an outlier effect on estimate especially important for UKBiobank studies
    - for any exposures examined in the context of the comparison population, make sure SNPs from non-Euro study doesn't have undo influence

- collider bias
- developmental compensation/canalization
    
# Next Steps
- directionality
- compare the results within the database, e.g. how were exposures defined, inclusion/exclusion of participants
- test correlation among four outcome definitions- further adjustment might be needed due to four outcomes

# Citations 

Stephen Burgess, Simon G Thompson, CRP CHD Genetics Collaboration, Avoiding bias from weak instruments in Mendelian randomization studies, International Journal of Epidemiology, Volume 40, Issue 3, June 2011, Pages 755–764, https://doi.org/10.1093/ije/dyr036


Davies NM, Holmes MV, Davey Smith G. Reading Mendelian randomisation studies: a guide, glossary, and checklist for clinicians. BMJ. 2018;362:k601. Published 2018 Jul 12. doi:10.1136/bmj.k601

Zheng, J., Baird, D., Borges, M. et al. Recent Developments in Mendelian Randomization Studies. Curr Epidemiol Rep 4, 330–345 (2017). https://doi.org/10.1007/s40471-017-0128-6

are there more exposures that are significant in shared among deleen 90 and pilling than in deleen and timmers