---
title: "Longevity"
author: "C le"
output: html_document
date :  "`r format(Sys.time(), '%d %B, %Y')`"
---



# Project Overview
Two sample Mendelian Randomization Analysis on GWAS of human lifespan looking at new and known exposure relationships related to longevity outcomes.

## A. Exposures

Exposures are obtained from the mrbase.org website from the following catalogs
  - MR BASE GWAS catalog
  - Gene Expression QTLs
  - Protein Level QTLs
  - Metabolite Level OTLs
  - Methylation Level QTLs

## B. Outcomes

Download Deelen outcomes and files for rsID mapping.

```{r, eval = FALSE}
#Mapping file from EasyQC. Place is data/outcomes/
system("wget https://homepages.uni-regensburg.de/~wit59712/easyqc/1000g/rsmid_map.1000G_ALL_p1v3.merged_mach_impute.v3.mergeindels.txt.gz")
#Deelen GWAS outcomes
system("wget ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/DeelenJ_31413261_GCST008598/Results_90th_percentile.txt.gz")
system("wget ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/DeelenJ_31413261_GCST008599/Results_99th_percentile.txt.gz")
```

Download Pilling outcomes
```{r, eval = FALSE}

system(wget ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/PillingLC_29227965_GCST006698/harmonised/

```

Download Timmers outcomes

```{r}

system("wget ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/PillingLC_29227965_GCST006698/harmonised/29227965-GCST006698-EFO_0007796-build37.f.tsv.gz")
```

### I. First Outcome: cases 90

Outcome Data obtained from CPMC that was utilized in Deelen et al., 2019 (https://pubmed.ncbi.nlm.nih.gov/31413261/). Exposure - Outcome relationships will be examined in persons whose longevity is in the 90th percentile using the rest of the cohort as the control.


Map cptids to rsids using EasyQC map files
```{r}
require(tidyverse)
outcome <- read_delim("data/outcomes/Results_90th_percentile.txt.gz", delim = "\t")

#create SNPorder to track dups caused by multiple matches
outcome <- outcome %>%
  mutate(SNPorder = seq_along(SNP))

mapfile <- read_delim("data/outcomes/rsmid_map.1000G_ALL_p1v3.merged_mach_impute.v3.mergeindels.txt", delim = "\t")

#rsmid contains rsids
mapfile %>%
  filter(grepl("^rs.*", rsmid)) %>%
  summarize(n())

outcome <- outcome %>%
  left_join(mapfile, by = c("SNP" = "cptid"))

# All 1:1 merge, no dups induced
outcome %>%
  filter(duplicated(SNPorder)) %>%
  summarize(n())
#Chr matches with chr
outcome %>%
  filter(Chr != chr) %>%
  summarize(n()) 
#Position mismatches with pos = 2
outcome %>%
  filter(Position != pos) %>%
  summarize(n()) 
map_int(outcome, function(x) sum(is.na(x))) #4 missing rsmids
#remove missing rsids, mismatching chr and pos, rename vars,
# remove unneeded vars
outcome <- outcome %>%
  filter(!is.na(rsmid)) %>%
  filter(Chr == chr) %>%
  filter(Position == pos) %>%
  rename(rsid = rsmid, pval = "P-value", cptid = SNP) %>%
  mutate(chr = NULL) %>%
  mutate(pos = NULL) %>%
  mutate(SNPorder = NULL) %>%
  select(rsid, everything())

write_csv(outcome ,"data/outcomes/cases90rsid.csv")
```


### II. Second outcome: cases 99

SNP data from persons whose longevity is in the 99th percentile. 
Map cptids to rsids using EasyQC map files

```{r}
require(tidyverse)
outcome <- read_delim("data/outcomes/Results_99th_percentile.txt.gz", delim = "\t")
#create SNPorder to track dups caused by multiple matches
outcome <- outcome %>%
  mutate(SNPorder = seq_along(SNP))

mapfile <- read_delim("data/outcomes/rsmid_map.1000G_ALL_p1v3.merged_mach_impute.v3.mergeindels.txt", delim = "\t")

#rsmid contains rsids
mapfile %>%
  filter(grepl("^rs.*", rsmid)) %>%
  summarize(n())

outcome <- outcome %>%
  left_join(mapfile, by = c("SNP" = "cptid"))

# All 1:1 merge, no dups induced
outcome %>%
  filter(duplicated(SNPorder)) %>%
  summarize(n())
#Chr matches with chr
outcome %>%
  filter(Chr != chr) %>%
  summarize(n()) 
#Position mismatches with pos = 2
outcome %>%
  filter(Position != pos) %>%
  summarize(n()) 
map_int(outcome, function(x) sum(is.na(x))) #4 missing rsmids
#remove missing rsids, mismatching chr and pos, rename vars,
# remove unneeded vars
outcome <- outcome %>%
  filter(!is.na(rsmid)) %>%
  filter(Chr == chr) %>%
  filter(Position == pos) %>%
  rename(rsid = rsmid, pval = "P-value", cptid = SNP) %>%
  mutate(chr = NULL) %>%
  mutate(pos = NULL) %>%
  mutate(SNPorder = NULL) %>%
  select(rsid, everything())

write_csv(outcome ,"data/outcomes/cases99rsid.csv")

```
rsid = rsid
cptid = 
chr = chromosome
position = position
EA = effect allele
NEA = non effect allele
EAF = effective allele frequency
Beta = Beta
SE = standard error
pval = pvalue
effective_N = sample size

### III. Third Outcome: UK Biobank

Persons whose parental longevity was in the 90th percentile from the UK Biobank

Map cptids to rsids using EasyQC map files
```{r}
require(tidyverse)
outcome <- read_delim("data/outcomes/29227965-GCST006698-EFO_0007796-build37.f.tsv.gz", delim = "\t")

# #filter out rsids already mapped
# outcome <- outcome %>%
#   filter(!grepl("^rs.*", variant_id))
# 
# #keep thos that have already been mapped
# out <- outcome %>%
#   filter(grepl("^rs.*", variant_id))

# remove alleles at tend of variant ID

outcome <- outcome %>%
  mutate(variant_id = str_replace_all(variant_id, "A", "")) %>%
  mutate(variant_id = str_replace_all(variant_id, "G", "")) %>%
  mutate(variant_id = str_replace_all(variant_id, "T", "")) %>%
  mutate(variant_id = str_replace_all(variant_id, "C", "")) %>%
  mutate(variant_id = str_replace_all(variant_id, "_", ""))
  
#create SNPorder to track dups caused by multiple matches 11516100
outcome <- outcome %>%
  mutate(SNPorder = seq_along(variant_id))

mapfile <- read_delim("data/outcomes/rsmid_map.1000G_ALL_p1v3.merged_mach_impute.v3.mergeindels.txt", delim = "\t")

#rsmid contains rsids
m <- mapfile %>%
  filter(grepl("^rs.*", rsmid)) %>%
  summarize(n())

#join map file by chromosome and position
outcome <- outcome %>%
  left_join(mapfile, by = c("chromosome"= "chr", "base_pair_location"= "pos"))



# All 1:1 merge, 5046 dups induced
x <- outcome %>%
  filter(duplicated(SNPorder)) %>%
  summarize(n())


map_int(outcome, function(x) sum(is.na(x))) #1360831 missing rsmids
#remove missing rsids, mismatching chr and pos, rename vars,
# remove unneeded vars
outcome <- outcome %>%
  filter(!is.na(rsmid)) %>%
  filter(chromosome == chr) %>%
  filter(base_pair_location == pos) %>%
  rename(rsid = rsmid, cptid = variant_id) %>%
  mutate(chr = NULL) %>%
  mutate(pos = NULL) %>%
  mutate(SNPorder = NULL) %>%
  select(rsid, everything())


out <- out %>%
  mutate(rsid = variant_id)%>%
  rename(cptid = variant_id)

#599 additional SNPs mapped.
write_csv(outcome ,"data/outcomes/pilling_mapped_all.csv")


#put mapped and already provided rsids together
pilling_download <- rbind(outcome, out)

write_csv(pilling_download, "data/outcomes/Pilling_90.csv")

```


### IV. Fourth Outcome: UK Biobank Parental Longevity

### V. Comparisons
Compare significant Exposures-Outcome relationships found via Two Sample Mendelian Randomization Analysis Among these four studies, using the 90th percentile outcomes in Deleen as the baseline. 

## C. Preparation of Data

### I. Load Libraries

Run this code chunk the first time you use this code to check that you have all the packages needed.
```{r,eval=FALSE}

if(!require(TwoSampleMR)){
  install.packages("devtools")
  devtools::install_github("MRCIEU/TwoSampleMR")
}
  
if(!require(ieugwasr)){
  #if devtools not installed yet, use line below
  #install.packages("devtools")
  devtools::install_github("mrcieu/ieugwasr")
}

if(!require(genetics.binaRies)){
  #if devtools not installed yet, use line below
  #install.packages("devtools")
  devtools::install_github("explodecomputer/genetics.binaRies")
}



if(!requireNamespace("BiocManager", quietly = TRUE)){
 install.packages("BiocManager") 
 BiocManager::install("multtest")
}
    




is_it_here <- function(package_list){
  for( i in seq_along(package_list)){
    if(!require(package_list[i])){
      install.packages(package_list[i])
    }
  }
}
packages_CRAN <- c("dplyr", "knitr", "arsenal","tidyverse", "data.table", "doParallel")
is_it_here(packages_CRAN)
```

Load the libraries you need for the subsequent code chunks
```{r}
library(TwoSampleMR)
library(MRInstruments)
library(tidyverse) #so we can use purrr::map, and no, not just because I got cats
library(knitr)
library(ieugwasr)
library(arsenal)
library(tidyverse)
library(multtest)
library(doParallel)
```

###II.  Exposures from each catalog
Local LD clumping was done per [reference url] and reference data was obtained from 1000 genomes at [ http://fileserve.mrcieu.ac.uk/ld/1kg.v3.tgz]

Below is the code to do local LD clumping. However in **Section F** there are examples to do it via the TwoSampleMR reference via the internet as well as an example extracting exposures as a function using apply.

Code in words:
For each unique phenotype, we format the data for use with the TwoSampleMR package, then we ensure independence of SNPs by clumping the formatted data to ensure that the SNPs are not too close to each other, thereby violating our independence assumption.

Make sure to set eval = **TRUE** if you want to run that chunk of code. 

Set variables for local clumping of data
```{r,eval=FALSE}


# plink library
plink <- genetics.binaRies::get_plink_binary()

# bfile is the filepath were your reference datasets are located
#no need to include extensions for population (e.g EUR.bin)
bfile <- "data/LD/LD_reference/EUR"
```

#### a. GWAS Catalog

```{r, eval=FALSE}

#cache = True will create a folder that saves objects, and if objects and code aren't updated, this code chunk will not re-run. This is good for long-running code sections.
#import gwas catalog into R

gwas_catalog <- available_outcomes()

#create a vector of unique gwas phenotypes
gwas_unique <- gwas_catalog %>%
  pull(id)
# 34513 unique id in catalog

  exposures_gwas <- extract_instruments(outcomes = gwas_unique ) 

#rbind list elements into DF
#old way of doing things
#exposures_gwas_DF <- do.call(rbind, exposures_gwas)
#faster way of doing things with dplyr
# exposures_gwas_DF <- bind_rows(exposures_gwas)

#check that there are still 34513 unique ids
#23282 unique ids left. 
x$id <-unique(exposures_gwas$id.exposure)

# 11231 ids that were not kept , test if ids results in empty vector (no rsids)
null <- anti_join(gwas_catalog, x, by = c("id"))

null_id <- null %>% pull(id)

null_exp <- extract_instruments(outcomes = null_id)
# results in empty vector
# 
save(exposures_gwas, file = "data/exposures/exposures_gwas_ieu.Rdata")


# check if there are duplicates in all other columns even if id is unique among same traits
traits_ls <- unique(gwas_catalog$trait)
result <- vector(mode = "list", length = length(traits_ls))
for(i in seq_along(traits_ls)){
  trait <- gwas_catalog %>%
    filter(trait == traits_ls[i])
  duplicated = (duplicated(trait[ , 3:28]))
  result[[i]] <-tibble( trait = traits_ls[i],
                        location = i,
                      duplicated = duplicated)
  
}

trait_df <- bind_rows(result)

unique_dups <- trait_df %>%
  filter(duplicated == TRUE) %>%
  filter(!duplicated(trait))

#pull info from gwas

dups <- unique_dups %>% pull(trait)

gwas_dups <- gwas_catalog %>%
  filter(trait %in% dups)

write_csv(gwas_dups, "data/possible_gwas_duplicates.csv")
```



#### b. Gene Expression QTLs
Search for SNPs can be done by gene and tissue or just one. Below exposure instruments are created for each unique gene.

```{r, eval= FALSE}

#import Gene Expression QTLs list
data(gtex_eqtl)
gtex_eqtl


#create a list of unique gene names

genes_unique <- gtex_eqtl[!duplicated(gtex_eqtl$gene_name),] %>% pull(gene_name)

# 32432 unique gene names

genes_table <- gtex_eqtl[!duplicated(gtex_eqtl$gene_name),]

exposures_genes <- data.frame()
for(i in 17583: length(genes_unique)){
  genes <- subset(gtex_eqtl, gene_name == genes_unique[i]) %>%
    format_gtex_eqtl()%>%
    rename(rsid = SNP,
         pval = pval.exposure,
         id = id.exposure)


  genes_format <-  ld_clump(exposures_genes, plink_bin = plink, bfile = bfile,
           clump_kb = 10000, clump_r2 = .001, clump_p = 1) 

  exposures_genes <- rbind(exposures_genes, genes_format)
}

saveRDS(exposures_genes, file = "data/exposures/exposures_genes.rds")
```

#### c. Protein Level QTLs

```{r, eval = FALSE}
#import Protein Level QTLs into R
data(proteomic_qtls)
proteomic_qtls

#create a list of unique protein analyte & make a list

proteins_unique <- proteomic_qtls[!duplicated(proteomic_qtls$analyte),] %>% pull(analyte)


exposures_protein <- data.frame()
for(i in 1 : length(proteins_unique)) {
  
  proteins_phenotypes<- subset(proteomic_qtls, analyte == proteins_unique[i])
  
  #format exposure data and clump by LD r2 <0.001 to reduce covariance
  proteins_format<- proteins_phenotypes %>% format_proteomic_qtls() %>%
    rename(rsid = SNP,
         pval = pval.exposure,
         id = id.exposure)
  
  proteins_clump <- ld_clump(proteins_format, 
                             plink_bin = plink, 
                             bfile = bfile,
                             clump_kb = 10000, clump_r2 = .001, clump_p = 1)

  exposures_protein <- rbind(exposures_protein, proteins_clump)
    
}

exposures_protein <- exposures_protein %>% rename (SNP = rsid, id.exposure = id)

saveRDS(exposures_genes, file = "data/exposures/exposures_proteins.rds")
```

#### d. Metabolite Level OTLs

```{r, eval=FALSE}

#import Metabolite Level QTLs into R
data(metab_qtls)
metab_qtls

#create a list of unique metabolomic phenotype & make a list

metabolites_unique <- metab_qtls[!duplicated(metab_qtls$phenotype),] %>% pull(phenotype)
# 121 unique phenotypes in catalog



#make a for loop to create a database of exposure SNPs formatted for exposure data and clumped
exposures_metabolite <- data.frame()
for(i in 1 : length(metabolites_unique)) {
  
  metabolites_phenotypes<- subset(metab_qtls, phenotype == metabolites_unique[i]) %>% 
    format_metab_qtls() %>% 
    rename(rsid = SNP,
         pval = pval.exposure,
         id = id.exposure) %>%
    ld_clump( plink_bin = plink, 
              bfile = bfile,
              clump_kb = 10000, clump_r2 = .001, clump_p = 1)
    

  exposures_metabolite <- rbind(exposures_metabolite, metabolites_phenotypes)
    
}

exposures_metabolite <- exposures_metabolite %>% rename (SNP = rsid, id.exposure = id)  
  
#check that 'phenotypes' is the same as the last entry in metabolites_unique
metabolites_unique[121]
metabolites_phenotypes$phenotype


saveRDS(exposures_metabolite, file = "data/exposures/exposures_metabolites.rds")
```

#### e. Methylation Level QTLs
by cpg site and age

```{r}

#import Methylation Level QTLs into R
data(aries_mqtl)
aries_mqtl

#create a list of unique Methylation  cpg & make a list

methylation_unique <- aries_mqtl[!duplicated(aries_mqtl$cpg),]%>% pull(cpg)
age <- c("Birth", "Adolescence", "Childhood", "Middle age", "Pregnancy")
#  33256 unique cpg sites in catalog
# 5 'ages'

#make a for loop to create a database of exposure SNPs formatted for exposure data and clumped


exposures_methylation <- data.frame()
for (j in 1:5){
  for(i in 1 : 33256) {
  
    methylation_phenotypes<- subset(aries_mqtl, cpg == methylation_unique[i] & age == age[j])%>% 
      format_aries_mqtl() %>% 
      rename(rsid = SNP,
         pval = pval.exposure,
         id = id.exposure) %>%
    ld_clump( plink_bin = plink, 
              bfile = bfile,
              clump_kb = 10000, clump_r2 = .001, clump_p = 1)
    
    
  exposures_methylation <- rbind(exposures_methylation, methylation_phenotypes)
     

    
    
  }


}

exposures_methylation <- exposures_methylation %>% rename (SNP = rsid, id.exposure = id)   
#check that 'phenotypes' is the same as the last entry in metabolites_unique
methylation_unique[33256]
methylation_phenotypes$cpg


saveRDS(exposures_methylation, file = "data/exposures/exposures_methylation.rds")
```

#### f. Other Methods for obtaining exposures

Ex. Genes Catalog

Utilizing the clump_data function in TwoSampleMR

```{r}
#Retrieve the catalog
data(gtex_eqtl)
gtex_eqtl

# then create a vector of unique genes

genes_table <- gtex_eqtl[!duplicated(gtex_eqtl$gene_name),]

exposures_genes <- data.frame()
for(i in 1: length(genes_unique)){
  genes <- subset(gtex_eqtl, gene_name == genes_unique[i])

  #create a separate clumping step in case LD lookup times out
  genes_format <- format_gtex_eqtl(genes) %>% clump_data()

  exposures_genes <- rbind(exposures_genes, genes_format)
}

saveRDS(exposures_genes, file = "data/exposures/exposures_genes.rds")
```

As a function for a list of genes 
```{r}
#As before retrieve the catalog
data(gtex_eqtl)
gtex_eqtl

# then create a vector of unique genes

genes_table <- gtex_eqtl[!duplicated(gtex_eqtl$gene_name),]

# Define a function
gene_clump <- function(genes){
  gene <- subset(gtex_eqtl, gene_name == genes) %>% format_gtex_eqtl() %>% clump_data()
}


test <- lapply(gwas_unique, genes_clump)
```

### III. Outcomes & Harmonize
Here we will use the TwoMRSample package to merge the exposure and outcome dataset by SNP/rsid for each unique exposure, then harmonize the data and compile it.

#### a. Loading Exposures

```{r}
#GWAS
load("data/exposures/exposures_gwas_ieu.Rdata")

#Proteins
exposures_proteins <- readRDS("data/exposures/exposures_proteins.rds")

#Methylation
exposures_methylation <- readRDS( "data/exposures/exposures_methylation.rds")

#Genes
exposures_genes <- readRDS( "data/exposures/exposures_genes.rds")

#Metabolites
exposures_metabolites <- readRDS("data/exposures/exposures_metabolites.rds")


```


#### b. Cases 90
GWAS

```{r}
error_log <- data.frame()

exp_unique <- exposures_gwas[!duplicated(exposures_gwas$id.exposure),] %>% pull(id.exposure)
cases_90 <- vector(mode = "list", length = length(exp_unique))

for(i in seq_along(exp_unique)){
# for(i in 1:1){
  exp <- exposures_gwas %>%
    filter(id.exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "data/outcomes/cases90rsid.csv",
    sep = ",",
    snp_col = "rsid",
    beta_col = "Beta",
    se_col = "SE",
    effect_allele_col = "EA",
    other_allele_col = "NEA",
    eaf_col = "EAF",
    pval_col = "pval",
    samplesize_col = "effective_N"),
    error=function(e) e)

  if(inherits(outcome, "error") == FALSE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      cases_90[[i]] <- merge(exp, outcome, by = "SNP") %>%
        mutate(error = "Harmonize")
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        cases_90[[i]] <- harmonize %>%
          mutate(error = "FALSE")
  

      }
    

  }else{
    cases_90[[i]] <- tibble(error = "no matching rsids in outcome")
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
  }
    
  print(paste0("Harmonized exposure ", exp_unique[i], 
               " which is iterator ", i)) 
}
 

cases_90_df <- bind_rows(cases_90)

saveRDS(error_log, "data/harmonized/error_log_cases_90_gwas_ieu.rds")


saveRDS(cases_90, "data/harmonized/cases_90_gwas_ieu_ls.rds")
saveRDS(cases_90_df, "data/harmonized/cases_90_gwas_ieu_df.rds")
```

789 exposure rsids not in outcome


PROTEIN EXPOSURES

```{r}
error_log <- data.frame()

exp_unique <- exposures_proteins[!duplicated(exposures_proteins$exposure),] %>% pull(exposure)

cases_90 <- vector(mode = "list", length = length(exp_unique))

for(i in seq_along(exp_unique)){ 
  
  exp <- exposures_proteins %>%
    filter(exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "data/outcomes/cases90rsid.csv",
    sep = ",",
    snp_col = "rsid",
    beta_col = "Beta",
    se_col = "SE",
    effect_allele_col = "EA",
    other_allele_col = "NEA",
    eaf_col = "EAF",
    pval_col = "pval",
    samplesize_col = "effective_N"),
    error=function(e) e)

  if(inherits(outcome, "error") == FALSE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      cases_90[[i]] <- merge(exp, outcome, by = "SNP") %>%
        mutate(error = "Harmonize")
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        cases_90[[i]] <- harmonize %>%
          mutate(error = "FALSE")
  

      }
    

  }else{
    cases_90[[i]] <- tibble(error = "no matching rsids in outcome")
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
  }
    
  print(paste0("Harmonized exposure ", exp_unique[i], 
               " which is iterator ", i)) 
}
 

cases_90_df <- bind_rows(cases_90)
 
saveRDS(error_log, "data/harmonized/error_log_cases_90_protein.rds")


saveRDS(cases_90, "data/harmonized/cases_90_protein.rds")
saveRDS(cases_90_df, "data/harmonized/cases_90_protein_df.rds")
```
0 errors in harmonizing
matching SNPs found for each protein
METABOLITES


```{r}

error_log <- data.frame()

exp_unique <- exposures_metabolites[!duplicated(exposures_metabolites$exposure),] %>% pull(exposure)

cases_90 <- vector(mode = "list", length = length(exp_unique))

for(i in seq_along(exp_unique)){ 
  
  exp <- exposures_metabolites %>%
    filter(exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "data/outcomes/cases90rsid.csv",
    sep = ",",
    snp_col = "rsid",
    beta_col = "Beta",
    se_col = "SE",
    effect_allele_col = "EA",
    other_allele_col = "NEA",
    eaf_col = "EAF",
    pval_col = "pval",
    samplesize_col = "effective_N"),
    error=function(e) e)

  if(inherits(outcome, "error") == FALSE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      cases_90[[i]] <- merge(exp, outcome, by = "SNP") %>%
        mutate(error = "Harmonize")
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        cases_90[[i]] <- harmonize %>%
          mutate(error = "FALSE")
  

      }
    

  }else{
    cases_90[[i]] <- tibble(error = "no matching rsids in outcome")
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
  }
    
  print(paste0("Harmonized exposure ", exp_unique[i], 
               " which is iterator ", i)) 
}
 

cases_90_df <- bind_rows(cases_90)
 
saveRDS(error_log, "data/harmonized/error_log_cases90_metabolites.rds")


saveRDS(cases_90, "data/harmonized/cases_90_metabolites.rds")
saveRDS(cases_90_df, "data/harmonized/cases_90_metabolites_df.rds")
```




GENES

```{r}
error_log <- data.frame()

exp_unique <- exposures_genes[!duplicated(exposures_genes$exposure),] %>% pull(exposure)

cases_90 <- vector(mode = "list", length = length(exp_unique))

# for(i in seq_along(exp_unique)){ 
for(i in 1:seq_along(exp_unique)){#finished
  
  exp <- exposures_genes %>%
    filter(exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "../data/outcomes/cases90rsid.csv",
    sep = ",",
    snp_col = "rsid",
    beta_col = "Beta",
    se_col = "SE",
    effect_allele_col = "EA",
    other_allele_col = "NEA",
    eaf_col = "EAF",
    pval_col = "pval",
    samplesize_col = "effective_N"),
    error=function(e) e)

  if(!inherits(outcome, "error") == TRUE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      cases_90[[i]] <- merge(exp, outcome, by = "SNP") %>%
        mutate(error = "Harmonize")
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        cases_90[[i]]<- harmonize %>%
          mutate(error = "FALSE")
  

      }
  

  }else{
    cases_90[[i]] <- tibble(error = "no matching rsids in outcome")
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
    
    
  }
  
   

  print(paste0("Harmonized exposure ", exp_unique[i], 
               " which is iterator ", i))    
}


cases_90_DF <- bind_rows(cases_90)
 
saveRDS(error_log, "data/harmonized/error_log_cases_90_genes.rds")


saveRDS(cases_90, "data/harmonized/cases_90_genes.rds")
saveRDS(cases_90_df, "data/harmonized/cases_90_genes_df.rds")
```
methylation
fixed and added methylation error exposures

```{r}
error_log <- data.frame()

exp_unique <- exposures_methylation[!duplicated(exposures_methylation$exposure),] %>% pull(exposure)

cases_90<- vector(mode = "list", length = length(exp_unique))

for(i in seq_along(exp_unique)){
# for(i in 120000:140099){
  
  exp <- exposures_methylation %>%
    filter(exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "data/outcomes/cases90rsid.csv",
    sep = ",",
    snp_col = "rsid",
    beta_col = "Beta",
    se_col = "SE",
    effect_allele_col = "EA",
    other_allele_col = "NEA",
    eaf_col = "EAF",
    pval_col = "pval",
    samplesize_col = "effective_N"),
    error=function(e) e)

  if(!inherits(outcome, "error") == TRUE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      cases_90[[i]] <- merge(exp, outcome, by = "SNP") %>%
        mutate(error = "Harmonize")
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        cases_90[[i]]<- harmonize %>%
          mutate(error = "FALSE")
  

      }
  

  }else{
    cases_90[[i]] <- tibble(error = "no matching rsids in outcome")
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
    
    
  }
  
   

  print(paste0("Harmonized exposure ", exp_unique[i], 
               " which is iterator ", i))    
}

cases_90_DF <- bind_rows(cases_90)
 
saveRDS(error_log, "data/harmonized/error_log_cases_90_methylation.rds")


saveRDS(cases_90, "data/harmonized/cases_90_methylation.rds")

saveRDS(cases_90_DF,"data/harmonized/cases_90_methylation_df.rds")

```



#### c. Cases 99
GWAS

```{r}
error_log <- data.frame()

exp_unique <- exposures_gwas[!duplicated(exposures_gwas$id.exposure),] %>% pull(id.exposure)
cases_99 <- vector(mode = "list", length = length(exp_unique))

# for(i in seq_along(exp_unique)){ 
# for(i in 1:5){  
  exp <- exposures_gwas %>%
    filter(id.exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "data/outcomes/cases99rsid.csv",
    sep = ",",
    snp_col = "rsid",
    beta_col = "Beta",
    se_col = "SE",
    effect_allele_col = "EA",
    other_allele_col = "NEA",
    eaf_col = "EAF",
    pval_col = "pval",
    samplesize_col = "effective_N"),
    error=function(e) e)

  if(inherits(outcome, "error") == FALSE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      cases_99[[i]] <- merge(exp, outcome, by = "SNP") %>%
        mutate(error = "Harmonize")
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        cases_99[[i]] <- harmonize %>%
          mutate(error = "FALSE")
  

      }
    

  }else{
    cases_99[[i]] <- tibble(error = "no matching rsids in outcome")
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
  }
    
  print(paste0("Harmonized exposure ", exp_unique[i], 
               " which is iterator ", i)) 
}
 

cases_99_df <- bind_rows(cases_99)

saveRDS(error_log, "data/harmonized/error_log_cases_99_gwas_ieu.rds")


saveRDS(cases_99, "data/harmonized/cases_99_gwas_ieu_ls.rds")
saveRDS(cases_99_df, "data/harmonized/cases_99_gwas_ieu_df.rds")
```
178 exposures with no matching rsids in outcome

1216 harmonizing errors
PROTEIN EXPOSURES

```{r}
error_log <- data.frame()

exp_unique <- exposures_proteins[!duplicated(exposures_proteins$exposure),] %>% pull(exposure)

cases_99 <- vector(mode = "list", length = length(exp_unique))

for(i in seq_along(exp_unique)){ 
  
  exp <- exposures_proteins %>%
    filter(exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "data/outcomes/cases99rsid.csv",
    sep = ",",
    snp_col = "rsid",
    beta_col = "Beta",
    se_col = "SE",
    effect_allele_col = "EA",
    other_allele_col = "NEA",
    eaf_col = "EAF",
    pval_col = "pval",
    samplesize_col = "effective_N"),
    error=function(e) e)

  if(inherits(outcome, "error") == FALSE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      cases_99[[i]] <- merge(exp, outcome, by = "SNP") %>%
        mutate(error = "Harmonize")
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        cases_99[[i]] <- harmonize %>%
          mutate(error = "FALSE")
  

      }
    

  }else{
    cases_99[[i]] <- tibble(error = "no matching rsids in outcome")
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
  }
    
  print(paste0("Harmonized exposure ", exp_unique[i], 
               " which is iterator ", i)) 
}
 

cases_99_df <- bind_rows(cases_99)
 
saveRDS(error_log, "data/harmonized/error_log_cases_99_protein.rds")


saveRDS(cases_99, "data/harmonized/cases_99_protein.rds")
saveRDS(cases_99_df, "data/harmonized/cases_99_protein_df.rds")
```

no errors


METABOLITES


```{r}

error_log <- data.frame()

exp_unique <- exposures_metabolites[!duplicated(exposures_metabolites$exposure),] %>% pull(exposure)

cases_99 <- vector(mode = "list", length = length(exp_unique))

for(i in seq_along(exp_unique)){ 
  
  exp <- exposures_metabolites %>%
    filter(exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "data/outcomes/cases99rsid.csv",
    sep = ",",
    snp_col = "rsid",
    beta_col = "Beta",
    se_col = "SE",
    effect_allele_col = "EA",
    other_allele_col = "NEA",
    eaf_col = "EAF",
    pval_col = "pval",
    samplesize_col = "effective_N"),
    error=function(e) e)

  if(inherits(outcome, "error") == FALSE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      cases_99[[i]] <- merge(exp, outcome, by = "SNP") %>%
        mutate(error = "Harmonize")
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        cases_99[[i]] <- harmonize %>%
          mutate(error = "FALSE")
  

      }
    

  }else{
    cases_99[[i]] <- tibble(error = "no matching rsids in outcome")
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
  }
    
  print(paste0("Harmonized exposure ", exp_unique[i], 
               " which is iterator ", i)) 
}
 

cases_99_df <- bind_rows(cases_99)
 
saveRDS(error_log, "data/harmonized/error_log_cases_99_metabolites.rds")


saveRDS(cases_99, "data/harmonized/cases_99_metabolites.rds")
saveRDS(cases_99_df, "data/harmonized/cases_99_metabolites_df.rds")
```

1 exposure with no matching rsids



GENES

```{r}
error_log <- data.frame()

exp_unique <- exposures_genes[!duplicated(exposures_genes$exposure),] %>% pull(exposure)

cases_99 <- vector(mode = "list", length = length(exp_unique))

# for(i in seq_along(exp_unique)){ 
for(i in 1:seq_along(exp_unique)){#finished
  
  exp <- exposures_genes %>%
    filter(exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "../data/outcomes/cases99rsid.csv",
    sep = ",",
    snp_col = "rsid",
    beta_col = "Beta",
    se_col = "SE",
    effect_allele_col = "EA",
    other_allele_col = "NEA",
    eaf_col = "EAF",
    pval_col = "pval",
    samplesize_col = "effective_N"),
    error=function(e) e)

  if(!inherits(outcome, "error") == TRUE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      cases_99[[i]] <- merge(exp, outcome, by = "SNP") %>%
        mutate(error = "Harmonize")
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        cases_99[[i]]<- harmonize %>%
          mutate(error = "FALSE")
  

      }
  

  }else{
    cases_99[[i]] <- tibble(error = "no matching rsids in outcome")
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
    
    
  }
  
   

  print(paste0("Harmonized exposure ", exp_unique[i], 
               " which is iterator ", i))    
}


cases_99_DF <- bind_rows(cases_99)
 
saveRDS(error_log, "C:/Users/me/Desktop/MPH/Internship/CPMC/Data/error_log_cases_99_genes.rds")


saveRDS(cases_99, "C:/Users/me/Desktop/MPH/Internship/CPMC/Data/cases_99_genes.rds")
saveRDS(cases_99_df, "..data/harmonized/cases_99_genes_df.rds")
```

METHYLATION
```{r}
error_log <- data.frame()

exp_unique <- exposures_methylation[!duplicated(exposures_methylation$exposure),] %>% pull(exposure)

cases_99<- vector(mode = "list", length = length(exp_unique))

for(i in seq_along(exp_unique)){
# for(i in 120000:140099){
  
  exp <- exposures_methylation %>%
    filter(exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "..data/outcomes/cases99rsid.csv",
    sep = ",",
    snp_col = "rsid",
    beta_col = "Beta",
    se_col = "SE",
    effect_allele_col = "EA",
    other_allele_col = "NEA",
    eaf_col = "EAF",
    pval_col = "pval",
    samplesize_col = "effective_N"),
    error=function(e) e)

  if(!inherits(outcome, "error") == TRUE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      cases_99[[i]] <- merge(exp, outcome, by = "SNP") %>%
        mutate(error = "Harmonize")
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        cases_99[[i]]<- harmonize %>%
          mutate(error = "FALSE")
  

      }
  

  }else{
    cases_99[[i]] <- tibble(error = "no matching rsids in outcome")
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
    
    
  }
  
   

  print(paste0("Harmonized exposure ", exp_unique[i], 
               " which is iterator ", i))    
}

cases_99_DF <- bind_rows(cases_99)
 
saveRDS(error_log, "..data/harmonized/error_log_cases_99_methylation.rds")


saveRDS(cases_99, "..data/harmonized/cases_99_methylation.rds")

saveRDS(cases_99_DF,"..data/harmonized/cases_99_methylation_df.rds")

```

#### d. UK 90



GWAS EXPOSURES

```{r}

#downloaded data



error_log <- data.frame()

exp_unique <- exposures_gwas[!duplicated(exposures_gwas$id.exposure),] %>% pull(id.exposure)

pilling<- vector(mode = "list", length = length(exp_unique))

for(i in 1:1){ 
  
  exp <- exposures_gwas %>%
    filter(id.exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "data/outcomes/29227965-GCST006698-EFO_0007796-build37.f.tsv",
    sep = "\t",
    snp_col = "variant_id",
    beta_col = "beta",
    se_col = "standard_error",
    effect_allele_col = "effect_allele",
    other_allele_col = "other_allele",
    eaf_col = "effect_allele_frequency",
    pval_col = "p_value"),
    error=function(e) e)

  if(inherits(outcome, "error") == FALSE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      pilling[[i]] <- merge(exp, outcome, by = "SNP") %>%
        mutate(error = "Harmonize")
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        pilling[[i]] <- harmonize %>%
          mutate(error = "FALSE")
  

      }
    

  }else{
    pilling[[i]]<- tibble(error = "no matching rsids in outcome")
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
  }
  print(paste0("harmonized exposure", exp_unique[i], 
               "which is iterator ", i))  
}

pilling_df <- bind_rows(pilling)
saveRDS(error_log, "data/harmonized/error_log_pilling_gwas_ieu.rds")


saveRDS(pilling, "data/harmonized/pilling_ls_gwas_ieu.rds")
saveRDS(pilling_df,"data/harmonized/pilling_df_gwas_ieu.rds" )
```

3612 exposures did not have matching rsids in the pilling data

545 harmonizing errors due to NA values




PROTEIN EXPOSURES

```{r}


pilling_download <- data.frame()
error_log <- data.frame()

exp_unique <- exposures_proteins[!duplicated(exposures_proteins$exposure),] %>% pull(exposure)



for(i in 1:47){ 
  
  exp <- exposures_proteins %>%
    filter(exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "data/outcomes/Pilling_90.csv",
    sep = ",",
    snp_col = "rsid",
    beta_col = "beta",
    se_col = "standard_error",
    effect_allele_col = "effect_allele",
    other_allele_col = "other_allele",
    eaf_col = "effect_allele_frequency",
    pval_col = "p_value"),
    error=function(e) e)

  if(!inherits(outcome, "error") == TRUE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      harmonize_dat <- merge(exp, outcome, by = "SNP") %>%
        mutate(remove = FALSE,
               palindromic = FALSE,
               ambiguous = FALSE,
               action = 2,
               mr_keep = FALSE, #can change to true later
               samplesize.outcome = NA,
               error = TRUE)
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        harmonize_dat <- harmonize %>%
          mutate(error = FALSE)
  

      }
    pilling_download <- rbind(pilling_download, harmonize_dat)

  }else{
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
  }
    
}
 
saveRDS(error_log, "data/harmonized/error_log_pilling_protein.rds")


saveRDS(pilling_download, "data/harmonized/pilling_download_protein.rds")
```

40 exposures with no matching SNPs


METABOLITES


```{r}
pilling_download <- data.frame()
error_log <- data.frame()

exp_unique <- exposures_metabolites[!duplicated(exposures_metabolites$exposure),] %>% pull(exposure)



for(i in 1:121){ 
  
  exp <- exposures_metabolites %>%
    filter(exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "data/outcomes/Pilling_90.csv",
    sep = ",",
    snp_col = "rsid",
    beta_col = "beta",
    se_col = "standard_error",
    effect_allele_col = "effect_allele",
    other_allele_col = "other_allele",
    eaf_col = "effect_allele_frequency",
    pval_col = "p_value"),
    error=function(e) e)

  if(!inherits(outcome, "error") == TRUE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      harmonize_dat <- merge(exp, outcome, by = "SNP") %>%
        mutate(remove = FALSE,
               palindromic = FALSE,
               ambiguous = FALSE,
               action = 2,
               mr_keep = FALSE, #can change to true later
               samplesize.outcome = NA,
               error = TRUE)
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        harmonize_dat <- harmonize %>%
          mutate(error = FALSE)
  

      }
    pilling_download <- rbind(pilling_download, harmonize_dat)

  }else{
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
  }
    
}
 
saveRDS(error_log, "data/harmonized/error_log_pilling_metabolites.rds")


saveRDS(pilling_download, "data/harmonized/pilling_download_metabolites.rds")
```
45 exposures with no mathing SNPs in the outcome


GENES

```{r}
error_log <- data.frame()

exp_unique <- exposures_genes[!duplicated(exposures_genes$exposure),] %>% pull(exposure)

pilling_download <- vector(mode = "list", length = length(exp_unique))

for(i in seq_along(exp_unique)){
# for(i in 15001:55000){
  
  exp <- exposures_genes %>%
    filter(exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "data/outcomes/Pilling_90.csv",
    sep = ",",
    snp_col = "rsid",
    beta_col = "beta",
    se_col = "standard_error",
    effect_allele_col = "effect_allele",
    other_allele_col = "other_allele",
    eaf_col = "effect_allele_frequency",
    pval_col = "p_value"),
    error=function(e) e)

  if(!inherits(outcome, "error") == TRUE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      pilling_download[[i]] <- merge(exp, outcome, by = "SNP") %>%
        mutate(error = "Harmonize")
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        pilling_download[[i]]<- harmonize %>%
          mutate(error = "FALSE")
  

      }
  

  }else{
    pilling_download[[i]] <- tibble(error = "no matching rsids in outcome")
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
    
    
  }
  
   

  print(paste0("Harmonized exposure ", exp_unique[i], 
               " which is iterator ", i))    
}


pilling_download_DF <- bind_rows(pilling_download)
 
saveRDS(error_log, "data/harmonized/error_log_pilling_genes.rds")


saveRDS(pilling_download, "data/harmonized/pilling_download_genes.rds")
```




methylation
fixed and added methylation error exposures

```{r}
error_log <- data.frame()

exp_unique <- exposures_methylation[!duplicated(exposures_methylation$exposure),] %>% pull(exposure)

pilling_download <- vector(mode = "list", length = length(exp_unique))

for(i in seq_along(exp_unique)){
# for(i in 120000:140099){
  
  exp <- exposures_methylation %>%
    filter(exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "data/outcomes/Pilling_90.csv",
    sep = ",",
    snp_col = "rsid",
    beta_col = "beta",
    se_col = "standard_error",
    effect_allele_col = "effect_allele",
    other_allele_col = "other_allele",
    eaf_col = "effect_allele_frequency",
    pval_col = "p_value"),
    error=function(e) e)

  if(!inherits(outcome, "error") == TRUE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      pilling_download[[i]] <- merge(exp, outcome, by = "SNP") %>%
        mutate(error = "Harmonize")
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        pilling_download[[i]]<- harmonize %>%
          mutate(error = "FALSE")
  

      }
  

  }else{
    pilling_download[[i]] <- tibble(error = "no matching rsids in outcome")
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
    
    
  }
  
   

  print(paste0("Harmonized exposure ", exp_unique[i], 
               " which is iterator ", i))    
}


pilling_download_DF <- bind_rows(pilling_download)
 
saveRDS(error_log, "data/harmonized/error_log_pilling_methylation.rds")


saveRDS(pilling_download, "data/harmonized/pilling_download_methylation.rds")

saveRDS(pilling_download_DF,"data/harmonized/pilling_download_methylation_df.rds")

```




#### e. UK Parental
GWAS

```{r}
error_log <- data.frame()

exp_unique <- exposures_gwas[!duplicated(exposures_gwas$exposure),] %>% pull(exposure)


for(i in seq_along(exp_unique)){ 
  
  exp <- exposures_proteins %>%
    filter(exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "data/outcomes/lifegen_phase2_bothpl_alldr_2017_09_18.tsv",
    sep = "\t",
    snp_col = "rsid",
    beta_col = "beta1",
    se_col = "se",
    effect_allele_col = "a1",
    other_allele_col = "a0",
    eaf_col = "freq1",
    pval_col = "p"),
    error=function(e) e)

  if(inherits(outcome, "error") == FALSE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      timmers[[i]] <- merge(exp, outcome, by = "SNP") %>%
        mutate(error = "Harmonize")
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        timmers[[i]] <- harmonize %>%
          mutate(error = FALSE)
  

      }
    

  }else{
    pilling_download[[i]] <- tibble(error = "no matching rsids in outcome")
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
  }
    
  print(paste0("Harmonized exposure ", exp_unique[i], 
               " which is iterator ", i)) 
}
 

timmers_df <- bind_rows(timmers)

saveRDS(error_log, "data/harmonized/error_log_timmrs_gwas.rds")


saveRDS(timmers, "data/harmonized/timmers_gwas_ls.rds")
saveRDS(timmers_df, "data/harmonized/timmers_gwas_df.rds")
```

Proteins
```{r}



error_log <- data.frame()

exp_unique <- exposures_proteins[!duplicated(exposures_proteins$exposure),] %>% pull(exposure)

timmers <- vector(mode = "list", length = length(exp_unique))

for(i in seq_along(exp_unique)){ 
  
  exp <- exposures_proteins %>%
    filter(exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "data/outcomes/lifegen_phase2_bothpl_alldr_2017_09_18.tsv",
    sep = "\t",
    snp_col = "rsid",
    beta_col = "beta1",
    se_col = "se",
    effect_allele_col = "a1",
    other_allele_col = "a0",
    eaf_col = "freq1",
    pval_col = "p"),
    error=function(e) e)

  if(inherits(outcome, "error") == FALSE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      timmers[[i]] <- merge(exp, outcome, by = "SNP") %>%
        mutate(error = "Harmonize")
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        timmers[[i]] <- harmonize %>%
          mutate(error = "FALSE")
  

      }
    

  }else{
    pilling_download[[i]] <- tibble(error = "no matching rsids in outcome")
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
  }
    
  print(paste0("Harmonized exposure ", exp_unique[i], 
               " which is iterator ", i)) 
}
 

timmers_df <- bind_rows(timmers)

saveRDS(error_log, "data/harmonized/error_log_timmrs_protein.rds")


saveRDS(timmers, "data/harmonized/timmers_protein_ls.rds")
saveRDS(timmers_df, "data/harmonized/timmers_protein_df.rds")
```


METABOLITES


```{r}

error_log <- data.frame()

exp_unique <- exposures_metabolites[!duplicated(exposures_metabolites$exposure),] %>% pull(exposure)

timmers <- vector(mode = "list", length = length(exp_unique))

for(i in seq_along(exp_unique)){ 
  
  exp <- exposures_metabolites %>%
    filter(exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "data/outcomes/lifegen_phase2_bothpl_alldr_2017_09_18.tsv",
    sep = "\t",
    snp_col = "rsid",
    beta_col = "beta1",
    se_col = "se",
    effect_allele_col = "a1",
    other_allele_col = "a0",
    eaf_col = "freq1",
    pval_col = "p"),
    error=function(e) e)

  if(inherits(outcome, "error") == FALSE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      timmers[[i]] <- merge(exp, outcome, by = "SNP") %>%
        mutate(error = "Harmonize")
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        timmers[[i]] <- harmonize %>%
          mutate(error = "FALSE")
  

      }
    

  }else{
    timmers[[i]] <- tibble(error = "no matching rsids in outcome")
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
  }
    
  print(paste0("Harmonized exposure ", exp_unique[i], 
               " which is iterator ", i)) 
}
 

timmers_df <- bind_rows(timmers)

saveRDS(error_log, "data/harmonized/error_log_timmrs_metabolite.rds")


saveRDS(timmers, "data/harmonized/timmers_metabolite_ls.rds")
saveRDS(timmers_df, "data/harmonized/timmers_metabolite_df.rds")
```





## D. MR Analysis

### I. UK 90 
Downloaded Pilling Data
```{r}
gwas <- readRDS("data/harmonized/Pilling/pilling_download_gwas.rds")
mr_gwas <- gwas %>% 
  filter(error == "FALSE") %>%
  mr()%>%
  mutate(exposure_dat = "GWAS")

protein <-readRDS("data/harmonized/Pilling/pilling_download_protein.rds")
mr_protein <- protein %>%
  filter(error == "FALSE") %>%
  mr() %>%
  mutate(exposure_dat = "protein")

metabolite <- readRDS("data/harmonized/Pilling/pilling_download_metabolites.rds")
mr_metabolite <- metabolite %>%
  filter(error == "FALSE") %>%
  mr()  %>%
  mutate(exposure_dat = "metabolite")

# methylation <- readRDS("data/harmonized/Pilling/pilling_download_methylation_df.rds")
# mr_methylation <- methylation %>%
#   filter(error =="FALSE") %>%
#   mr()%>%
#   mutate(exposure_dat = "methylation")

# genes <- readRDS("data/harmonized/Pilling/pilling_download_genes_df.rds")
# mr_genes <- genes %>%
#   filter(error == "FALSE") %>%
#   mr() %>%
#   mutate(exposure_dat = "genes")


mr <- rbind(mr_gwas, mr_protein, mr_metabolite)

saveRDS(mr, "data/MR/MR_pilling_filterF.rds")
```

Control for multiple testing
```{r}

# Look only at the Inverse variance weighted method for unique exposures
MR <- mr %>%
  filter(method == "Inverse variance weighted") 

mt.pval <- mt.rawp2adjp(MR$pval, proc = "BH", alpha = 0.05)

mt.pval$adjp

MR <- MR %>%
   mutate(pval_BH = mt.pval[["adjp"]][order(mt.pval$index), "BH"])

#check for duplicate significant exposures

dup <- MR %>%
  filter(method == "Inverse variance weighted", pval_BH < 0.05)


result <- dup %>%
    filter(!duplicated(exposure))

saveRDS(MR, "data/MR/Pilling_multtest_F.rds")

```

no duplicates in significant exposures after adjusting for multiple testing

### II. Timmers data

```{r}
gwas <- readRDS("data/harmonized/Timmers/timmers_gwas_df.rds")
mr_gwas <- gwas %>%
  filter(error == "FALSE") %>%
  mr()%>%
  mutate(exposure_dat = "GWAS")

protein <-readRDS("data/harmonized/Timmers/timmers_protein_df.rds")

mr_protein <- protein %>%
  filter(error == "FALSE") %>%
  mr() %>%
  mutate(exposure_dat = "protein")

metabolite <- readRDS("data/harmonized/Timmers/timmers_metabolite_df.rds")
mr_metabolite <- metabolite %>%
  filter(error == "FALSE") %>%
  mr()  %>%
  mutate(exposure_dat = "metabolite")


mr <- rbind(mr_gwas, mr_protein, mr_metabolite)

saveRDS(mr, "data/MR/MR_timmers_F.rds")
```

Control for multiple testing
```{r}

# Look only at the Inverse variance weighted method for unique exposures
MR <- mr %>%
  filter(method == "Inverse variance weighted")

mt.pval <- mt.rawp2adjp(MR$pval, proc = "BH", alpha = 0.05)

mt.pval$adjp

MR <- MR %>%
   mutate(pval_BH = mt.pval[["adjp"]][order(mt.pval$index), "BH"])


#check for duplicate significant exposures

dup <- MR %>%
  filter(method == "Inverse variance weighted", pval_BH < 0.05)


result <- dup%>%
    filter(!duplicated(exposure))

#no duplicate exposures in signficant results where adjusted pvalue exists

saveRDS(MR, "data/MR/Timmers_multtest.rds")
```
### III. Cases 90 from Deleen et al
Deleen cases 90

```{r}
gwas <- readRDS("data/harmonized/Cases90/cases_90_gwas_df.rds")
mr_gwas <- gwas %>%
  filter(error == "FALSE") %>%
  mr()%>%
  mutate(exposure_dat = "GWAS")
saveRDS(mr_gwas, "data/harmonized/Cases90/MR_cases_90_gwas_F.rds")


protein <-readRDS("data/harmonized/Cases90/cases_90_protein_df.rds")
mr_protein <- protein %>%
  filter(error == "FALSE") %>%
  mr() %>%
  mutate(exposure_dat = "protein")

metabolite <- readRDS("data/harmonized/Cases90/cases_90_metabolites_df.rds")
mr_metabolite <- metabolite %>%
  filter(error == "FALSE") %>%
  mr()  %>%
  mutate(exposure_dat = "metabolite")


mr <- rbind(mr_gwas, mr_protein, mr_metabolite)

saveRDS(mr, "data/MR/MR_cases_90_F.rds")
```


Control for multiple testing
```{r}

# Look only at the Inverse variance weighted method for unique exposures
MR <- mr %>%
  filter(method == "Inverse variance weighted")

mt.pval <- mt.rawp2adjp(MR$pval, proc = "BH", alpha = 0.05)

mt.pval$adjp

MR <- MR %>%
   mutate(pval_BH = mt.pval[["adjp"]][order(mt.pval$index), "BH"])

#check for duplicate significant exposures

dup <- MR %>%
  filter(method == "Inverse variance weighted", pval_BH <= 0.05)


result <- dup %>%
    filter(!duplicated(exposure))
#no duplicate exposures for adjusted p-values < = 0.05

saveRDS(MR, "data/MR/Cases90_multtest_F.rds")
```

### IV. Cases 99 from Deleen et al
```{r}
gwas <- readRDS("data/harmonized/Cases99/cases_99_gwas_df.rds")

mr_gwas <- gwas %>%
  filter(error == "FALSE") %>%
  mr() %>%
  mutate(exposure_dat = "GWAS")


protein <-readRDS("data/harmonized/Cases99/cases_99_protein_df.rds")
mr_protein <- protein %>%
  filter(error == "FALSE") %>%
  mr() %>%
  mutate(exposure_dat = "protein")

metabolite <- readRDS("data/harmonized/Cases99/cases_99_metabolites_df.rds")
mr_metabolite <- metabolite %>%
  filter(error == "FALSE") %>%
  mr()  %>%
  mutate(exposure_dat = "metabolite")


mr <- rbind(mr_gwas, mr_protein, mr_metabolite)

saveRDS(mr, "data/MR/MR_cases_99_F.rds")
```

Control for multiple testing
```{r}

# Look only at the Inverse variance weighted method for unique exposures
MR <- mr %>%
  filter(method == "Inverse variance weighted")

mt.pval <- mt.rawp2adjp(MR$pval, proc = "BH", alpha = 0.05)

mt.pval$adjp

MR <- MR %>%
   mutate(pval_BH = mt.pval[["adjp"]][order(mt.pval$index), "BH"])

#check for duplicate significant exposures

dup <- MR %>%
  filter(method == "Inverse variance weighted", pval_BH <= 0.05)


result <- dup %>%
    filter(!duplicated(exposure))

#no duplicate exposures

saveRDS(MR, "data/MR/Cases99_multtest_F.rds")
```


```{r}
sum(exposure_Timmers %>% exposure_case90)/length(exposure_Timmerspar)
```



### V. Check Assumptions
For significant results (adjusted p-value <0.05), we will check MR assumptions to see if further analysis or other MR methods should be used to obtain a better estimate of the relationship between the exposure and longevity, our outcome of interest.


First, lets get the significant exposures
```{r}
library(tidyverse)
library(reactable)
library(TwoSampleMR)
dat <- readRDS( "results/full_results_201908.rds")

timmers_exp <- dat %>%
  filter(Timmers == "YES") %>%
  pull(exposure)

pilling_exp <- dat %>%
  filter(Pilling == "YES") %>%
  pull(exposure)

cases90_exp <- dat %>%
  filter(Cases90 == "YES") %>%
  pull(exposure)

cases99_exp <- dat %>%
  filter(Cases99 == "YES") %>%
  pull(exposure)
```

Then let's pull the exposures from the harmonized datasets
Timmers Dataset
```{r}
gwas <- readRDS("data/harmonized/Timmers/timmers_gwas_df.rds") %>%
  mutate(data_source.exposure = "gwas")
metabolite <- readRDS("data/harmonized/Timmers/timmers_metabolite_df.rds") %>%
  rename(pval.exposure = pval)%>%
  mutate(units.exposure = NA, 
         units.exposure_dat = NA,
         gene.exposure = NA,
         chr.exposure = NA)
protein <- readRDS("data/harmonized/Timmers/timmers_protein_df.rds")%>%
  mutate(units.exposure = NA, 
         units.exposure_dat = NA) %>%
  rename(pval.exposure = pval)

timmers <- rbind(gwas, metabolite, protein)

dat <- timmers %>%
  filter(exposure %in% timmers_exp)
```

Run the assumptions test
```{r}
#Test for heterogeneity

het_test <- mr_heterogeneity(dat)
pleio_test <- mr_pleiotropy_test(dat)
leaveoo <- mr_leaveoneout(dat)
```

Plots 
```{r}
result <-readRDS("data/MR/MR_timmers_F.rds")

p1 <- mr_scatter_plot(result, dat)
p1[8]
```

Cases99
Then let's pull the exposures from the harmonized datasets
Timmers Dataset
```{r}
gwas <- readRDS("data/harmonized/Cases99/cases_99_gwas_df.rds") %>%
  mutate(data_source.exposure = "gwas")
metabolite <- readRDS("data/harmonized/Cases99/cases_99_metabolites_df.rds") %>%
  rename(pval.exposure = pval)%>%
  mutate(units.exposure = NA, 
         units.exposure_dat = NA,
         gene.exposure = NA,
         chr.exposure = NA)
protein <- readRDS("data/harmonized/Cases99/cases_99_protein_df.rds")%>%
  mutate(units.exposure = NA, 
         units.exposure_dat = NA) %>%
  rename(pval.exposure = pval)

cases99<- rbind(gwas, metabolite, protein)

dat <- cases99%>%
  filter(exposure %in% cases99_exp)
```

Run the assumptions test
```{r}
#Test for heterogeneity

het_test <- mr_heterogeneity(dat)
pleio_test <- mr_pleiotropy_test(dat)
leaveoo <- mr_leaveoneout(dat)
```

Plots 
```{r}
result <-readRDS("data/MR/MR_cases_99_F.rds")

p1 <- mr_scatter_plot(result, dat)
p1[5]
```