---
title: "Longevity"
author: "C le"
output: html_document
---



# Project Overview
Two sample Mendelian Randomization Analysis on GWAS of human lifespan looking at new and known exposure relationships related to longevity outcomes.

## A. Exposures

Exposures are obtained from the mrbase.org website from the following catalogs
  - NHGRI-EBI GWAS catalog
  - MR BASE GWAS catalog
  - Gene Expression QTLs
  - Protein Level QTLs
  - Metabolite Level OTLs
  - Methylation Level QTLs

## B. Outcomes

### I. First Outcome: cases 90

Outcome Data obtained from CPMC that was utilized in Deelen et al., 2019 (https://pubmed.ncbi.nlm.nih.gov/31413261/). Exposure - Outcome relationships will be examined in persons whose longevity is in the 90th percentile using the rest of the cohort as the control.


### II. Second outcome: cases 99

Using data from section I, compare persons in the 90th percentile with persons in the 99th percentile. 

### III. Third Outcome: UK Biobank

Compare persons in the 90th percentile from section I to those in the 90th percentile from the UK Biobank

### IV. Fourth Outcome: UK Biobank Parental Longevity
## C. Preparation of Data

### I. Load Libraries

Run this code chunk the first time you use this code to check that you have all the packages needed.
```{r,eval=FALSE}

if(!require(TwoSampleMR)){
  install.packages("devtools")
  devtools::install_github("MRCIEU/TwoSampleMR")
}
  
if(!require(ieugwasr)){
  #if devtools not installed yet, use line below
  #install.packages("devtools")
  devtools::install_github("mrcieu/ieugwasr")
}

if(!require(genetics.binaRies)){
  #if devtools not installed yet, use line below
  #install.packages("devtools")
  devtools::install_github("explodecomputer/genetics.binaRies")
}

is_it_here <- function(package_list){
  for( i in seq_along(package_list)){
    if(!require(i)){
      install.packages(i)
    }
  }
}
packages_CRAN <- c("dplyr", "knitr", "arsenal","tidyverse", "data.table")
is_it_here(packages_CRAN)

````


```{r}
library(TwoSampleMR)
library(MRInstruments)
library(tidyverse) #so we can use purrr::map, and no, not just because I got cats
library(knitr)
library(ieugwasr)
library(arsenal)
library(tidyverse)
```

###II.  Exposures from each catalog
Local LD clumping was done per [reference url] and reference data was obtained from 1000 genomes at [ http://fileserve.mrcieu.ac.uk/ld/1kg.v3.tgz]


Below is the method to do local LD clumping. However in **Section F** there are examples to do it via the TwoSampleMR reference via the internet as well as an example extracting exposures as a function using apply. 

Make sure to set eval = **TRUE** if you want to run that chunk of code. 



Set variables for local clumping of data
```{r,eval=FALSE}
#set working directory
setwd("path/to/file")

# plink library
plink <- genetics.binaRies::get_plink_binary()

# bfile is the filepath were your reference datasets are located
#no need to include extensions for population (e.g EUR.bin)
bfile <- "../data/LD/LD_reference/EUR"
```

#### a. GWAS Catalog

```{r, eval=FALSE}

#cache = True will create a folder that saves objects, and if objects and code aren't updated, this code chunk will not re-run. This is good for long-running code sections.
#import gwas catalog into R
data(gwas_catalog)
gwas_catalog

#create a vector of unique gwas phenotypes
gwas_unique <- gwas_catalog %>%
  filter(!duplicated(Phenotype)) %>%
  pull(Phenotype)
# 3590 unique phenotypes in catalog

#make a for loop to create a database of exposure SNPs formatted for exposure data and clumped
#Initialize a list to save results. The problem with creating an empty dataframe with data.frame() is that it doesn't preallocate memory for 3590 rows. Instead, it creates an empty DF with zero rows.

exposures_gwas <- vector(mode = "list", length = length(gwas_unique))
for(i in seq_along(gwas_unique)){
  #you can test with for(i in 1:5)
# for(i in 1 : length(gwas_unique)) {
  #format exposure data and clump by LD r2 <0.001 to reduce covariance
  #combined both steps together. Send output to each element of list
  exposures_gwas[[i]] <- gwas_catalog %>%
    filter(Phenotype == gwas_unique[i]) %>%
    format_data %>%
    ld_clump(plink_bin = plink, bfile = bfile, 
           clump_kb = 10000, clump_r2 = .001, 
           clump_p = 1) 
  
  print(paste0("Last successful run was ", gwas_unique[i], 
               " which is iterator ", i))
}

#rbind list elements into DF
#old way of doing things
#exposures_gwas_DF <- do.call(rbind, exposures_gwas)
#faster way of doing things with dplyr
exposures_gwas_DF <- bind_rows(exposures_gwas)

#check that 'phenotypes' is the same as the last entry in gwas_unique
# gwas_unique[3590]
# gwas_phenotypes$Phenotype

# 
save(exposures_gwas, file = "~data/exposures_gwas.Rdata")
```


#### b. Gene Expression QTLs
Search for SNPs can be done by gene and tissue or just one. Below exposure instruments are created for each unique gene.

```{r, eval= FALSE}

#import Gene Expression QTLs list
data(gtex_eqtl)
gtex_eqtl


#create a list of unique gene names

genes_unique <- gtex_eqtl[!duplicated(gtex_eqtl$gene_name),] %>% pull(gene_name)

# 32432 unique gene names

genes_table <- gtex_eqtl[!duplicated(gtex_eqtl$gene_name),]

exposures_genes <- data.frame()
for(i in 17583: length(genes_unique)){
  genes <- subset(gtex_eqtl, gene_name == genes_unique[i]) %>%
    format_gtex_eqtl()%>%
    rename(rsid = SNP,
         pval = pval.exposure,
         id = id.exposure)


  genes_format <-  ld_clump(exposures_genes, plink_bin = plink, bfile = bfile,
           clump_kb = 10000, clump_r2 = .001, clump_p = 1) 

  exposures_genes <- rbind(exposures_genes, genes_format)
}

saveRDS(exposures_genes, file = "~data/exposures_genes.rds")
```



#### c. Protein Level QTLs

```{r, eval = FALSE}
#import Protein Level QTLs into R
data(proteomic_qtls)
proteomic_qtls

#create a list of unique protein analyte & make a list

proteins_unique <- proteomic_qtls[!duplicated(proteomic_qtls$analyte),] %>% pull(analyte)


exposures_protein <- data.frame()
for(i in 1 : length(proteins_unique)) {
  
  proteins_phenotypes<- subset(proteomic_qtls, analyte == proteins_unique[i])
  
  #format exposure data and clump by LD r2 <0.001 to reduce covariance
  proteins_format<- proteins_phenotypes %>% format_proteomic_qtls() %>%
    rename(rsid = SNP,
         pval = pval.exposure,
         id = id.exposure)
  
  proteins_clump <- ld_clump(proteins_format, 
                             plink_bin = plink, 
                             bfile = bfile,
                             clump_kb = 10000, clump_r2 = .001, clump_p = 1)

  exposures_protein <- rbind(exposures_protein, proteins_clump)
    
}

exposures_protein <- exposures_protein %>% rename (SNP = rsid, id.exposure = id)

```



#### d. Metabolite Level OTLs


```{r, eval=FALSE}

#import Metabolite Level QTLs into R
data(metab_qtls)
metab_qtls

#create a list of unique metabolomic phenotype & make a list

metabolites_unique <- metab_qtls[!duplicated(metab_qtls$phenotype),] %>% pull(phenotype)
# 121 unique phenotypes in catalog



#make a for loop to create a database of exposure SNPs formatted for exposure data and clumped
exposures_metabolite <- data.frame()
for(i in 1 : length(metabolites_unique)) {
  
  metabolites_phenotypes<- subset(metab_qtls, phenotype == metabolites_unique[i]) %>% 
    format_metab_qtls() %>% 
    rename(rsid = SNP,
         pval = pval.exposure,
         id = id.exposure) %>%
    ld_clump( plink_bin = plink, 
              bfile = bfile,
              clump_kb = 10000, clump_r2 = .001, clump_p = 1)
    

  exposures_metabolite <- rbind(exposures_metabolite, metabolites_phenotypes)
    
}

exposures_metabolite <- exposures_metabolite %>% rename (SNP = rsid, id.exposure = id)  
  
#check that 'phenotypes' is the same as the last entry in metabolites_unique
metabolites_unique[121]
metabolites_phenotypes$phenotype


saveRDS(exposures_metabolite, file = "~data/exposures_metabolites.rds")
```

#### e. Methylation Level QTLs
by cpg site and age

```{r}

#import Methylation Level QTLs into R
data(aries_mqtl)
aries_mqtl

#create a list of unique Methylation  cpg & make a list

methylation_unique <- aries_mqtl[!duplicated(aries_mqtl$cpg),]%>% pull(cpg)
age <- c("Birth", "Adolescence", "Childhood", "Middle age", "Pregnancy")
#  33256 unique cpg sites in catalog
# 5 'ages'

#make a for loop to create a database of exposure SNPs formatted for exposure data and clumped


exposures_methylation <- data.frame()
for (j in 1:5){
  for(i in 1 : 33256) {
  
    methylation_phenotypes<- subset(aries_mqtl, cpg == methylation_unique[i] & age == age[j])%>% 
      format_aries_mqtl() %>% 
      rename(rsid = SNP,
         pval = pval.exposure,
         id = id.exposure) %>%
    ld_clump( plink_bin = plink, 
              bfile = bfile,
              clump_kb = 10000, clump_r2 = .001, clump_p = 1)
    
    
  exposures_methylation <- rbind(exposures_methylation, methylation_phenotypes)
     

    
    
  }


}

exposures_methylation <- exposures_methylation %>% rename (SNP = rsid, id.exposure = id)   
#check that 'phenotypes' is the same as the last entry in metabolites_unique
methylation_unique[33256]
methylation_phenotypes$cpg


saveRDS(exposures_methylation, file = "~data/exposures_methylation.rds")
```

#### f. Other Methods for obtaining exposure

Ex. GWAS Catalog

Utilizing the clump_data function in TwoSampleMR

```{r}
#Retrieve the catalog
data(gwas_catalog)
gwas_catalog

# then create a vector of unique gwas phenotypes
gwas_unique <- gwas_catalog %>%
  filter(!duplicated(Phenotype)) %>%
  pull(Phenotype)

genes_table <- gtex_eqtl[!duplicated(gtex_eqtl$gene_name),]

exposures_genes <- data.frame()
for(i in 17583: length(genes_unique)){
  genes <- subset(gtex_eqtl, gene_name == genes_unique[i])

  #create a separate clumping step in case LD lookup times out
  genes_format <- format_gtex_eqtl(genes) %>% clump_data()

  exposures_genes <- rbind(exposures_genes, genes_format)
}

saveRDS(exposures_genes, file = "~data/exposures_genes.rds")
```



As a function for a list of genes 
```{r}
#As before retrieve the catalog
data(gwas_catalog)
gwas_catalog

# then create a vector of unique gwas phenotypes
gwas_unique <- gwas_catalog %>%
  filter(!duplicated(Phenotype)) %>%
  pull(Phenotype)

# Define a function
gwas_clump <- function(genes){
  gene <- subset(gtex_eqtl, gene_name == genes) %>% format_gtex_eqtl() %>% clump_data()
}


test <- lapply(gwas_unique, genes_clump)
```


### III. Outcomes & Harmonize

#### a. Load Exposures

```{r}
#GWAS
load("C:/Users/me/Desktop/MPH/Internship/CPMC/exposures_gwas.Rdata")

#Proteins
exposures_proteins <- readRDS("C:/Users/me/Desktop/MPH/Internship/CPMC/exposures_proteins.rds")

#Methylation
exposures_methylation <- readRDS( "C:/Users/me/Desktop/MPH/Internship/CPMC/exposures_methylation.rds")

#Genes
exposures_genes <- readRDS( "C:/Users/me/Desktop/MPH/Internship/CPMC/exposures_genes.rds")

exposures_metabolites <- readRDS("C:/Users/me/Desktop/MPH/Internship/CPMC/exposures_metabolites.rds")

# wait for Dan's answer
```


#### b. Cases 90

#### c. Cases 99

#### d. UK 90

DOWNLOADED PILLING DATA

GWAS EXPOSURES

```{r}

#downloaded data


pilling_download <- data.frame()
error_log <- data.frame()

exp_unique <- exposures_gwas[!duplicated(exposures_gwas$exposure),] %>% pull(exposure)



for(i in 1:5917){ 
  
  exp <- exposures_gwas %>%
    filter(exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "C:/Users/me/Desktop/MPH/Internship/CPMC/Data/Pilling_90.csv",
    sep = ",",
    snp_col = "SNP",
    beta_col = "BETA",
    se_col = "SE",
    effect_allele_col = "ALLELE1",
    other_allele_col = "ALLELE0",
    eaf_col = "A1FREQ",
    pval_col = "P_BOLT"),
    error=function(e) e)

  if(!inherits(outcome, "error") == TRUE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome, action = 3),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      harmonize_dat <- merge(exp, outcome, by = "SNP") %>%
        mutate(remove = FALSE,
               palindromic = FALSE,
               ambiguous = FALSE,
               action = 2,
               mr_keep = FALSE, #can change to true later
               samplesize.outcome = NA,
               error = TRUE)
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        harmonize_dat <- harmonize %>%
          mutate(error = FALSE)
  

      }
    pilling_download <- rbind(pilling_download, harmonize_dat)

  }else{
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
  }
    
}
 
saveRDS(error_log, "C:/Users/me/Desktop/MPH/Internship/CPMC/Data/error_log_pilling_gwas.rds")


saveRDS(pilling_download, "C:/Users/me/Desktop/MPH/Internship/CPMC/Data/pilling_download_gwas.rds")
```

3612 exposures did not have matching rsids in the pilling data

545 harmonizing errors due to NA values




PROTEIN EXPOSURES

```{r}


pilling_download <- data.frame()
error_log <- data.frame()

exp_unique <- exposures_proteins[!duplicated(exposures_proteins$exposure),] %>% pull(exposure)



for(i in 1:47){ 
  
  exp <- exposures_proteins %>%
    filter(exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "C:/Users/me/Desktop/MPH/Internship/CPMC/Data/Pilling_90.csv",
    sep = ",",
    snp_col = "SNP",
    beta_col = "BETA",
    se_col = "SE",
    effect_allele_col = "ALLELE1",
    other_allele_col = "ALLELE0",
    eaf_col = "A1FREQ",
    pval_col = "P_BOLT"),
    error=function(e) e)

  if(!inherits(outcome, "error") == TRUE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      harmonize_dat <- merge(exp, outcome, by = "SNP") %>%
        mutate(remove = FALSE,
               palindromic = FALSE,
               ambiguous = FALSE,
               action = 2,
               mr_keep = FALSE, #can change to true later
               samplesize.outcome = NA,
               error = TRUE)
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        harmonize_dat <- harmonize %>%
          mutate(error = FALSE)
  

      }
    pilling_download <- rbind(pilling_download, harmonize_dat)

  }else{
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
  }
    
}
 
saveRDS(error_log, "C:/Users/me/Desktop/MPH/Internship/CPMC/Data/error_log_pilling_protein.rds")


saveRDS(pilling_download, "C:/Users/me/Desktop/MPH/Internship/CPMC/Data/pilling_download_protein.rds")
```

40 exposures with no matching SNPs


METABOLITES


```{r}
pilling_download <- data.frame()
error_log <- data.frame()

exp_unique <- exposures_metabolites[!duplicated(exposures_metabolites$exposure),] %>% pull(exposure)



for(i in 1:121){ 
  
  exp <- exposures_metabolites %>%
    filter(exposure == exp_unique[i])
  
  outcome <- tryCatch( 
    read_outcome_data(
    snps = exp$SNP,
    filename = "C:/Users/me/Desktop/MPH/Internship/CPMC/Data/Pilling_90.csv",
    sep = ",",
    snp_col = "SNP",
    beta_col = "BETA",
    se_col = "SE",
    effect_allele_col = "ALLELE1",
    other_allele_col = "ALLELE0",
    eaf_col = "A1FREQ",
    pval_col = "P_BOLT"),
    error=function(e) e)

  if(!inherits(outcome, "error") == TRUE){

    harmonize <- tryCatch(
    harmonise_data(exp, outcome),
    error=function(e) e)

    if(inherits(harmonize, "error") == TRUE){
      harmonize_dat <- merge(exp, outcome, by = "SNP") %>%
        mutate(remove = FALSE,
               palindromic = FALSE,
               ambiguous = FALSE,
               action = 2,
               mr_keep = FALSE, #can change to true later
               samplesize.outcome = NA,
               error = TRUE)
      error <-tibble(error = harmonize$message, 
               location_i = i,
               exposure = exp_unique[i],
               message = "harmonizing error")
      
      error_log <- rbind(error_log, error)

      }else {
        harmonize_dat <- harmonize %>%
          mutate(error = FALSE)
  

      }
    pilling_download <- rbind(pilling_download, harmonize_dat)

  }else{
    error <- tibble(error = outcome$message,
                    location_i = i,
                    exposure = exp_unique[i],
                    message = "no matching rsids in outcome")
    error_log <- rbind(error_log, error)
  }
    
}
 
saveRDS(error_log, "C:/Users/me/Desktop/MPH/Internship/CPMC/Data/error_log_pilling_metabolites.rds")


saveRDS(pilling_download, "C:/Users/me/Desktop/MPH/Internship/CPMC/Data/pilling_download_metabolites.rds")
```
45 exposures with no mathing SNPs in the outcome

RETRIEVED DATA
```{r}
#retrieved data



pilling <- harmonise_data(
    exposure_dat =exp_pilling_gwas,
    outcome_dat = pilling_out_gwas, action = 2
 )

```

#### e. UK Parental


### IV. MR Analysis

#### a. UK 90 
Downloaded Pilling Data
```{r}
gwas <- readRDS("C:/Users/me/Desktop/MPH/Internship/CPMC/Data/pilling_download_gwas.rds")
mr_gwas <- mr(gwas)%>%
  mutate(exposure_dat = "GWAS")

protein <-readRDS("C:/Users/me/Desktop/MPH/Internship/CPMC/Data/pilling_download_protein.rds")
mr_protein <- mr(protein) %>%
  mutate(exposure_dat = "protein")

metabolite <- readRDS("C:/Users/me/Desktop/MPH/Internship/CPMC/Data/pilling_download_metabolites.rds")
mr_metabolite <- mr(metabolite)  %>%
  mutate(exposure_dat = "metabolite")


mr <- rbind(mr_gwas, mr_protein, mr_metabolite)

saveRDS(mr, "C:/Users/me/Desktop/MPH/Internship/CPMC/Data/MR_pilling.rds")
```

