---
title: "Two Sample Mendelian Randomization on Longevity GWAS"
author: "Christine Le"
date:  "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  slidy_presentation
---
#code folding lookup
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## Outline
- Background
- Methods
- Data Description
- Results
- Discussion
- Conclusion


## Background

```{r}
knitr::include_graphics("images/An_Aging_Nation.png")
```


## Methods

```{r}
knitr::include_graphics("images/2MR_graphic.png")
```


**Two Sample Mendelian Randomization**

- Exposure: SNPs as a proxy for Exposure from one dataset
- Outcome: outcome from another dataset using SNPs from Exposure to assess the causal relationship between Exposure and Outcome, bypassing confounders that influence the E-O relationship.

**Advantages**
  - use publicly available data from two separate studies to inform researchers on the possible relationships between an exposure and outcome
  - simple analyses
  - cost-efficient

citation: Zheng, J., Baird, D., Borges, M. et al. Recent Developments in Mendelian Randomization Studies. Curr Epidemiol Rep 4, 330–345 (2017). https://doi.org/10.1007/s40471-017-0128-6


## Methods for Determining Estimates
- Inverse Variance Weighted
- Egger
- Median Weighted
- Mode Weighted

## Assumption 1

SNPs truly associated with Exposure, as seen from independent studies


**Violations**
- LD
  - clump SNPs that are close to each other
- low power
  - limit SNP-exposure associations to only those that have been validated in multiple studies
- weak gIV
  - solution: limit gIV associated with exposure with high significance in studies from GWAS, and other databases

citation: Zheng, J., Baird, D., Borges, M. et al. Recent Developments in Mendelian Randomization Studies. Curr Epidemiol Rep 4, 330–345 (2017). https://doi.org/10.1007/s40471-017-0128-6

## Assumption 2 


SNPs not associated with E-O confounders
- difficult to prove positively to due complex biological pathways and unknown confounders
- can be tested with known confounders of the Exposure-Outcome relationship

citation: Zheng, J., Baird, D., Borges, M. et al. Recent Developments in Mendelian Randomization Studies. Curr Epidemiol Rep 4, 330–345 (2017). https://doi.org/10.1007/s40471-017-0128-6

## Assumption 3

SNPs are associated with outcome through Exposure only

**Violations**
- Horizontal Pleiotropy
  - when a SNP is related to multiple traits independent of the exposure
  - we can use different methods to try and detect the prescence of horizontal pleiotropy
  - requires adjusted methods to estimate our association of interest even when present 
    - MR Egger 
      - InSIDE(Instrument Strength Independent of Direct Effect)
        assumes no correlation between SNP-exposure association and direct efect of SNP-outcome
      - suffers from low power
      - weak instruments can influence the regression and therefore the estimate
    - Weighted Median approach
      - allows up to 50% of SNPs to be weak or invalid instruments
    - Mode based approach
      - futher relaxes pleiotropy assumption requirements
- weak instruments    
    
**Violations**

- trait heterogeneity
  - difficult to determine without further understanding of biological pathway

citation: Zheng, J., Baird, D., Borges, M. et al. Recent Developments in Mendelian Randomization Studies. Curr Epidemiol Rep 4, 330–345 (2017). https://doi.org/10.1007/s40471-017-0128-6


## Other Limitations

- developmenal compensation /canalization
   - solution: careful interpretation of results, realizing that both environmental and genetic factors influence an outcome.
   - bias towards null
- collider bias
  - be careful in how we interpret the results of our analysis. Based on current knowledge, could the relationship we see in our study be due to other biases like loss to follow up.
- heterogeneity of effect
   - estimates are not consistent across independent instruments
   - use Cochran's Q statistic(IVW) or Rucker's Q statistic(Egger) to test for this


## TWoSampleMR Package

**Software- TwoSampleMR Package**

R package developed to utilize 2-sample Mendelian Randomization on GWAS studies with curated databases. 
Allows researchers to more easily perform 2-sample Mendelian Randomization on GWASs to discover and eludicate possible causal relationships between phenotypes and evaluate assumptions of this statistical strategy. 

source: Gibran Hemani, Jie Zheng, Kaitlin H Wade, Charles Laurin, Benjamin Elsworth, Stephen Burgess, Jack Bowden, Ryan Langdon, Vanessa Tan, James Yarmolinsky, Hashem A. $ The MR-Base platform supports systematic causal inference across the human phenome. eLife 2018. doi: https://doi.org/10.7554/eLife.34408 


## Ways to Assess Violations

**Statistical Tests**
Test for Heterogeneity
 - mr_heterogeneity() function
  - will perform a test of heterogeneity on harmonized data
  - if significant and there is the presence of horizontal pleiotropy MR Egger method would result in a better estimate than IVW. 
  If there is no or negligible horizontal pleiotropy, IVW method i

**Graphically**
Scatter plot 
  - can evaluate whether there are outliers in the IV estimates
  
Funnel Plot

**Steps**
1. Find top SNPs associated with each phenotype/exposure, use ld_clump() function  from the ieugwasr package to ensure instruments/SNPs for an exposure are independent.
Note: can use the clump_data() function for non-local clumping. 
2. For each exposure with associated SNPs, find those SNPs in the outcome data, harmonize the merged data to ensure that the SNP association we are seeing is occurring on the same allele and located on the forward strand. (Aligned to human genome build 37)
3. Perform MR
4. Analyze results for violations of assumptions, correct for multiple testing errors
5. Redo 1-3 to account for any violations found in 4

Code : https://github.com/c-le-ds/Longevity/blob/master/Longevity.Rmd

## Data Description - Exposures
Data obtained using IEU GWAS database

**GWAS **
Curated from the NHGRI-EBI GWAS Catalog of significant associations found in different GWAS studies to include only data that has the required information to perform MR analyis. Catalog was further cleaned so that the unit of report is the same.

**Proteins**

Significant associations from GWAS studies on 47 protein levels obtained from whole blood

**Metabolites**

Database of significant associations on 121 metabolites from GWAS studies provided by Peter Wurtz and Mika Ala-Korpela. 

**Genes**

Signficant associations of tissue-specific gene expression  GWASs from the GTEX study. 


**Methylation**

Significant hits from GWASs on DNA methylation levels performed on mothers and children, with five timepoints during the lifecourse (e.g. child - birth, child - childhood, child - adolescence, mother - pregnancy, mother - middle age).


## Data Description - Outcomes

**Pilling**
Results from GWAS study of parental longevity (combined monther's and father's age) on UK Biobank participants. Participants whose parental longevity was in the top 10 percent were compared with all other UK Biobank participants who met the inclusion criteria. 

**Timmers**
GWAS results from 1 million parental lifespans of 500,000 particpants with data on mortalilty risk factors. 

**Deleen**


## Results - Pilling
```{r, echo=FALSE, results='hide',message=FALSE}
library(tidyverse)
library(knitr)
```
```{r}
readRDS("data/MR/Pilling_multtest.rds")%>%
  filter(method == "Inverse variance weighted", pval_BH < 0.05) %>%
  select(-id.exposure, -id.outcome, -outcome, -method)%>%
  arrange(pval)%>%
  kable()
  
```

## Results - Timmers
```{r}
readRDS("data/MR/Timmers_multtest.rds")%>%
  filter(method == "Inverse variance weighted", pval_BH < 0.05) %>%
  select(-id.exposure, -id.outcome, -outcome, -method)%>%
  arrange(pval)%>%
  kable()
  
```



## Results - Deleen Cases 90
```{r}
readRDS("data/MR/Cases90_multtest.rds")%>%
  filter(method == "Inverse variance weighted", pval_BH < 0.05) %>%
  select(-id.exposure, -id.outcome, -outcome, -method)%>%
  arrange(pval)%>%
  kable()
  
```

## Results - Deleen Cases 99
```{r}
readRDS("data/MR/Cases99_multtest.rds")%>%
  filter(method == "Inverse variance weighted", pval_BH < 0.05) %>%
  select(-id.exposure, -id.outcome, -outcome, -method)%>%
  arrange(pval)%>%
  kable()
  
```



## Common and Unique Exposures in Each Study - Venn Diagram


```{r, echo=FALSE, results='hide',message=FALSE}
library(VennDiagram)
library(tidyverse)
Pilling_dat <- readRDS("data/MR/Pilling_multtest.rds")
Timmers_dat <- readRDS("data/MR/Timmers_multtest.rds")
Cases90_dat <- readRDS("data/MR/Cases90_multtest.rds")
Cases99_dat <- readRDS("data/MR/Cases99_multtest.rds")


```


```{r, eval=FALSE}
venn.diagram(
  x = list(
    Pilling_dat %>% 
      filter(method == "Inverse variance weighted", pval_BH < 0.05) %>% 
      select(exposure) %>% unlist(),
    Timmers_dat %>% 
      filter(method == "Inverse variance weighted", pval_BH < 0.05) %>% 
      select(exposure) %>% unlist(),
    Cases90_dat %>% 
      filter(method == "Inverse variance weighted", pval_BH < 0.05) %>% 
      select(exposure) %>% unlist(),
    Cases99_dat %>% 
      filter(method == "Inverse variance weighted", pval_BH < 0.05) %>% 
      select(exposure) %>% unlist()
  ),
  category.names = c("Pilling", "Timmers", "Deleen 90th", "Deleen 99th"),
  filename = "images/FourOutcomesVenn.png",
  output = TRUE,
          imagetype="png" ,
          height = 900 , 
          width = 1200 , 
          resolution = 300,
          compression = "lzw",
          lwd = 1,
          col=c("#440154ff", '#21908dff', '#fde725ff', '#ff3374'),
          fill = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3), alpha('#ff3374', 0.3))

  
)
```
```{r, out.width = "50%", fig.pos="h"}
knitr::include_graphics("images/FourOutcomesVenn.png")
```



Table of Common and Unique Exposures among Four Outcomes who are significant(adjusted p-value $\leq$ 0.05) using the Inverse Variance Weighted Method

```{r, echo=FALSE, results='hide',message=FALSE}}
library(reactable)
library(tidyverse)
```
```{r}

#Create a table containing adjusted p-values and informationon exposure dataset, and outcome dataset sources
PillingA <- Pilling_dat %>% 
  mutate(Pilling = "Yes", Exposure.Dataset.Pilling = exposure_dat, Adjusted.p.Pilling = pval_BH) %>% 
  select(exposure,  Exposure.Dataset.Pilling, Adjusted.p.Pilling, Pilling) 
TimmersA <- Timmers_dat %>% 
  mutate(Timmers = "Yes", Exposure.Dataset.Timmers = exposure_dat, Adjusted.p.Timmers = pval_BH) %>% 
  select(exposure, Exposure.Dataset.Timmers, Adjusted.p.Timmers, Timmers) 
Cases90A <- Cases90_dat %>% 
  mutate(Cases90 = "Yes", Exposure.Dataset.Cases90 = exposure_dat, Adjusted.p.Cases90 = pval_BH) %>% 
  select(exposure, Exposure.Dataset.Cases90, Adjusted.p.Cases90, Cases90)
Cases99A <- Cases99_dat %>% 
  mutate(Cases99 = "Yes", Exposure.Dataset.Cases99 = exposure_dat, Adjusted.p.Cases99 = pval_BH) %>% 
  select(exposure, Exposure.Dataset.Cases99, Adjusted.p.Cases99, Cases99)

dat <- merge(TimmersA, PillingA, by = "exposure", all = TRUE)
dat <- merge(dat, Cases90A, by = "exposure", all = TRUE)
dat <- merge(dat, Cases99A, by = "exposure", all = TRUE) %>%
  select(exposure, Pilling, Timmers, Cases90, Cases99, Adjusted.p.Pilling, Adjusted.p.Timmers, Adjusted.p.Cases90, Adjusted.p.Cases99, Exposure.Dataset.Pilling, Exposure.Dataset.Timmers, Exposure.Dataset.Cases90, Exposure.Dataset.Cases99)


#replace NAs in Dataset columns with "No"
dat2 <- dat
dat$Pilling <- ifelse(is.na(dat$Pilling), "NO", "YES")
dat$Timmers <- ifelse(is.na(dat$Timmers), "NO", "YES")
dat$Cases90 <- ifelse(is.na(dat$Cases90), "NO", "YES")
dat$Cases99 <- ifelse(is.na(dat$Cases99), "NO", "YES")


#Make one Exposure Dataset column

dat$Exposure.Dataset <- ifelse(dat$Pilling == "YES", dat$Exposure.Dataset.Pilling,
                               ifelse(dat$Timmers == "YES", dat$Exposure.Dataset.Timmers,
                                      ifelse(dat$Cases90 == "YES", dat$Exposure.Dataset.Cases90,
                                             ifelse(dat$Cases99 == "YES", dat$Exposure.Dataset.Cases99, NA)
                                             )
                                      )
                               )
#remove extra exposure dataset columns
dat <- dat %>%
  select(-Exposure.Dataset.Pilling, -Exposure.Dataset.Timmers, -Exposure.Dataset.Cases90, -Exposure.Dataset.Cases99)

#create a filterable table
reactable(dat, filterable = TRUE, paginationType = "jump", defaultPageSize = 20, showSortable = TRUE,
          columns = list(
            exposure = colDef(name = "Exposure"),
            Adjusted.p.Timmers = colDef(name = "Adjusted p-value (Timmers)", 
                                        align = "center", headerStyle = list(background = "#f7f7f8"),
                                        filterable = FALSE),         
            Adjusted.p.Pilling = colDef(name = "Adjusted p-value (Pilling)", 
                                        align = "center",headerStyle = list(background = "#f7f7f8"),
                                        filterable = FALSE),
            Adjusted.p.Cases90 = colDef(name = "Adjusted p-value (Cases90)", 
                                   align = "center",headerStyle = list(background = "#f7f7f8"),
                                   filterable = FALSE),
            Adjusted.p.Cases99 = colDef(name = "Adjusted p-value (Cases99)", 
                                   align = "center",headerStyle = list(background = "#f7f7f8"),
                                   filterable = FALSE),
            Exposure.Dataset = colDef(name = "Exposure Dataset", 
                                              align = "center",headerStyle = list(background = "#f7f7f8")),
            Pilling = colDef(cell = function(value){
              if(value == "NO") "\u2718" else "\u2713"
            }),
            Timmers = colDef(cell = function(value){
              if(value == "NO") "\u2718" else "\u2713"
            }),
            Cases90 = colDef(name = "90th Percentile", cell = function(value){
              if(value == "NO") "\u2718" else "\u2713"
            }),
            Cases99 = colDef(name = "99th Percentile", cell = function(value){
              if(value == "NO") "\u2718" else "\u2713"
            })
          ))
```

## Discussion


##Conclusion

##Limitations and Next Steps