---
title: "Tables Figures and Results"
author: "Christine Le"
date: "11/4/2020"
output:
  html_document: default
  pdf_document: default
  word_document: default
always_allow_html: yes
---


```{r , include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
library(VennDiagram)
library(reactable)
library(tidyverse)
library(kableExtra)
library(TwoSampleMR)
library(patchwork)
library(cowplot)
library(gridExtra)
library(MRInstruments)
library(ggplot2)

#data



Pilling_dat <- readRDS("data/MR/Pilling_multtest_gwas_ieu.rds")
Timmers_dat <- readRDS("data/MR/Timmers_multtest_gwas_ieu.rds")
Cases90_dat <- readRDS("data/MR/Cases90_multtest_gwas_ieu.rds")
Cases99_dat <- readRDS("data/MR/Cases99_multtest_gwas_ieu.rds")
```




```{r results='hide', include=FALSE}
#merge all four outcomes
#Create a table containing adjusted p-values and information on exposure dataset, and outcome dataset sources
PillingA <- Pilling_dat %>% 
  filter(method == "Inverse variance weighted", exposure_dat != "methylation") %>%
  mutate(Pilling = "YES") %>%
  rename(Exposure.Dataset.Pilling = exposure_dat, Adjusted.p.Pilling = pval_BH, 
         b.pilling = b, se.pilling = se, nsnp.pilling = nsnp, pval.pilling = pval ) %>% 
  select(-id.exposure, -id.outcome, -outcome, -method) 
TimmersA <- Timmers_dat %>% 
  filter(method == "Inverse variance weighted") %>%
  mutate(Timmers = "YES") %>%
  rename(Exposure.Dataset.Timmers = exposure_dat, Adjusted.p.Timmers = pval_BH, 
         b.timmers = b, se.timmers = se, nsnp.timmers = nsnp, pval.timmers = pval) %>% 
  select(-id.exposure, -id.outcome, -outcome, -method) 
Cases90A <- Cases90_dat %>%
  filter(method == "Inverse variance weighted") %>%
  mutate(Cases90 = "YES") %>%
  rename(Exposure.Dataset.Cases90 = exposure_dat, Adjusted.p.Cases90 = pval_BH, 
         b.90 = b, se.90 = se, nsnp.90 = nsnp, pval.90 = pval) %>% 
  select(-id.exposure, -id.outcome, -outcome, -method)
Cases99A <- Cases99_dat %>%
  filter(method == "Inverse variance weighted") %>%
  mutate(Cases99 = "YES") %>%
  rename(Exposure.Dataset.Cases99 = exposure_dat, Adjusted.p.Cases99 = pval_BH,
         b.99 = b, se.99 = se, nsnp.99 = nsnp, pval.99 = pval) %>% 
  select(-id.exposure, -id.outcome, -outcome, -method)

dat_all <- merge(TimmersA, PillingA, by = "exposure", all = TRUE)
dat_all <- merge(dat_all, Cases90A, by = "exposure", all = TRUE)
dat_all <- merge(dat_all, Cases99A, by = "exposure", all = TRUE)


#replace NAs in Dataset columns with "NO"

dat_all$Pilling <- ifelse(is.na(dat_all$Pilling), "NO", dat_all$Pilling)
dat_all$Timmers <- ifelse(is.na(dat_all$Timmers), "NO", dat_all$Timmers)
dat_all$Cases90 <- ifelse(is.na(dat_all$Cases90), "NO", dat_all$Cases90)
dat_all$Cases99 <- ifelse(is.na(dat_all$Cases99), "NO", dat_all$Cases99)



#Make one Exposure Dataset column

dat_all$Exposure.Dataset <- ifelse(dat_all$Pilling == "YES", dat_all$Exposure.Dataset.Pilling,
                               ifelse(dat_all$Timmers == "YES", dat_all$Exposure.Dataset.Timmers,
                                      ifelse(dat_all$Cases90 == "YES", dat_all$Exposure.Dataset.Cases90,
                                             ifelse(dat_all$Cases99 == "YES", dat_all$Exposure.Dataset.Cases99, NA)
                                             )
                                      )
                               )






```



# Table1 in Four Parts.



```{r}
# Pilling significant for adjusted p-value <= 0.05 IVW only
table1a <- readRDS("data/MR/Pilling_multtest_gwas_ieu.rds")%>%
  filter(method == "Inverse variance weighted", pval_BH < 0.05) %>%
  filter(nsnp >2)%>%
  select(-id.exposure, -id.outcome, -outcome, -method, - exposure_dat) 

# Timmers significant for adjusted p-value <= 0.05 IVW only
table1b <- readRDS("data/MR/Timmers_multtest_gwas_ieu.rds")%>%
  filter(method == "Inverse variance weighted", pval_BH < 0.05) %>%
  filter(nsnp >2)%>%
  select(-id.exposure, -id.outcome, -outcome, -method, - exposure_dat)

# Deleen 90th percentile significant for adjusted p-value <= 0.05 IVW only
table1c <- readRDS("data/MR/Cases90_multtest_gwas_ieu.rds")%>%
  filter(method == "Inverse variance weighted", pval_BH < 0.05) %>%
  filter(nsnp >2)%>%
  select(-id.exposure, -id.outcome, -outcome, -method, - exposure_dat)

# Deleen 90th percentile significant for adjusted p-value <= 0.05 IVW only
table1d <- readRDS("data/MR/Cases99_multtest_gwas_ieu.rds")%>%
  filter(method == "Inverse variance weighted", pval_BH < 0.05) %>%
  filter(nsnp >2)%>%
  select(-id.exposure, -id.outcome, -outcome, -method, - exposure_dat)

table1 <- rbind(table1a, table1b, table1c, table1d)
# rename columns
colnames(table1)<- c("Exposure", "No of SNPs", "Beta", "SE", "p-value", "adjusted p-value")

##make a table
kbl(table1, booktabs = T, escape = F, row.names = FALSE)%>%
  kable_paper("striped", full_width = T)%>%
  pack_rows("Pilling", 1,353) %>%
  pack_rows("Timmers", 354, 946) %>%
  pack_rows("Deleen 90th Percentile", 947, 1181)%>%
  pack_rows("Deleen 99th Percentile",1182,1320 )
  
  
```

# Supplemntal Table 1. Exposures that are shared among all four studies with at least one significant association (adjusted p-value $\leq$ 0.05)


```{r ,include=FALSE}
#create groups

table5b_groups <- c(
"Neurological" ,                                                                                                                   
"Behavioral",                                                                                                       
"Behavioral",                                                                                                 
"Behavioral",                                                                                            
"Behavioral",                                                                                             
"Behavioral",                                                                                             
"Cardiovascular Risk Factors" ,                                                                                              
"Behavioral",                                                                                            
"Endocrine",                                                                                            
"Endocrine",                                                                                           
"Behavioral",                                                                                            
"Behavioral",                                                                                            
"Endocrine",                                                                       
"Neurological" ,                                                                                         
"Neurological" ,                                                                                             
"Neurological" ,                                                                                    
"Neurological" ,                                                                                                              
"Neurological" ,                                                                                                              
"Neurological" ,                                                                                                             
"Neurological" ,                                                                               
"Neurological" ,                                                                                                   
"Neurological" ,                                                                            
"Neurological" ,                                                                                                         
"Cardiovascular Outcome", 
"Other Body Measures",
"Neurological" ,                                                                                                       
"Other" ,                                                                   
"Cardiovascular Risk Factors" ,                                                                                                  
"Cardiovascular Risk Factors" ,                                                                                                 "Cardiovascular Risk Factors" ,                                                                                                 "Cardiovascular Risk Factors" ,  
"Other Body Measures",                                                                                                       
"Other Body Measures",                                                                                                    
"Other Body Measures",                                                                                                      
"Other Body Measures",                                                                                                     
"Other Body Measures",                                                                                                         
"Other Body Measures",                                                                                                           
"Other Body Measures",                                                                                                          
"Other Body Measures",                                                                                                         
"Other Body Measures",                                                                                                     
"Other Body Measures",                                                                                                  
"Other Body Measures",                                                                                                    
"Other Body Measures",                                                                                                  
"Other Body Measures",                                                                                                     
"Other Body Measures",                                                                                                     
"Other Body Measures",                                                                                                     
"Other Body Measures",                                                                                                  
"Cardiovascular Outcome",          
"Respiratory",                                                                                                                  
"Respiratory" ,                                                                                                                
"Respiratory",                                                                                     
"Cardiovascular Outcome",                                                                                                         "Cardiovascular Outcome",                                                                                                      
"Cardiovascular Outcome",                                                                                                       
"Cardiovascular Outcome",                                                                                                       
"Cardiovascular Outcome",                                                                                               
"Other",                                                                      
"Other",                                                        
"Other",                                                                  
"Other",                                                                                   
"Behavioral",                                                                                          
"Endocrine",                                                                                                             
"Endocrine",  
"Neurological",                                                                                                       
"Other",                                                                                                      
"Other",                                                                                                     
"Respiratory",                           
 "Other",           
"Respiratory",                          
 "Other",    
 "Other",   
"Cardiovascular Risk Factors" ,                                                                                                  "Cardiovascular Risk Factors" ,                                                                                                  
"Cardiovascular Risk Factors" ,                                                                                                  
"Cardiovascular Risk Factors" ,                                                                                                 "Cardiovascular Risk Factors" ,                                                                                                  
"Cardiovascular Risk Factors" ,                                                                                                 "Cardiovascular Risk Factors" ,                                                                                                 "Cardiovascular Risk Factors" ,                                                                                                 "Cardiovascular Risk Factors" ,                                                                                                 "Cardiovascular Risk Factors" ,                                                                                                 "Cardiovascular Risk Factors" ,                                                                                                 "Cardiovascular Risk Factors" ,                                                                                                 "Cardiovascular Risk Factors" ,                                                                                                 "Cardiovascular Risk Factors" ,                                                                                                 "Cardiovascular Risk Factors" ,                                                                                                 "Cardiovascular Risk Factors" ,                                                                                                 "Cancer",                                                                
"Cancer",                                                                                                          
"Cancer",                                                                                                          
"Cancer",                                                                                                         
"Cancer",                                                                                                     
"Cancer",                                                                                                                   
"Cancer",                                                                                                           
"Cancer",                                                                                                                   
"Biomarker",                                                                                                     
"Cancer",                                                                                        
"Cancer",                                                                                       
"Cancer",                                                                                      
"Cancer",                                                                                    
"Cancer",                                                                                                
"Cardiovascular Outcome",                                                                            
"Cardiovascular Outcome",                                                                            
"Cardiovascular Outcome",                                                                                                    
"Gastrointestinal",                                                                                                           
"Gastrointestinal",                                                                                                               
"Gastrointestinal",                                                                                                                
"Behavioral",                                                                                                        
"Neurological" ,                                                                                                               
"Behavioral",                                                                                                                
"Cardiovascular Risk Factors" ,                                                                                                            
"Cardiovascular Risk Factors" ,                                                                                                              
"Cardiovascular Risk Factors" ,                                                                                                             
"Cardiovascular Risk Factors" ,                                                                                                       
"Cardiovascular Risk Factors" ,                                                                                                      
"Behavioral",                                                                                                             
"Behavioral",                                                                   
"Neurological" ,                                                                                                    
"Gastrointestinal",                                                                                                              
"Cardiovascular Risk Factors" ,                                                                                                   
"Cardiovascular Risk Factors" ,                                                                                                
"Other",                                                                                            
"Respiratory",                                                                                        
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
                                                                                               
"Cardiovascular Outcome", 
                                                                                                   
"Cardiovascular Outcome", 
                                                                                                           
"Cardiovascular Outcome", 
                                                                                                            
"Cardiovascular Outcome", 
                                                                                                            
"Cardiovascular Outcome", 
                                                                                                           
"Cardiovascular Outcome", 
                                                                               
"Behavioral",                                                                                                          
"Behavioral",                                                                                                        
"Biomarker",                                                                                                              
"Biomarker",                                                                                   
"Neurological" ,                                                                                                             
"Neurological" ,                                                                                                       
"Other",
"Other",           
"Endocrine"  ,                                                                                                  
"Endocrine"  ,                                                                                                     
"Endocrine"  ,            
"Cancer",                                                                         
"Other",               
"Cancer",                                                                      
"Cancer",                                                                    
"Cancer",                                                                         
"Ocular" ,                                                                          
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Respiratory", 
"Respiratory", 
"Cardiovascular Outcome", 
"Musculoskeletal",                     
"Respiratory",                                                                                        
"Gastrointestinal",
"Gastrointestinal",
"Musculoskeletal",                                                                 
"Musculoskeletal",                                                         
"Endocrine"   ,                                                                          
"Renal",
"Other"     ,                                                   
"Endocrine",                                                               
"Other"     , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Respiratory", 
"Musculoskeletal",           
"Behavioral",
"Cardiovascular Risk Factors" , 
"Cancer", 
"Cardiovascular Risk Factors" , 
"Cardiovascular Drug",
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Behavioral",
"Behavioral",
"Cardiovascular Risk Factors" , 
"Other",                         
"Ocular",                                                                                                            
"Respiratory", 
"Behavioral",
"Behavioral",
"Neurological" , 
"Other"  ,                                                                    
"Other"  ,                                                                                                                                          
"Cardiovascular Outcome", 
"Other"  ,                                                                    
"Other"  ,                                                                            
"Cancer", 
"Cancer", 
"Cancer", 
"Cancer", 
"Cancer", 
"Ocular",      
"Behavioral",               
"Neurological"  ,                                                                                                
"Other"  , 
"Cardiovascular Risk Factors" , 
"Ocular",          
"Other"  ,                 
"Other"  ,                 
"Neurological" , 
"Parental Longevity",
"Parental Longevity",
"Neurological"  ,                                                                                                               "Neurological"  ,                                                                                                               "Neurological"  ,                                                                                                               
"Neurological" , 
"Neurological" , 
"Respiratory", 
"Respiratory", 
"Behavioral",
"Other",
"Other",            
"Neurological"  ,               
"Behavioral",
"Biomarker",                                                                                                                   
"Biomarker",                                                                                                   
"Biomarker",                                                                                                     
"Musculoskeletal",                                                                                
"Musculoskeletal",                                                                                         
"Biomarker",                                                                                                                                                                                                             
"Biomarker",                                                                                                                                                                                                                  
"Musculoskeletal",
"Musculoskeletal",
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Other",                                                                                                                    
"Other",                                                                                                                 
"Biomarker",                                                                                                                 
"Biomarker",                                                                                                        
"Biomarker",                                                                                     
"Biomarker",                                                                                    
"Biomarker",                                                                                 
"Other",                                                                                                              
"Other",                                                                                                              
"Other",                                                                                                            
"Other",                                                                                                          
"Other",                                                                                                          
"Other",                                                                                                         
"Other",                                                                   
"Other",                                                   
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Neurological" , 
"Neurological" , 
"Neurological" , 
"Neurological" , 
"Neurological" , 
"Neurological" , 
"Neurological" , 
"Neurological" , 
"Neurological" , 
"Neurological" , 
"Neurological" ,                
"Neurological" ,                 
"Cardiovascular Outcome",                                                                                           
"Cardiovascular Outcome", 
"Cardiovascular Risk Factors" ,                                                                                        
"Cardiovascular Risk Factors" ,                                                                                    
"Cancer",                                                                                          
"Cancer",
"Other",
"Other",                                                                            
"Cancer",                                                                                     
"Neurological" ,                                                                              
"Neurological" ,                                                                              
"Cancer",                                                                                          
"Cancer",                                                                                            
"Respiratory",                                                                             
"Cardiovascular Outcome",                                                                                             
"Cardiovascular Outcome",                                                                                            
"Cardiovascular Risk Factors" , 
"Other",                                                                               
"Other",                
"Other",                 
"Cancer",                 
"Endocrine" ,                                                                                              
"Cardiovascular Outcome",                
"Cardiovascular Outcome",                
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" ,                  
"Other",                
"Other",                 
"Other",               
"Other",                
"Other",                
"Other",                
"Other",                
"Other",                 
"Other",                 
"Other",                  
"Other",                 
"Other",                 
"Gastrointestinal",
"Endocrine" ,                  
"Endocrine" ,                  
"Neurological" , 
"Other",                  
"Biomarker"  ,                                                                                                                         
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Behavioral",                
"Behavioral",                
"Behavioral",                
"Behavioral",                
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Other",                
"Other",                
"Other",                 
"Other",                
"Other",                 
"Other",                 
"Other",                  
"Other",                 
"Other",                  
"Other",                   
"Other",                  
"Other",                 
"Other",                  
"Other",                 
"Other",                  
"Other",                  
"Behavioral",
"Behavioral",
"Behavioral",
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Other",                  
"Other",                  
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Ocular" ,                                                                        
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cancer", 
"Cancer", 
"Cancer", 
"Cancer", 
"Neurological",                                                                                  
"Behavioral",
"Behavioral",              
"Cardiovascular Risk Factors" , 
"Biomarker" ,                                                                                       
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Respiratory", 
"Respiratory",                                                                        
"Behavioral",                 
"Behavioral",
"Behavioral",
"Behavioral",
"Behavioral",
"Behavioral",
"Neurological" ,                                                                                                                     
"Neurological" ,                                                                                                               
"Other" ,                                                                                                                 
"Musculoskeletal",
"Musculoskeletal",
"Musculoskeletal",
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Neurological" , 
"Neurological" , 
"Neurological" , 
"Biomarker"  ,                                                                                  
"Behavioral",
"Behavioral",
"Other" ,                                                                                          
"Other" ,                                                                                                          
"Other" ,                                                                                                                
"Other" ,                                                                                                                
"Cardiovascular Risk Factors" , 
                                                                                  
"Respiratory", 
"Endocrine"  ,                                                                               
"Cardiovascular Outcome", 
"Cardiovascular Risk Factors" , 
                                                                       
"Cardiovascular Risk Factors" , 
                                                                           
 "Endocrine"  ,                                                           
"Respiratory", 
"Cardiovascular Risk Factors" , 
                                                                            
"Respiratory", 
"Cardiovascular Outcome", 
"Other" ,
"Endocrine"  ,                                                                         
"Cardiovascular Outcome", 
"Cardiovascular Risk Factors" , 
                                                                     
"Cardiovascular Risk Factors" , 
"Endocrine"  ,                                                     
"Endocrine"  ,                                                          
"Respiratory", 
"Musculoskeletal",
"Behavioral",
"Behavioral",
"Behavioral",
"Behavioral",
"Other" ,                                                                                                      
"Other" ,                                                                                            
"Cancer",                                                                                               
"Cancer",                                                                                   
"Cancer",                                                                                    
"Other" ,                                                                                         
"Other" ,                                                                                        
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Other" , 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Musculoskeletal",
"Musculoskeletal",
"Other" ,                                                                                                    
"Other" ,                                                               
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Renal",
"Other" ,                                                
"Other" ,                 
 "Other" ,                  
"Other" ,                 
"Musculoskeletal",                 
"Musculoskeletal",              
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Musculoskeletal",
"Musculoskeletal",                 
"Other" ,                
"Other" ,                
"Cardiovascular Risk Factors" , 
"Behavioral",
"Behavioral",
"Behavioral",
"Behavioral",
"Other" ,                
"Parental Longevity",
"Parental Longevity",
"Parental Longevity",
"Parental Longevity",
"Parental Longevity",
"Parental Longevity",
"Behavioral",
"Behavioral",
"Neurological" ,                                                                                
"Other" ,                
"Respiratory" ,                                                                                                   
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Other" ,                 
"Other" ,                
"Cancer", 
"Respiratory", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Neurological" , 
"Neurological" , 
"Neurological" , 
"Neurological" , 
"Neurological" , 
"Neurological" , 
"Neurological" , 
"Neurological" , 
"Neurological" , 
"Biomarker",                                                                                                       
"Biomarker",                 
"Musculoskeletal",
"Musculoskeletal",
"Musculoskeletal",
"Musculoskeletal",
"Musculoskeletal",
"Musculoskeletal",
"Musculoskeletal",
"Musculoskeletal",
"Behavioral",
"Neurological" ,                 
"Neurological" ,                  
"Other" ,                 
"Other" ,          
"Other" ,                
"Other" ,                
"Other" ,                
"Behavioral",
"Behavioral",
"Other" ,                 
"Other" ,                 
"Cancer", 
"Neurological" ,                 
"Neurological" ,                  
"Other" ,                  
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Other" ,                  
"Other" ,                 
"Behavioral",
"Behavioral",
"Behavioral",                                                                                                   
"Behavioral",                                                                                                  
"Behavioral",                                                                                                       
"Behavioral",                                                                                                     
"Behavioral",                                                                                               
"Behavioral",                                                                                          
"Cardiovascular Risk Factors" , 
                                                                                                               
"Cardiovascular Risk Factors" , 
                                                                                                               
"Cardiovascular Risk Factors" , 
                                                                                                            
"Other" ,                  
"Other" ,                
"Behavioral",                 
"Cardiovascular Risk Factors" ,                  
"Other" ,                  
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Endocrine",                                                                              
"Endocrine",               
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Cardiovascular Drug",
"Endocrine",                 
"Cardiovascular Drug",
"Respiratory",                                                                           
"Cardiovascular Drug",
"Cardiovascular Drug",
"Respiratory",                
"Respiratory",                 
"Cardiovascular Drug",
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Other",                                                                                                            
"Other",                 
"Other",                 
"Other",                  
"Other",                  
"Other",                 
"Other",                  
"Other",                  
"Endocrine",                 
"Endocrine",                 
"Endocrine",                   
"Endocrine",                    
"Endocrine",                  
"Endocrine",                  
"Endocrine",                   
"Endocrine",                 
"Endocrine",                 
"Endocrine",                  
"Endocrine",                   
"Endocrine", 
"Endocrine",                  
"Cancer", 
"Cancer", 
"Cancer", 
"Behavioral",
"Behavioral",
"Behavioral",
"Behavioral",
"Behavioral",
"Behavioral",
"Behavioral",
"Behavioral",
"Behavioral",
"Behavioral",
"Other",               
"Renal",
"Renal",
"Renal",
"Behavioral",
"Behavioral",
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Cardiovascular Outcome", 
"Other",                
"Other",                
"Behavioral",                 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Behavioral",
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Respiratory", 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Cardiovascular Risk Factors" , 
"Other",              
 "Neurological",                                                                                               
"Neurological"  ,                                                                                                          
"Neurological"   ,                                                                                                        
 "Neurological",             
 "Neurological",                   
 "Neurological" )
```

```{r, include= FALSE}
table.1 <- dat_all %>% filter(Pilling == "YES" & Timmers == "YES" & Cases90 == "YES" & Cases99 == "YES") %>% 
  filter(Adjusted.p.Pilling <= 0.05 | Adjusted.p.Timmers <= 0.05 | Adjusted.p.Cases90 <= 0.05 | Adjusted.p.Cases99 <= 0.05) %>%
  mutate(Groups = table5b_groups)%>%
  filter(nsnp.pilling > 2& nsnp.timmers >2 & nsnp.90 > 2 & nsnp.99 > 2)%>%
  select(Groups, exposure,  b.pilling, b.timmers,  b.90, b.99, se.pilling, se.timmers, se.90, se.99, nsnp.pilling, nsnp.timmers, nsnp.90, nsnp.99,pval.pilling, pval.timmers,  pval.90,  pval.99, Adjusted.p.Pilling,  Adjusted.p.Timmers,  Adjusted.p.Cases90,  Adjusted.p.Cases99, Exposure.Dataset)
```




```{r, include= FALSE}
#conditional formatting

table.1$Adjusted.p.Pilling <- round(table.1$Adjusted.p.Pilling, digits = 5)%>% 
  cell_spec(color = ifelse(table.1$Adjusted.p.Pilling > 0.05,"red", "green"))

table.1$Adjusted.p.Timmers <- round(table.1$Adjusted.p.Timmers, digits = 5)%>% 
  cell_spec( color = ifelse(table.1$Adjusted.p.Timmers > 0.05,"red", "green"))

table.1$Adjusted.p.Cases90<- round(table.1$Adjusted.p.Cases90, digits = 5)%>% 
  cell_spec( color = ifelse(table.1$Adjusted.p.Cases90 > 0.05,"red", "green"))

table.1$Adjusted.p.Cases99 <- round(table.1$Adjusted.p.Cases99, digits = 5)%>% 
  cell_spec(color = ifelse(table.1$Adjusted.p.Cases99 > 0.05,"red", "green"))
```




```{r, include= FALSE}
#label columns
colnames(table.1) <- c("Groups", "Exposure", "Pilling", "Timmers", "90th Percentile", "99th Percentile","Pilling", "Timmers", "90th Percentile", "99th Percentile","Pilling", "Timmers", "90th Percentile", "99th Percentile","Pilling", "Timmers", "90th Percentile", "99th Percentile", "Pilling", "Timmers", "90th Percentile", "99th Percentile","Exposure Dataset")

```




```{r}
#create table
new_tab <- table.1[order(table.1$Groups),]

kbl(new_tab,booktabs = T, escape = F, row.names = FALSE, digits = 4) %>%
    add_header_above(c(" " = 2, "Beta"= 4, "SE"= 4,"No. SNPs" =4, "p-value" = 4, "Adjusted p-value"= 4, " " = 1))%>%
  kable_paper("striped", full_width = T)

```


# Table 2.Selected Exposures

```{r}
sel_tab <- dat_all %>% filter(Pilling == "YES" & Timmers == "YES" & Cases90 == "YES" & Cases99 == "YES") %>%
  filter(Adjusted.p.Pilling <= 0.05 | Adjusted.p.Timmers <= 0.05 | Adjusted.p.Cases90 <= 0.05 | Adjusted.p.Cases99 <= 0.05) %>%
  mutate(Groups = table5b_groups)%>%
  select(Groups, exposure,  b.pilling, b.timmers,  b.90, b.99, se.pilling, se.timmers, se.90, se.99, nsnp.pilling, nsnp.timmers, nsnp.90, nsnp.99,pval.pilling, pval.timmers,  pval.90,  pval.99, Adjusted.p.Pilling,  Adjusted.p.Timmers,  Adjusted.p.Cases90,  Adjusted.p.Cases99, Exposure.Dataset)%>%
  filter(nsnp.pilling > 100& nsnp.timmers >100 & nsnp.90 > 100 & nsnp.99 > 100) %>%
  filter(exposure %in% c("Basal metabolic rate || id:ukb-b-16446", "Age first had sexual intercourse || id:ukb-b-6591", 
                         "Overall health rating || id:ukb-b-6306", 
                         "Non-cancer illness code, self-reported: hypertension || id:ukb-b-14057",
                         "Arm fat mass (left) || id:ukb-a-287", "Glycated haemoglobin || id:ukb-d-30750_irnt",
                         "Prostate cancer || id:ieu-b-85", "Coronary artery disease || id:ebi-a-GCST005194", 
                         "Type 2 diabetes || id:ebi-a-GCST006867", "Years of schooling || id:ieu-a-1239"))

 sel_graph <- sel_tab
sel_tab$Adjusted.p.Pilling <- round(sel_tab$Adjusted.p.Pilling, digits = 5) %>% cell_spec( color = ifelse(sel_tab$Adjusted.p.Pilling > 0.05,"red", "green"))

sel_tab$Adjusted.p.Timmers <- round(sel_tab$Adjusted.p.Timmers, digits = 5) %>%
  cell_spec( color = ifelse(sel_tab$Adjusted.p.Timmers < 0.05,"green", "red"))

sel_tab$Adjusted.p.Cases90<- round(sel_tab$Adjusted.p.Cases90, digits = 5) %>% cell_spec( color = ifelse(sel_tab$Adjusted.p.Cases90 > 0.05,"red", "green"))

sel_tab$Adjusted.p.Cases99 <- round(sel_tab$Adjusted.p.Cases99, digits = 5) %>% cell_spec( color = ifelse(sel_tab$Adjusted.p.Cases99 > 0.05,"red", "green"))


colnames(sel_tab) <- c("Groups", "Exposure", "Pilling", "Timmers", "90th Percentile", "99th Percentile","Pilling", "Timmers", "90th Percentile", "99th Percentile","Pilling", "Timmers", "90th Percentile", "99th Percentile", "Pilling", "Timmers", "90th Percentile", "99th Percentile","Pilling", "Timmers", "90th Percentile", "99th Percentile", "Exposure Dataset")

kbl(sel_tab,booktabs = T, escape = F, row.names = FALSE) %>%
  add_header_above(c(" " = 2, "Beta"= 4, "SE"= 4, "No. of SNPS" = 4, "p-value" = 4, "Adjusted p-value"= 4, " " = 1)) %>%
  kable_paper("striped", full_width = T)


```


# Functions
## for 2x2 LOO plots
Modify function from TwoSampleMR Package

```{r }
res <-  function(d, x, title) # must use data filtered by outcome
	{
    d <- dplyr::filter(d, `Longevity Definition` == x)
		d <- plyr::mutate(d)
		# Need to have at least 3 SNPs because IVW etc methods can't be performed with fewer than 2 SNPs
		if(sum(!grepl("All", d$SNP)) < 3) {
			return(
				ggplot() +
  annotate("text", x = 0, y = 0, label = "Insufficient number of SNPs")+
  theme(legend.position = "none",
		        axis.text.y = element_text(size =6),
		        axis.ticks.y = element_line(size =0),
		        axis.title.x = element_text(size =8),
		        panel.border = element_blank(),
		        panel.grid.major = element_blank(),
		        panel.grid.minor = element_blank(),
		        panel.background = element_blank())+
		  labs(y = "", title = title))

		}
		else {
		d$up <- d$Beta + 1.96 * d$SE
		d$lo <- d$Beta - 1.96 * d$SE
		d$tot <- 1
		d$tot[d$SNP != "All"] <- 0.01
		d$SNP <- as.character(d$SNP)
		nom <- d$SNP[d$SNP != "All"]
		nom <- nom[order(d$Beta)]
		d <- rbind(d, d[nrow(d),])
		d$SNP[nrow(d)-1] <- ""
		d$Beta[nrow(d)-1] <- NA
		d$up[nrow(d)-1] <- NA
		d$lo[nrow(d)-1] <- NA
		d$SNP <- factor(d$SNP, levels=c("All" , "", nom))


		plot <- ggplot(d, aes(x = Beta, y = SNP)) +
		  geom_vline(xintercept = 0, linetype = "dotted") +
		  geom_errorbarh(aes(xmin = lo, xmax = up, size = as.factor(tot), colour = as.factor(tot)), height = 0, na.rm = TRUE) +
		  geom_point(aes(color = as.factor(tot)),  na.rm = TRUE)+
		  geom_hline(aes(yintercept = which(levels(SNP) %in% "")), color = "grey") +
		  scale_colour_manual(values = c("black", "red")) +
		  scale_size_manual(values =  c(0.3,1))+
		  theme(legend.position = "none",
		        axis.text.y = element_text(size =6),
		        axis.ticks.y = element_line(size =0),
		        axis.title.x = element_text(size =8),
		        panel.border = element_blank(),
		        panel.grid.major = element_blank(),
		        panel.grid.minor = element_blank(),
		        panel.background = element_blank())+
		  labs(y = "", title = title)
		return(plot)
		}
}
```

Make a function for all four outcomes!

```{r }
#doo for all individual LOO tables 1 -15

loo_graphs <- function(loo_test, exposure, outcome){
library(patchwork)
plot1 <- res(loo_test, "Pilling", "(a) Pilling")
plot2 <- res(loo_test, "Timmers", "(b) Timmers")
plot3 <- res(loo_test, "Cases90", "(c) Deleen 90th Percentile")
plot4 <- res(loo_test, "Cases99", "(d) Deleen 90th Percentile")

plot1 + plot2 + plot3 + plot4 + plot_layout(nrow = 2) +
  plot_annotation(caption = paste0("MR leave-one-out sensitivity analysis for ", exposure, " on ", outcome))
}
```

## Function for scatter plot of three methods 2x2

### Legend
```{r}
# Modify mr_scatter_plot function for our purposes

mrscatter_legend <- function(mr_results, dat, title)
{
	# dat <- subset(dat, paste(id.outcome, id.exposure) %in% paste(mr_results$id.outcome, mr_results$id.exposure))
	requireNamespace("ggplot2", quietly=TRUE)
	requireNamespace("plyr", quietly=TRUE)
	mrres <- plyr::dlply(dat, c("id.exposure", "id.outcome"), function(d)
	{
		d <- plyr::mutate(d)
		if(nrow(d) < 2 | sum(d$mr_keep) == 0)
		{
			return(blank_plot("Insufficient number of SNPs"))
		}
		d <- subset(d, mr_keep)
		index <- d$beta.exposure < 0
		d$beta.exposure[index] <- d$beta.exposure[index] * -1
		d$beta.outcome[index] <- d$beta.outcome[index] * -1
		mrres <- subset(mr_results, id.exposure == d$id.exposure[1] & id.outcome == d$id.outcome[1])
		mrres$a <- 0
		if("MR Egger" %in% mrres$method)
		{
			temp <- mr_egger_regression(d$beta.exposure, d$beta.outcome, d$se.exposure, d$se.outcome, default_parameters())
			mrres$a[mrres$method == "MR Egger"] <- temp$b_i
		}

		if("MR Egger (bootstrap)" %in% mrres$method)
		{
			temp <- mr_egger_regression_bootstrap(d$beta.exposure, d$beta.outcome, d$se.exposure, d$se.outcome, default_parameters())
			mrres$a[mrres$method == "MR Egger (bootstrap)"] <- temp$b_i
		}

		ggplot2::ggplot(data=d, ggplot2::aes(x=beta.exposure, y=beta.outcome)) +
			ggplot2::geom_errorbar(ggplot2::aes(ymin=beta.outcome-se.outcome, ymax=beta.outcome+se.outcome), colour="grey", width=0) +
			ggplot2::geom_errorbarh(ggplot2::aes(xmin=beta.exposure-se.exposure, xmax=beta.exposure+se.exposure), colour="grey", height=0) +
			ggplot2::geom_point(ggplot2::aes(text=paste("SNP:", SNP))) +
			ggplot2::geom_abline(data=mrres, ggplot2::aes(intercept=a, slope=b, colour=method), show.legend=TRUE) +
			ggplot2::scale_colour_manual(values=c("#a6cee3", "#1f78b4", "#b2df8a", "#33a02c", "#fb9a99", "#e31a1c", "#fdbf6f", "#ff7f00", "#cab2d6", "#6a3d9a", "#ffff99", "#b15928")) +
			ggplot2::labs(colour="MR Test", x = "", y=paste("SNP effect on", d$outcome[1]), title = title) +
			ggplot2::theme(legend.position= c(-.1,-.3), legend.key = element_rect(fill = alpha("white" , 0.1)), legend.background = element_rect(fill = alpha("white" , 0.1)), legend.title = element_text(size = 4), legend.text = element_text(size =4)) +
			ggplot2::guides(colour=ggplot2::guide_legend(ncol=3))
	})
	mrres
}


blank_plot <- function(message)
{
	requireNamespace("ggplot2", quietly=TRUE)
	ggplot2::ggplot(data.frame(a=0,b=0,n=message)) + ggplot2::geom_text(ggplot2::aes(x=a,y=b,label=n)) + ggplot2::labs(x=NULL,y=NULL) + ggplot2::theme(axis.text=ggplot2::element_blank(), axis.ticks=ggplot2::element_blank())
}



```
### No legend

```{r}
mrscatter_nolegend <- function(mr_results, dat, title)
{
	# dat <- subset(dat, paste(id.outcome, id.exposure) %in% paste(mr_results$id.outcome, mr_results$id.exposure))
	requireNamespace("ggplot2", quietly=TRUE)
	requireNamespace("plyr", quietly=TRUE)
	mrres <- plyr::dlply(dat, c("id.exposure", "id.outcome"), function(d)
	{
		d <- plyr::mutate(d)
		if(nrow(d) < 2 | sum(d$mr_keep) == 0)
		{
			return(blank_plot("Insufficient number of SNPs"))
		}
		d <- subset(d, mr_keep)
		index <- d$beta.exposure < 0
		d$beta.exposure[index] <- d$beta.exposure[index] * -1
		d$beta.outcome[index] <- d$beta.outcome[index] * -1
		mrres <- subset(mr_results, id.exposure == d$id.exposure[1] & id.outcome == d$id.outcome[1])
		mrres$a <- 0
		if("MR Egger" %in% mrres$method)
		{
			temp <- mr_egger_regression(d$beta.exposure, d$beta.outcome, d$se.exposure, d$se.outcome, default_parameters())
			mrres$a[mrres$method == "MR Egger"] <- temp$b_i
		}

		if("MR Egger (bootstrap)" %in% mrres$method)
		{
			temp <- mr_egger_regression_bootstrap(d$beta.exposure, d$beta.outcome, d$se.exposure, d$se.outcome, default_parameters())
			mrres$a[mrres$method == "MR Egger (bootstrap)"] <- temp$b_i
		}

		ggplot2::ggplot(data=d, ggplot2::aes(x=beta.exposure, y=beta.outcome)) +
			ggplot2::geom_errorbar(ggplot2::aes(ymin=beta.outcome-se.outcome, ymax=beta.outcome+se.outcome), colour="grey", width=0) +
			ggplot2::geom_errorbarh(ggplot2::aes(xmin=beta.exposure-se.exposure, xmax=beta.exposure+se.exposure), colour="grey", height=0) +
			ggplot2::geom_point(ggplot2::aes(text=paste("SNP:", SNP))) +
			ggplot2::geom_abline(data=mrres, ggplot2::aes(intercept=a, slope=b, colour=method), show.legend=TRUE) +
			ggplot2::scale_colour_manual(values=c("#a6cee3", "#1f78b4", "#b2df8a", "#33a02c", "#fb9a99", "#e31a1c", "#fdbf6f", "#ff7f00", "#cab2d6", "#6a3d9a", "#ffff99", "#b15928")) +
			ggplot2::labs(colour="MR Test", x=paste(""), y=paste("SNP effect on", d$outcome[1]), title = title) +
			ggplot2::theme(legend.position="none")
	})
	mrres
}


blank_plot <- function(message)
{
	requireNamespace("ggplot2", quietly=TRUE)
	ggplot2::ggplot(data.frame(a=0,b=0,n=message)) + ggplot2::geom_text(ggplot2::aes(x=a,y=b,label=n)) + ggplot2::labs(x=NULL,y=NULL) + ggplot2::theme(axis.text=ggplot2::element_blank(), axis.ticks=ggplot2::element_blank())
}


```

# Basal Metabolic Rate 1

```{r}
betas <- sel_graph %>%
  filter(exposure == c("Basal metabolic rate || id:ukb-b-16446"))%>%
  dplyr::select( b.pilling, b.timmers, b.90, b.99) %>%
  unlist()
se <- sel_graph %>%
  filter(exposure == c("Basal metabolic rate || id:ukb-b-16446"))%>%
  dplyr::select(se.pilling, se.timmers, se.90, se.99)%>%
  unlist()
nsnp <- sel_graph %>%
  filter(exposure == c("Basal metabolic rate || id:ukb-b-16446"))%>%
  dplyr::select(nsnp.pilling, nsnp.timmers, nsnp.90, nsnp.99)%>%
  unlist()

Outcome <- c( "Pilling", "Timmers", "Deleen 90", "Deleen 99")
exposure <- c("Basal metabolic rate || id:ukb-b-16446")
graph.data <- as.data.frame(cbind(Outcome,exposure, betas, se, nsnp))

graph.data$betas <- as.numeric(graph.data$betas)
graph.data$se <- as.numeric(graph.data$se)

graph.data1 <- graph.data %>%
  mutate(lower.CI = betas - (1.96*se),
         upper.CI = betas + (1.96*se),
         OR = exp(betas),
         lower.OR.CI = exp(lower.CI),
         upper.OR.CI = exp(upper.CI))
graph.data1
ggplot(graph.data1) +
  geom_segment(aes(x = lower.CI, xend = upper.CI, y= 1:4, yend = 1:4, col = Outcome)) +
  xlab("") +
  ylab("") +
  ggtitle("Confidence Intervals for Four Longevity Outcomes")
```


As expected, the most overlap is between the two Deleen definitions and the Pilling and Timmers datasets (UK Biobank).

## Sensivity Analysis

```{r include=FALSE}
Pilling <- readRDS("C:/Users/me/Desktop/MPH/Internship/CPMC/Longevity/data/harmonized/pilling_df_gwas_ieu.rds")
Timmers <- timmers_gwas_df <- readRDS("C:/Users/me/Desktop/MPH/Internship/CPMC/Longevity/data/harmonized/timmers_gwas_ieu_df.rds")
Cases90 <-  readRDS("C:/Users/me/Desktop/MPH/Internship/CPMC/Longevity/data/harmonized/cases_90_gwas_ieu_df.rds")
Cases99 <- readRDS("C:/Users/me/Desktop/MPH/Internship/CPMC/Longevity/data/harmonized/cases_99_gwas_ieu_df.rds")

```
```{r}
#filter by Basal Metabolic Rate

Pilling_harm_dat <- Pilling %>%
  filter(exposure == "Basal metabolic rate || id:ukb-b-16446")

Timmers_harm_dat <- Timmers %>%
  filter(exposure == "Basal metabolic rate || id:ukb-b-16446")

Cases90_harm_dat <- Cases90 %>%
  filter(exposure == "Basal metabolic rate || id:ukb-b-16446")

Cases99_harm_dat <- Cases99 %>%
  filter(exposure == "Basal metabolic rate || id:ukb-b-16446")
```

### Heterogeneity Test

```{r include=FALSE}
het.pilling <- mr_heterogeneity(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
het.timmers <- mr_heterogeneity(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
het.cases90 <- mr_heterogeneity(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
het.cases99 <- mr_heterogeneity(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

het_test1 <- rbind(het.pilling,het.timmers,het.cases90,het.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, method, Q, Q_df, Q_pval) %>%
  rename(`Longevity Definition` = Outcome.dat)
```

```{r}
kbl(het_test1, booktabs = T, caption = "Test for Heterogeneity using Cochran's Q-test")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Evidence of Heterogeneity in the Timmers data.

### Check for Pleiotropy

```{r include=FALSE}
plei.pilling <- mr_pleiotropy_test(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
plei.timmers <- mr_pleiotropy_test(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
plei.cases90 <- mr_pleiotropy_test(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
plei.cases99 <- mr_pleiotropy_test(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

pli_test1 <- rbind(plei.pilling, plei.timmers,plei.cases90,plei.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, egger_intercept, se, pval) %>%
  rename(`Longevity Definition` = Outcome.dat,
         `Egger Intercept` = egger_intercept)

```

```{r}
kbl(pli_test1, booktabs = T, caption = "Test for Pleiotropy ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

no indication from this test that there is the presence of horizontal pleiotropy


## Leave on out Analysis - Check to see if there is one SNP influencing the results
```{r include=FALSE}
loo.pilling <- mr_leaveoneout(Pilling_harm_dat)%>%
  mutate(outcome.data = "Pilling")
loo.timmers <-mr_leaveoneout(Timmers_harm_dat)%>%
  mutate(outcome.data = "Timmers")
loo.cases90 <-mr_leaveoneout(Cases90_harm_dat)%>%
  mutate(outcome.data = "Cases90")
loo.cases99 <- mr_leaveoneout(Cases99_harm_dat)%>%
  mutate(outcome.data = "Cases99")

loo_test1 <- rbind(loo.pilling, loo.timmers,loo.cases90,loo.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(outcome.data, exposure, samplesize, SNP, b,se, p) %>%
  rename(`Longevity Definition` = outcome.data,
         `Sample Size` = samplesize,
         Beta = b,
         SE = se,
         `p-value` = p)

p1 <- mr_leaveoneout_plot(loo.pilling)
p2 <- mr_leaveoneout_plot(loo.timmers)
p3 <- mr_leaveoneout_plot(loo.cases90)
p4 <- mr_leaveoneout_plot(loo.cases99)

p1 <- grid.arrange(grobs = p1)
p2 <- grid.arrange(grobs = p2)
p3 <- grid.arrange(grobs = p3)
p4 <- grid.arrange(grobs = p4)
```


```{r}
plot_grid(p1, p2, labels = c("Pilling", "Timmers"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -1, label_colour = "blue")

plot_grid(p3, p4, labels = c("Deleen 90", "Deleen 99"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")

kbl(loo_test1, booktabs = T, caption = "Leave one out Analysis ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Although there are some SNPs which significantly affect the point estimate, most SNPs overlap.


## Scatter plot of Three Methods
```{r cache= TRUE, include=FALSE}
mr.pilling <- mr(Pilling_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.timmers <- mr(Timmers_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases90 <- mr(Cases90_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases99 <- mr(Cases99_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))

sc.pilling <- mrscatter_nolegend(mr.pilling, Pilling_harm_dat, title = "(a) Pilling")
sc.timmers <- mrscatter_nolegend(mr.timmers, Timmers_harm_dat, title ="(b) Timmers")
sc.cases90 <- mrscatter_nolegend(mr.cases90, Cases90_harm_dat, title ="(c) Deleen 90th Percentile")
sc.cases99 <- mrscatter_legend(mr.cases99, Cases99_harm_dat, title ="(d) Deleen 99th Percentile")

```

```{r}
exposure <- Pilling_harm_dat$exposure[1]

scatter <- plot_grid(sc.pilling[[1]],sc.timmers[[1]], sc.cases90[[1]], sc.cases99[[1]], nrow = 2)

scatter1 <- add_sub(scatter, paste("SNP effect on ", exposure), y = 0.2, vjust = 0, size = 8 )
ggdraw(scatter1)
# mr_scatter_plot(mr.timmers, Timmers_harm_dat)
# to check if the graph is the same
ggsave("images/Basal_met_rate_scatter.png")
```

## Forest Plot of Single SNP estimates compared to estimates with all SNPs

```{r include=FALSE}
pilling.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
timmers.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases90.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases99.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
```

```{r}
plot_grid(pilling.forest, timmers.forest,labels =  c("Pilling", "Timmers"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
plot_grid(cases90.forest, cases99.forest,labels =  c("Deleen 90", "Deleen 99"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
```

## Funnel plot of Single SNP analysis compared to the estimates from all SNPs



```{r include=FALSE}
pilling.funnel <- mr_funnel_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
timmers.funnel <-  mr_funnel_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases90.funnel <- mr_funnel_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases99.funnel <- mr_funnel_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))

pilling.funnel$`ukb-b-16446.3d2fYT`$theme$legend.direction <- 'none'
pilling.funnel$`ukb-b-16446.3d2fYT`$theme$legend.position <- 'none'
timmers.funnel$`ukb-b-16446.qdpgbc`$theme$legend.direction <- 'none'
timmers.funnel$`ukb-b-16446.qdpgbc`$theme$legend.position <- 'none'
cases90.funnel$`ukb-b-16446.Sr7BYr`$theme$legend.direction <- 'none'
cases90.funnel$`ukb-b-16446.Sr7BYr`$theme$legend.position <- 'none'
cases99.funnel$`ukb-b-16446.JdTzPB`$theme$legend.direction <- 'none'
cases99.funnel$`ukb-b-16446.JdTzPB`$theme$legend.position <- 'none'

```

```{r}


p1 <- pilling.funnel$`ukb-b-16446.3d2fYT` + 
  ggtitle("Parental 90") + theme(plot.title = element_text(color = "blue"))
p2 <- timmers.funnel$`ukb-b-16446.qdpgbc`+ 
  ggtitle("Parental") + theme(plot.title = element_text(color = "blue"))
p3 <- cases90.funnel$`ukb-b-16446.Sr7BYr`+ 
  ggtitle("Cases 90" ) + theme(plot.title = element_text(color = "blue"))
p4 <- cases99.funnel$`ukb-b-16446.JdTzPB` + 
  ggtitle("Cases 99") + theme(plot.title = element_text(color = "blue"))

annotate_figure(ggarrange(p1,p2,p3,p4, nrow = 2, ncol = 2) ,
                top = paste0("Single SNP estimates of ", Pilling_harm_dat$exposure[1], " on Longevity"))

ggsave("images/basal_met_rate_funnel.png")
```




# Age first had sexual intercourse 2
```{r}
betas <- sel_graph %>%
  filter(exposure %in% c("Age first had sexual intercourse || id:ukb-b-6591"))%>%
  select( b.pilling, b.timmers, b.90, b.99) %>%
  unlist()
se <- sel_graph %>%
   filter(exposure %in% c("Age first had sexual intercourse || id:ukb-b-6591"))%>%
  select(se.pilling, se.timmers, se.90, se.99)%>%
  unlist()
nsnp <- sel_graph %>%
  filter(exposure %in% c("Age first had sexual intercourse || id:ukb-b-6591"))%>%
  dplyr::select(nsnp.pilling, nsnp.timmers, nsnp.90, nsnp.99)%>%
  unlist()

Outcome <- c( "Pilling", "Timmers", "Deleen 90", "Deleen 99")
exposure <- c("Age first had sexual intercourse || id:ukb-b-6591")
graph.data <- as.data.frame(cbind(Outcome, exposure, betas, se, nsnp))

graph.data$betas <- as.numeric(graph.data$betas)
graph.data$se <- as.numeric(graph.data$se)

graph.data2 <- graph.data %>%
  mutate(lower.CI = betas - (1.96*se),
         upper.CI = betas + (1.96*se),
         OR = exp(betas),
         lower.OR.CI = exp(lower.CI),
         upper.OR.CI = exp(upper.CI))
graph.data
ggplot(graph.data2) +
  geom_segment(aes(x = lower.CI, xend = upper.CI, y= 1:4, yend = 1:4, col = Outcome)) +
  xlab("") +
  ylab("") +
  ggtitle("Confidence Intervals for Four Longevity Outcomes")
```


As expected, the most overlap is between the two Deleen definitions and the Pilling and Timmers datasets (UK Biobank).

## Sensivity Analysis

```{r include=FALSE}

#filter by Alzheimer

Pilling_harm_dat <- Pilling %>%
  filter(exposure == "Age first had sexual intercourse || id:ukb-b-6591")

Timmers_harm_dat <- Timmers %>%
  filter(exposure == "Age first had sexual intercourse || id:ukb-b-6591")

Cases90_harm_dat <- Cases90 %>%
  filter(exposure == "Age first had sexual intercourse || id:ukb-b-6591")

Cases99_harm_dat <- Cases99 %>%
  filter(exposure == "Age first had sexual intercourse || id:ukb-b-6591")
```

### Heterogeneity Test

```{r include=FALSE}
het.pilling <- mr_heterogeneity(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
het.timmers <- mr_heterogeneity(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
het.cases90 <- mr_heterogeneity(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
het.cases99 <- mr_heterogeneity(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

het_test2 <- rbind(het.pilling,het.timmers,het.cases90,het.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, method, Q, Q_df, Q_pval) %>%
  rename(`Longevity Definition` = Outcome.dat)
```

```{r}
kbl(het_test2, booktabs = T, caption = "Test for Heterogeneity using Cochran's Q-test")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Evidence of Heterogeneity in the Timmers data.

### Check for Pleiotropy

```{r include=FALSE}
plei.pilling <- mr_pleiotropy_test(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
plei.timmers <- mr_pleiotropy_test(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
plei.cases90 <- mr_pleiotropy_test(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
plei.cases99 <- mr_pleiotropy_test(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

pli_test2 <- rbind(plei.pilling, plei.timmers,plei.cases90,plei.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, egger_intercept, se, pval) %>%
  rename(`Longevity Definition` = Outcome.dat,
         `Egger Intercept` = egger_intercept)

```

```{r}
kbl(pli_test2, booktabs = T, caption = "Test for Pleiotropy ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

no indication from this test that there is the presence of horizontal pleiotropy


## Leave on out Analysis - Check to see if there is one SNP influencing the results
```{r include=FALSE}
loo.pilling <- mr_leaveoneout(Pilling_harm_dat)%>%
  mutate(outcome.data = "Pilling")
loo.timmers <-mr_leaveoneout(Timmers_harm_dat)%>%
  mutate(outcome.data = "Timmers")
loo.cases90 <-mr_leaveoneout(Cases90_harm_dat)%>%
  mutate(outcome.data = "Cases90")
loo.cases99 <- mr_leaveoneout(Cases99_harm_dat)%>%
  mutate(outcome.data = "Cases99")

loo_test2 <- rbind(loo.pilling, loo.timmers,loo.cases90,loo.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(outcome.data, exposure, samplesize, SNP, b,se, p) %>%
  rename(`Longevity Definition` = outcome.data,
         `Sample Size` = samplesize,
         Beta = b,
         SE = se,
         `p-value` = p)

p1 <- mr_leaveoneout_plot(loo.pilling)
p2 <- mr_leaveoneout_plot(loo.timmers)
p3 <- mr_leaveoneout_plot(loo.cases90)
p4 <- mr_leaveoneout_plot(loo.cases99)

p1 <- grid.arrange(grobs = p1)
p2 <- grid.arrange(grobs = p2)
p3 <- grid.arrange(grobs = p3)
p4 <- grid.arrange(grobs = p4)
```


```{r}
plot_grid(p1, p2, labels = c("Pilling", "Timmers"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -1, label_colour = "blue")

plot_grid(p3, p4, labels = c("Deleen 90", "Deleen 99"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")

kbl(loo_test2, booktabs = T, caption = "Leave one out Analysis ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Although there are some SNPs which significantly affect the point estimate, most SNPs overlap.


## Scatter plot of Three Methods
```{r cache= TRUE, include=FALSE}
mr.pilling <- mr(Pilling_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.timmers <- mr(Timmers_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases90 <- mr(Cases90_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases99 <- mr(Cases99_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))

sc.pilling <- mrscatter_nolegend(mr.pilling, Pilling_harm_dat, title = "(a) Pilling")
sc.timmers <- mrscatter_nolegend(mr.timmers, Timmers_harm_dat, title ="(b) Timmers")
sc.cases90 <- mrscatter_nolegend(mr.cases90, Cases90_harm_dat, title ="(c) Deleen 90th Percentile")
sc.cases99 <- mrscatter_legend(mr.cases99, Cases99_harm_dat, title ="(d) Deleen 99th Percentile")

```

```{r}
exposure <- Pilling_harm_dat$exposure[1]

scatter <- plot_grid(sc.pilling[[1]],sc.timmers[[1]], sc.cases90[[1]], sc.cases99[[1]], nrow = 2)

scatter2 <- add_sub(scatter, paste("SNP effect on ", exposure), y = 0.2, vjust = 0, size = 8 )
ggdraw(scatter2)

# to check if the graph is the same
#mr_scatter_plot(mr.timmers, Timmers_harm_dat)

ggsave("images/first_sex_scatter.png")
```


## Forest Plot of Single SNP estimates compared to estimates with all SNPs

```{r include=FALSE}
pilling.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
timmers.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases90.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases99.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
```

```{r}
plot_grid(pilling.forest, timmers.forest,labels =  c("Pilling", "Timmers"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
plot_grid(cases90.forest, cases99.forest,labels =  c("Deleen 90", "Deleen 99"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
```

## Funnel plot of Single SNP analysis compared to the estimates from all SNPs



```{r include=FALSE}
pilling.funnel <- mr_funnel_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
timmers.funnel <-  mr_funnel_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases90.funnel <- mr_funnel_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases99.funnel <- mr_funnel_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))

pilling.funnel$`ukb-b-6591.CbyrS2`$theme$legend.direction <- 'none'
pilling.funnel$`ukb-b-6591.CbyrS2`$theme$legend.position <- 'none'
timmers.funnel$`ukb-b-6591.JS7hw1`$theme$legend.direction <- 'none'
timmers.funnel$`ukb-b-6591.JS7hw1`$theme$legend.position <- 'none'
cases90.funnel$`ukb-b-6591.j4BMdl`$theme$legend.direction <- 'none'
cases90.funnel$`ukb-b-6591.j4BMdl`$theme$legend.position <- 'none'
cases99.funnel$`ukb-b-6591.CrMByu`$theme$legend.direction <- 'none'
cases99.funnel$`ukb-b-6591.CrMByu`$theme$legend.position <- 'none'

Pilling_harm_dat$exposure[1]

```

```{r}


p1 <- pilling.funnel$`ukb-b-6591.CbyrS2` + 
  ggtitle("Parental 90") + theme(plot.title = element_text(color = "blue"))
p2 <- timmers.funnel$`ukb-b-6591.JS7hw1`+ 
  ggtitle("Parental") + theme(plot.title = element_text(color = "blue"))
p3 <- cases90.funnel$`ukb-b-6591.j4BMdl`+ 
  ggtitle("Cases 90" ) + theme(plot.title = element_text(color = "blue"))
p4 <- cases99.funnel$`ukb-b-6591.CrMByu` + 
  ggtitle("Cases 99") + theme(plot.title = element_text(color = "blue"))

annotate_figure(ggarrange(p1,p2,p3,p4, nrow = 2, ncol = 2) ,
                top = paste0("Single SNP estimates of ", Pilling_harm_dat$exposure[1], " on Longevity"))

ggsave("images/first_sex_funnel.png")


```

# Overall Health 3

```{r}
betas <- sel_graph %>%
  filter(exposure ==  "Overall health rating || id:ukb-b-6306")%>%
  select( b.pilling, b.timmers, b.90, b.99) %>%
  unlist()
se <- sel_graph %>%
  filter(exposure ==  "Overall health rating || id:ukb-b-6306")%>%
  select(se.pilling, se.timmers, se.90, se.99)%>%
  unlist()
nsnp <- sel_graph %>%
  filter(exposure ==  "Overall health rating || id:ukb-b-6306")%>%
  dplyr::select(nsnp.pilling, nsnp.timmers, nsnp.90, nsnp.99)%>%
  unlist()

Outcome <- c( "Pilling", "Timmers", "Deleen 90", "Deleen 99")
exposure = c( "Overall health rating || id:ukb-b-6306")
graph.data <- as.data.frame(cbind(Outcome, exposure, betas, se, nsnp))

graph.data$betas <- as.numeric(graph.data$betas)
graph.data$se <- as.numeric(graph.data$se)

graph.data3 <- graph.data %>%
  mutate(lower.CI = betas - (1.96*se),
         upper.CI = betas + (1.96*se),
         OR = exp(betas),
         lower.OR.CI = exp(lower.CI),
         upper.OR.CI = exp(upper.CI))
graph.data
ggplot(graph.data3) +
  geom_segment(aes(x = lower.CI, xend = upper.CI, y= 1:4, yend = 1:4, col = Outcome)) +
  xlab("") +
  ylab("") +
  ggtitle("Confidence Intervals for Four Longevity Outcomes")
```




## Sensivity Analysis

```{r include=FALSE}


#filter by Alzheimer

Pilling_harm_dat <- Pilling %>%
  filter(exposure ==  "Overall health rating || id:ukb-b-6306")

Timmers_harm_dat <- Timmers %>%
  filter(exposure ==  "Overall health rating || id:ukb-b-6306")

Cases90_harm_dat <- Cases90 %>%
  filter(exposure == "Overall health rating || id:ukb-b-6306")

Cases99_harm_dat <- Cases99 %>%
  filter(exposure ==  "Overall health rating || id:ukb-b-6306")
```

### Heterogeneity Test

```{r include=FALSE}
het.pilling <- mr_heterogeneity(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
het.timmers <- mr_heterogeneity(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
het.cases90 <- mr_heterogeneity(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
het.cases99 <- mr_heterogeneity(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

het_test3 <- rbind(het.pilling,het.timmers,het.cases90,het.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, method, Q, Q_df, Q_pval) %>%
  rename(`Longevity Definition` = Outcome.dat)
```

```{r}
kbl(het_test3, booktabs = T, caption = "Test for Heterogeneity using Cochran's Q-test")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Evidence of Heterogeneity in the Pilling data.

### Check for Pleiotropy

```{r include=FALSE}
plei.pilling <- mr_pleiotropy_test(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
plei.timmers <- mr_pleiotropy_test(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
plei.cases90 <- mr_pleiotropy_test(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
plei.cases99 <- mr_pleiotropy_test(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

pli_test3 <- rbind(plei.pilling, plei.timmers,plei.cases90,plei.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, egger_intercept, se, pval) %>%
  rename(`Longevity Definition` = Outcome.dat,
         `Egger Intercept` = egger_intercept)

```

```{r}
kbl(pli_test3, booktabs = T, caption = "Test for Pleiotropy ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

no indication from this test that there is the presence of horizontal pleiotropy


## Leave on out Analysis - Check to see if there is one SNP influencing the results
```{r include=FALSE}
loo.pilling <- mr_leaveoneout(Pilling_harm_dat)%>%
  mutate(outcome.data = "Pilling")
loo.timmers <-mr_leaveoneout(Timmers_harm_dat)%>%
  mutate(outcome.data = "Timmers")
loo.cases90 <-mr_leaveoneout(Cases90_harm_dat)%>%
  mutate(outcome.data = "Cases90")
loo.cases99 <- mr_leaveoneout(Cases99_harm_dat)%>%
  mutate(outcome.data = "Cases99")

loo_test3 <- rbind(loo.pilling, loo.timmers,loo.cases90,loo.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(outcome.data, exposure, samplesize, SNP, b,se, p) %>%
  rename(`Longevity Definition` = outcome.data,
         `Sample Size` = samplesize,
         Beta = b,
         SE = se,
         `p-value` = p)

p1 <- mr_leaveoneout_plot(loo.pilling)
p2 <- mr_leaveoneout_plot(loo.timmers)
p3 <- mr_leaveoneout_plot(loo.cases90)
p4 <- mr_leaveoneout_plot(loo.cases99)

p1 <- grid.arrange(grobs = p1)
p2 <- grid.arrange(grobs = p2)
p3 <- grid.arrange(grobs = p3)
p4 <- grid.arrange(grobs = p4)
```


```{r}
plot_grid(p1, p2, labels = c("Pilling", "Timmers"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -1, label_colour = "blue")

plot_grid(p3, p4, labels = c("Deleen 90", "Deleen 99"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")

kbl(loo_test3, booktabs = T, caption = "Leave one out Analysis ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Although there are some SNPs which significantly affect the point estimate, most SNPs overlap with IVW estimate. However, Pilling data has fewer SNPs overall, with one SNP, rs6679677, whose point estimate is outside the IVW estimate.


## Scatter plot of Three Methods
```{r cache= TRUE, include=FALSE}
mr.pilling <- mr(Pilling_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.timmers <- mr(Timmers_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases90 <- mr(Cases90_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases99 <- mr(Cases99_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))

sc.pilling <- mrscatter_nolegend(mr.pilling, Pilling_harm_dat, title = "(a) Pilling")
sc.timmers <- mrscatter_nolegend(mr.timmers, Timmers_harm_dat, title ="(b) Timmers")
sc.cases90 <- mrscatter_nolegend(mr.cases90, Cases90_harm_dat, title ="(c) Deleen 90th Percentile")
sc.cases99 <- mrscatter_legend(mr.cases99, Cases99_harm_dat, title ="(d) Deleen 99th Percentile")

```

```{r}
exposure <- Pilling_harm_dat$exposure[1]

scatter <- plot_grid(sc.pilling[[1]],sc.timmers[[1]], sc.cases90[[1]], sc.cases99[[1]], nrow = 2)

scatter3 <- add_sub(scatter, paste("SNP effect on ", exposure), y = 0.2, vjust = 0, size = 8 )
ggdraw(scatter3)

# to check if the graph is the same
#mr_scatter_plot(mr.timmers, Timmers_harm_dat)

ggsave("images/overall_health_scatter.png")
```

## Forest Plot of Single SNP estimates compared to estimates with all SNPs

```{r include=FALSE}
pilling.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
timmers.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases90.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases99.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
```

```{r}
plot_grid(pilling.forest, timmers.forest,labels =  c("Pilling", "Timmers"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
plot_grid(cases90.forest, cases99.forest,labels =  c("Deleen 90", "Deleen 99"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
```

## Funnel plot of Single SNP analysis compared to the estimates from all SNPs



```{r include=FALSE}
pilling.funnel <- mr_funnel_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
timmers.funnel <-  mr_funnel_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases90.funnel <- mr_funnel_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases99.funnel <- mr_funnel_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
pilling.funnel$`ukb-b-6306.TtpwPi`$theme$legend.direction <- 'none'
pilling.funnel$`ukb-b-6306.TtpwPi`$theme$legend.position <- 'none'
timmers.funnel$`ukb-b-6306.3xmq7y`$theme$legend.direction <- 'none'
timmers.funnel$`ukb-b-6306.3xmq7y`$theme$legend.position <- 'none'
cases90.funnel$`ukb-b-6306.L0AiO9`$theme$legend.direction <- 'none'
cases90.funnel$`ukb-b-6306.L0AiO9`$theme$legend.position <- 'none'
cases99.funnel$`ukb-b-6306.9hTG4T`$theme$legend.direction <- 'none'
cases99.funnel$`ukb-b-6306.9hTG4T`$theme$legend.position <- 'none'

Pilling_harm_dat$exposure[1]

```

```{r}


p1 <- pilling.funnel$`ukb-b-6306.TtpwPi` + 
  ggtitle("Parental 90") + theme(plot.title = element_text(color = "blue"))
p2 <- timmers.funnel$`ukb-b-6306.3xmq7y`+ 
  ggtitle("Parental") + theme(plot.title = element_text(color = "blue"))
p3 <- cases90.funnel$`ukb-b-6306.L0AiO9`+ 
  ggtitle("Cases 90" ) + theme(plot.title = element_text(color = "blue"))
p4 <- cases99.funnel$`ukb-b-6306.9hTG4T` + 
  ggtitle("Cases 99") + theme(plot.title = element_text(color = "blue"))

annotate_figure(ggarrange(p1,p2,p3,p4, nrow = 2, ncol = 2) ,
                top = paste0("Single SNP estimates of ", Pilling_harm_dat$exposure[1], " on Longevity"))

ggsave("images/overall_health_funnel.png")
```


# Hypertension 4

```{r}
betas <- sel_graph %>%
  filter(exposure == "Non-cancer illness code, self-reported: hypertension || id:ukb-b-14057")%>%
  dplyr::select(b.pilling, b.timmers, b.90, b.99) %>%
  unlist()
se <- sel_graph %>%
  filter(exposure == "Non-cancer illness code, self-reported: hypertension || id:ukb-b-14057")%>%
  dplyr::select(se.pilling, se.timmers, se.90, se.99)%>%
  unlist()
nsnp <- sel_graph %>%
  filter(exposure == "Non-cancer illness code, self-reported: hypertension || id:ukb-b-14057")%>%
  dplyr::select(nsnp.pilling, nsnp.timmers, nsnp.90, nsnp.99)%>%
  unlist()
Outcome <- c( "Pilling", "Timmers", "Deleen 90", "Deleen 99")
exposure = c("Non-cancer illness code, self-reported: hypertension || id:ukb-b-14057")
graph.data <- as.data.frame(cbind(Outcome, exposure,  betas, se, nsnp))

graph.data$betas <- as.numeric(graph.data$betas)
graph.data$se <- as.numeric(graph.data$se)

graph.data4 <- graph.data %>%
  mutate(lower.CI = betas - (1.96*se),
         upper.CI = betas + (1.96*se),
         OR = exp(betas),
         lower.OR.CI = exp(lower.CI),
         upper.OR.CI = exp(upper.CI))
graph.data4
ggplot(graph.data4) +
  geom_segment(aes(x = lower.CI, xend = upper.CI, y= 1:4, yend = 1:4, col = Outcome)) +
  xlab("") +
  ylab("") +
  ggtitle("Confidence Intervals for Four Longevity Outcomes")
```


As expected, the most overlap is between the two Deleen definitions and the Pilling and Timmers datasets (UK Biobank).

## Sensivity Analysis

```{r include=FALSE}



Pilling_harm_dat <- Pilling %>%
  filter(exposure == "Non-cancer illness code, self-reported: hypertension || id:ukb-b-14057")

Timmers_harm_dat <- Timmers %>%
  filter(exposure == "Non-cancer illness code, self-reported: hypertension || id:ukb-b-14057")

Cases90_harm_dat <- Cases90 %>%
  filter(exposure == "Non-cancer illness code, self-reported: hypertension || id:ukb-b-14057")

Cases99_harm_dat <- Cases99 %>%
  filter(exposure == "Non-cancer illness code, self-reported: hypertension || id:ukb-b-14057")
```

### Heterogeneity Test

```{r include=FALSE}
het.pilling <- mr_heterogeneity(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
het.timmers <- mr_heterogeneity(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
het.cases90 <- mr_heterogeneity(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
het.cases99 <- mr_heterogeneity(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

het_test4 <- rbind(het.pilling,het.timmers,het.cases90,het.cases99) %>%
  dplyr::select(-id.exposure, -id.outcome, -outcome) %>%
  dplyr::select(Outcome.dat, exposure, method, Q, Q_df, Q_pval) %>%
  rename(`Longevity Definition` = Outcome.dat)
```

```{r}
kbl(het_test4, booktabs = T, caption = "Test for Heterogeneity using Cochran's Q-test")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Evidence of Heterogeneity in the Timmers data.

### Check for Pleiotropy

```{r include=FALSE}
plei.pilling <- mr_pleiotropy_test(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
plei.timmers <- mr_pleiotropy_test(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
plei.cases90 <- mr_pleiotropy_test(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
plei.cases99 <- mr_pleiotropy_test(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

pli_test4 <- rbind(plei.pilling, plei.timmers,plei.cases90,plei.cases99) %>%
  dplyr::select(-id.exposure, -id.outcome, -outcome) %>%
  dplyr::select(Outcome.dat, exposure, egger_intercept, se, pval) %>%
  rename(`Longevity Definition` = Outcome.dat,
         `Egger Intercept` = egger_intercept)

```

```{r}
kbl(pli_test4, booktabs = T, caption = "Test for Pleiotropy ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

no indication from this test that there is the presence of horizontal pleiotropy


## Leave on out Analysis - Check to see if there is one SNP influencing the results
```{r include=FALSE}
loo.pilling <- mr_leaveoneout(Pilling_harm_dat)%>%
  mutate(outcome.data = "Pilling")
loo.timmers <-mr_leaveoneout(Timmers_harm_dat)%>%
  mutate(outcome.data = "Timmers")
loo.cases90 <-mr_leaveoneout(Cases90_harm_dat)%>%
  mutate(outcome.data = "Cases90")
loo.cases99 <- mr_leaveoneout(Cases99_harm_dat)%>%
  mutate(outcome.data = "Cases99")

loo_test4 <- rbind(loo.pilling, loo.timmers,loo.cases90,loo.cases99) %>%
  dplyr::select(-id.exposure, -id.outcome, -outcome) %>%
  dplyr::select(outcome.data, exposure, samplesize, SNP, b,se, p) %>%
  rename(`Longevity Definition` = outcome.data,
         `Sample Size` = samplesize,
         Beta = b,
         SE = se,
         `p-value` = p)

p1 <- mr_leaveoneout_plot(loo.pilling)
p2 <- mr_leaveoneout_plot(loo.timmers)
p3 <- mr_leaveoneout_plot(loo.cases90)
p4 <- mr_leaveoneout_plot(loo.cases99)

p1 <- grid.arrange(grobs = p1)
p2 <- grid.arrange(grobs = p2)
p3 <- grid.arrange(grobs = p3)
p4 <- grid.arrange(grobs = p4)
```


```{r}
plot_grid(p1, p2, labels = c("Pilling", "Timmers"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -1, label_colour = "blue")

plot_grid(p3, p4, labels = c("Deleen 90", "Deleen 99"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")

kbl(loo_test4, booktabs = T, caption = "Leave one out Analysis ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Although there are some SNPs which significantly affect the point estimate, most SNPs overlap.


## Scatter plot of Three Methods
```{r cache= TRUE, include=FALSE}
mr.pilling <- mr(Pilling_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.timmers <- mr(Timmers_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases90 <- mr(Cases90_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases99 <- mr(Cases99_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))

sc.pilling <- mrscatter_nolegend(mr.pilling, Pilling_harm_dat, title = "(a) Pilling")
sc.timmers <- mrscatter_nolegend(mr.timmers, Timmers_harm_dat, title ="(b) Timmers")
sc.cases90 <- mrscatter_nolegend(mr.cases90, Cases90_harm_dat, title ="(c) Deleen 90th Percentile")
sc.cases99 <- mrscatter_legend(mr.cases99, Cases99_harm_dat, title ="(d) Deleen 99th Percentile")

```

```{r}
exposure <- Pilling_harm_dat$exposure[1]

scatter <- plot_grid(sc.pilling[[1]],sc.timmers[[1]], sc.cases90[[1]], sc.cases99[[1]], nrow = 2)

scatter4 <- add_sub(scatter, paste("SNP effect on ", exposure), y = 0.2, vjust = 0, size = 8 )
ggdraw(scatter4)

# to check if the graph is the same
#mr_scatter_plot(mr.timmers, Timmers_harm_dat)

ggsave("images/hypertension_scatter.png")
```

## Forest Plot of Single SNP estimates compared to estimates with all SNPs

```{r include=FALSE}
pilling.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
timmers.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases90.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases99.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
```

```{r}
plot_grid(pilling.forest, timmers.forest,labels =  c("Pilling", "Timmers"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
plot_grid(cases90.forest, cases99.forest,labels =  c("Deleen 90", "Deleen 99"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
```

## Funnel plot of Single SNP analysis compared to the estimates from all SNPs



```{r include=FALSE}
pilling.funnel <- mr_funnel_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
timmers.funnel <- mr_funnel_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases90.funnel <- mr_funnel_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases99.funnel <- mr_funnel_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
pilling.funnel$`ukb-b-14057.fXnZ1H`$theme$legend.direction <- 'none'
pilling.funnel$`ukb-b-14057.fXnZ1H`$theme$legend.position <- 'none'
timmers.funnel$`ukb-b-14057.HE5yJs`$theme$legend.direction <- 'none'
timmers.funnel$`ukb-b-14057.HE5yJs`$theme$legend.position <- 'none'
cases90.funnel$`ukb-b-14057.3jqzJU`$theme$legend.direction <- 'none'
cases90.funnel$`ukb-b-14057.3jqzJU`$theme$legend.position <- 'none'
cases99.funnel$`ukb-b-14057.hmU1Yx`$theme$legend.direction <- 'none'
cases99.funnel$`ukb-b-14057.hmU1Yx`$theme$legend.position <- 'none'

Pilling_harm_dat$exposure[1]

```

```{r}


p1 <- pilling.funnel$`ukb-b-14057.fXnZ1H` + 
  ggtitle("Parental 90") + theme(plot.title = element_text(color = "blue"))
p2 <- timmers.funnel$`ukb-b-14057.HE5yJs`+ 
  ggtitle("Parental") + theme(plot.title = element_text(color = "blue"))
p3 <- cases90.funnel$`ukb-b-14057.3jqzJU`+ 
  ggtitle("Cases 90" ) + theme(plot.title = element_text(color = "blue"))
p4 <- cases99.funnel$`ukb-b-14057.hmU1Yx` + 
  ggtitle("Cases 99") + theme(plot.title = element_text(color = "blue"))

annotate_figure(ggarrange(p1,p2,p3,p4, nrow = 2, ncol = 2) ,
                top = paste0("Single SNP estimates of hypertension || id:ukb-b-14057 on Longevity"))

ggsave("images/hypertension_funnel.png")
```



# Arm Fat 5

```{r}
betas <- sel_graph %>%
  filter(exposure == "Arm fat mass (left) || id:ukb-a-287" )%>%
  select( b.pilling, b.timmers, b.90, b.99) %>%
  unlist()
se <- sel_graph %>%
  filter(exposure == "Arm fat mass (left) || id:ukb-a-287" )%>%
  select(se.pilling, se.timmers, se.90, se.99)%>%
  unlist()
nsnp <- sel_graph %>%
  filter(exposure == "Arm fat mass (left) || id:ukb-a-287" )%>%
  dplyr::select(nsnp.pilling, nsnp.timmers, nsnp.90, nsnp.99)%>%
  unlist()

Outcome <- c( "Pilling", "Timmers", "Deleen 90", "Deleen 99")
exposure <- c("Arm fat mass (left) || id:ukb-a-287")
graph.data <- as.data.frame(cbind(Outcome,exposure, betas, se, nsnp))

graph.data$betas <- as.numeric(graph.data$betas)
graph.data$se <- as.numeric(graph.data$se)

graph.data5 <- graph.data %>%
  mutate(lower.CI = betas - (1.96*se),
         upper.CI = betas + (1.96*se),
         OR = exp(betas),
         lower.OR.CI = exp(lower.CI),
         upper.OR.CI = exp(upper.CI))
graph.data
ggplot(graph.data5) +
  geom_segment(aes(x = lower.CI, xend = upper.CI, y= 1:4, yend = 1:4, col = Outcome)) +
  xlab("") +
  ylab("") +
  ggtitle("Confidence Intervals for Four Longevity Outcomes")
```


As expected, the most overlap is between the two Deleen definitions and the Pilling and Timmers datasets (UK Biobank).

## Sensivity Analysis

```{r include=FALSE}


#filter by Alzheimer

Pilling_harm_dat <- Pilling %>%
  filter(exposure == "Arm fat mass (left) || id:ukb-a-287" )

Timmers_harm_dat  <- Timmers %>%
  filter(exposure == "Arm fat mass (left) || id:ukb-a-287" )

Cases90_harm_dat  <- Cases90 %>%
  filter(exposure == "Arm fat mass (left) || id:ukb-a-287" )

Cases99_harm_dat  <- Cases99 %>%
  filter(exposure == "Arm fat mass (left) || id:ukb-a-287" )
```

### Heterogeneity Test

```{r include=FALSE}
het.pilling <- mr_heterogeneity(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
het.timmers <- mr_heterogeneity(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
het.cases90 <- mr_heterogeneity(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
het.cases99 <- mr_heterogeneity(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

het_test5 <- rbind(het.pilling,het.timmers,het.cases90,het.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, method, Q, Q_df, Q_pval) %>%
  rename(`Longevity Definition` = Outcome.dat)
```

```{r}
kbl(het_test5, booktabs = T, caption = "Test for Heterogeneity using Cochran's Q-test")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Evidence of Heterogeneity in the Timmers data.

### Check for Pleiotropy

```{r include=FALSE}
plei.pilling <- mr_pleiotropy_test(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
plei.timmers <- mr_pleiotropy_test(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
plei.cases90 <- mr_pleiotropy_test(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
plei.cases99 <- mr_pleiotropy_test(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

pli_test5 <- rbind(plei.pilling, plei.timmers,plei.cases90,plei.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, egger_intercept, se, pval) %>%
  rename(`Longevity Definition` = Outcome.dat,
         `Egger Intercept` = egger_intercept)

```

```{r}
kbl(pli_test5, booktabs = T, caption = "Test for Pleiotropy ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

no indication from this test that there is the presence of horizontal pleiotropy


## Leave on out Analysis - Check to see if there is one SNP influencing the results
```{r include=FALSE}
loo.pilling <- mr_leaveoneout(Pilling_harm_dat)%>%
  mutate(outcome.data = "Pilling")
loo.timmers <-mr_leaveoneout(Timmers_harm_dat)%>%
  mutate(outcome.data = "Timmers")
loo.cases90 <-mr_leaveoneout(Cases90_harm_dat)%>%
  mutate(outcome.data = "Cases90")
loo.cases99 <- mr_leaveoneout(Cases99_harm_dat)%>%
  mutate(outcome.data = "Cases99")

loo_test5 <- rbind(loo.pilling, loo.timmers,loo.cases90,loo.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(outcome.data, exposure, samplesize, SNP, b,se, p) %>%
  rename(`Longevity Definition` = outcome.data,
         `Sample Size` = samplesize,
         Beta = b,
         SE = se,
         `p-value` = p)


```


```{r}
plot_grid(p1, p2, labels = c("Pilling", "Timmers"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -1, label_colour = "blue")

plot_grid(p3, p4, labels = c("Deleen 90", "Deleen 99"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")

kbl(loo_test5, booktabs = T, caption = "Leave one out Analysis ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Although there are some SNPs which significantly affect the point estimate, most SNPs overlap.


## Scatter plot of Three Methods
```{r cache= TRUE, include=FALSE}
mr.pilling <- mr(Pilling_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.timmers <- mr(Timmers_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases90 <- mr(Cases90_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases99 <- mr(Cases99_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))

sc.pilling <- mrscatter_nolegend(mr.pilling, Pilling_harm_dat, title = "(a) Pilling")
sc.timmers <- mrscatter_nolegend(mr.timmers, Timmers_harm_dat, title ="(b) Timmers")
sc.cases90 <- mrscatter_nolegend(mr.cases90, Cases90_harm_dat, title ="(c) Deleen 90th Percentile")
sc.cases99 <- mrscatter_legend(mr.cases99, Cases99_harm_dat, title ="(d) Deleen 99th Percentile")

```

```{r}
exposure <- Pilling_harm_dat$exposure[1]

scatter <- plot_grid(sc.pilling[[1]],sc.timmers[[1]], sc.cases90[[1]], sc.cases99[[1]], nrow = 2)

scatter5 <- add_sub(scatter, paste("SNP effect on ", exposure), y = 0.2, vjust = 0, size = 8 )
ggdraw(scatter5)

# to check if the graph is the same
#mr_scatter_plot(mr.timmers, Timmers_harm_dat)

ggsave("images/arm_fat_scatter.png")
```

## Forest Plot of Single SNP estimates compared to estimates with all SNPs

```{r include=FALSE}
pilling.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
timmers.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases90.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases99.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
```

```{r}
plot_grid(pilling.forest, timmers.forest,labels =  c("Pilling", "Timmers"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
plot_grid(cases90.forest, cases99.forest,labels =  c("Deleen 90", "Deleen 99"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
```

## Funnel plot of Single SNP analysis compared to the estimates from all SNPs



```{r include=FALSE}
pilling.funnel <- mr_funnel_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
timmers.funnel <-  mr_funnel_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases90.funnel <- mr_funnel_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases99.funnel <- mr_funnel_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
pilling.funnel$`ukb-a-287.rBgTxY`$theme$legend.direction <- 'none'
pilling.funnel$`ukb-a-287.rBgTxY`$theme$legend.position <- 'none'
timmers.funnel$`ukb-a-287.fyBExD`$theme$legend.direction <- 'none'
timmers.funnel$`ukb-a-287.fyBExD`$theme$legend.position <- 'none'
cases90.funnel$`ukb-a-287.4LUVlt`$theme$legend.direction <- 'none'
cases90.funnel$`ukb-a-287.4LUVlt`$theme$legend.position <- 'none'
cases99.funnel$`ukb-a-287.dOy4Ro`$theme$legend.direction <- 'none'
cases99.funnel$`ukb-a-287.dOy4Ro`$theme$legend.position <- 'none'

Pilling_harm_dat$exposure[1]

```

```{r}


p1 <- pilling.funnel$`ukb-a-287.rBgTxY` + 
  ggtitle("Parental 90") + theme(plot.title = element_text(color = "blue"))
p2 <- timmers.funnel$`ukb-a-287.fyBExD`+ 
  ggtitle("Parental") + theme(plot.title = element_text(color = "blue"))
p3 <- cases90.funnel$`ukb-a-287.4LUVlt`+ 
  ggtitle("Cases 90" ) + theme(plot.title = element_text(color = "blue"))
p4 <- cases99.funnel$`ukb-a-287.dOy4Ro` + 
  ggtitle("Cases 99") + theme(plot.title = element_text(color = "blue"))

annotate_figure(ggarrange(p1,p2,p3,p4, nrow = 2, ncol = 2) ,
                top = paste0("Single SNP estimates of ", Pilling_harm_dat$exposure[1], " on Longevity"))

ggsave("images/arm_fat_funnel.png")
```





# Glycalated hameoglobin HbA1c (unit increase) 6
```{r}
betas <- sel_graph %>%
  filter(exposure == c( "Glycated haemoglobin || id:ukb-d-30750_irnt"))%>%
  dplyr::select( b.pilling, b.timmers, b.90, b.99) %>%
  unlist()
se <- sel_graph %>%
  filter(exposure == c( "Glycated haemoglobin || id:ukb-d-30750_irnt"))%>%
  dplyr::select(se.pilling, se.timmers, se.90, se.99)%>%
  unlist()
nsnp <- sel_graph %>%
  filter(exposure == c( "Glycated haemoglobin || id:ukb-d-30750_irnt"))%>%
  dplyr::select(nsnp.pilling, nsnp.timmers, nsnp.90, nsnp.99)%>%
  unlist()

Outcome <- c( "Pilling", "Timmers", "Deleen 90", "Deleen 99")
exposure <- c( "Glycated haemoglobin || id:ukb-d-30750_irnt")
graph.data <- as.data.frame(cbind(Outcome,exposure, betas, se, nsnp))

graph.data$betas <- as.numeric(graph.data$betas)
graph.data$se <- as.numeric(graph.data$se)

graph.data6 <- graph.data %>%
  mutate(lower.CI = betas - (1.96*se),
         upper.CI = betas + (1.96*se),
         OR = exp(betas),
         lower.OR.CI = exp(lower.CI),
         upper.OR.CI = exp(upper.CI))
graph.data6
ggplot(graph.data6) +
  geom_segment(aes(x = lower.CI, xend = upper.CI, y= 1:4, yend = 1:4, col = Outcome)) +
  xlab("") +
  ylab("") +
  ggtitle("Confidence Intervals for Four Longevity Outcomes")
```




## Sensivity Analysis

```{r include=FALSE}

#filter by Alzheimer

Pilling_harm_dat <- Pilling %>%
  filter(exposure ==  "Glycated haemoglobin || id:ukb-d-30750_irnt")

Timmers_harm_dat <- Timmers %>%
  filter(exposure ==  "Glycated haemoglobin || id:ukb-d-30750_irnt")

Cases90_harm_dat <- Cases90 %>%
  filter(exposure ==  "Glycated haemoglobin || id:ukb-d-30750_irnt")

Cases99_harm_dat <- Cases99 %>%
  filter(exposure ==  "Glycated haemoglobin || id:ukb-d-30750_irnt")
```

### Heterogeneity Test

```{r include=FALSE}
het.pilling <- mr_heterogeneity(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
het.timmers <- mr_heterogeneity(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
het.cases90 <- mr_heterogeneity(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
het.cases99 <- mr_heterogeneity(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

het_test6 <- rbind(het.pilling,het.timmers,het.cases90,het.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, method, Q, Q_df, Q_pval) %>%
  rename(`Longevity Definition` = Outcome.dat)
```

```{r}
kbl(het_test6, booktabs = T, caption = "Test for Heterogeneity using Cochran's Q-test")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Evidence of Heterogeneity in the Timmers data.

### Check for Pleiotropy

```{r include=FALSE}
plei.pilling <- mr_pleiotropy_test(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
plei.timmers <- mr_pleiotropy_test(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
plei.cases90 <- mr_pleiotropy_test(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
plei.cases99 <- mr_pleiotropy_test(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

pli_test6 <- rbind(plei.pilling, plei.timmers,plei.cases90,plei.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, egger_intercept, se, pval) %>%
  rename(`Longevity Definition` = Outcome.dat,
         `Egger Intercept` = egger_intercept)

```

```{r}
kbl(pli_test6, booktabs = T, caption = "Test for Pleiotropy ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

no indication from this test that there is the presence of horizontal pleiotropy


## Leave on out Analysis - Check to see if there is one SNP influencing the results
```{r include=FALSE}
loo.pilling <- mr_leaveoneout(Pilling_harm_dat)%>%
  mutate(outcome.data = "Pilling")
loo.timmers <-mr_leaveoneout(Timmers_harm_dat)%>%
  mutate(outcome.data = "Timmers")
loo.cases90 <-mr_leaveoneout(Cases90_harm_dat)%>%
  mutate(outcome.data = "Cases90")
loo.cases99 <- mr_leaveoneout(Cases99_harm_dat)%>%
  mutate(outcome.data = "Cases99")

loo_test6 <- rbind(loo.pilling, loo.timmers,loo.cases90,loo.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(outcome.data, exposure, samplesize, SNP, b,se, p) %>%
  rename(`Longevity Definition` = outcome.data,
         `Sample Size` = samplesize,
         Beta = b,
         SE = se,
         `p-value` = p)


```


```{r}
plot_grid(p1, p2, labels = c("Pilling", "Timmers"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -1, label_colour = "blue")

plot_grid(p3, p4, labels = c("Deleen 90", "Deleen 99"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")

kbl(loo_test6, booktabs = T, caption = "Leave one out Analysis ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Although there are some SNPs which significantly affect the point estimate, SNPs overlap.


## Scatter plot of Three Methods
```{r cache= TRUE, include=FALSE}
mr.pilling <- mr(Pilling_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.timmers <- mr(Timmers_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases90 <- mr(Cases90_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases99 <- mr(Cases99_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))

sc.pilling <- mrscatter_nolegend(mr.pilling, Pilling_harm_dat, title = "(a) Pilling")
sc.timmers <- mrscatter_nolegend(mr.timmers, Timmers_harm_dat, title ="(b) Timmers")
sc.cases90 <- mrscatter_nolegend(mr.cases90, Cases90_harm_dat, title ="(c) Deleen 90th Percentile")
sc.cases99 <- mrscatter_legend(mr.cases99, Cases99_harm_dat, title ="(d) Deleen 99th Percentile")

```

```{r}
exposure <- Pilling_harm_dat$exposure[1]

scatter <- plot_grid(sc.pilling[[1]],sc.timmers[[1]], sc.cases90[[1]], sc.cases99[[1]], nrow = 2)

scatter6 <- add_sub(scatter, paste("SNP effect on ", exposure), y = 0.2, vjust = 0, size = 8 )
ggdraw(scatter6)

# to check if the graph is the same
#mr_scatter_plot(mr.timmers, Timmers_harm_dat)

ggsave("images/HbA1c_scatter.png")
```

## Forest Plot of Single SNP estimates compared to estimates with all SNPs

```{r include=FALSE}
pilling.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
timmers.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases90.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases99.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
```

```{r}
plot_grid(pilling.forest, timmers.forest,labels =  c("Pilling", "Timmers"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
plot_grid(cases90.forest, cases99.forest,labels =  c("Deleen 90", "Deleen 99"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
```

## Funnel plot of Single SNP analysis compared to the estimates from all SNPs



```{r include=FALSE}
pilling.funnel <- mr_funnel_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
timmers.funnel <- mr_funnel_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases90.funnel <- mr_funnel_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases99.funnel <- mr_funnel_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
pilling.funnel$`ukb-d-30750_irnt.e11jiR`$theme$legend.direction <- 'none'
pilling.funnel$`ukb-d-30750_irnt.e11jiR`$theme$legend.position <- 'none'
timmers.funnel$`ukb-d-30750_irnt.IC5Hqz`$theme$legend.direction <- 'none'
timmers.funnel$`ukb-d-30750_irnt.IC5Hqz`$theme$legend.position <- 'none'
cases90.funnel$`ukb-d-30750_irnt.S5PUzu`$theme$legend.direction <- 'none'
cases90.funnel$`ukb-d-30750_irnt.S5PUzu`$theme$legend.position <- 'none'
cases99.funnel$`ukb-d-30750_irnt.gFljFJ`$theme$legend.direction <- 'none'
cases99.funnel$`ukb-d-30750_irnt.gFljFJ`$theme$legend.position <- 'none'

Pilling_harm_dat$exposure[1]

```

```{r}


p1 <- pilling.funnel$`ukb-d-30750_irnt.e11jiR` + 
  ggtitle("Parental 90") + theme(plot.title = element_text(color = "blue"))
p2 <- timmers.funnel$`ukb-d-30750_irnt.IC5Hqz`+ 
  ggtitle("Parental") + theme(plot.title = element_text(color = "blue"))
p3 <- cases90.funnel$`ukb-d-30750_irnt.S5PUzu`+ 
  ggtitle("Cases 90" ) + theme(plot.title = element_text(color = "blue"))
p4 <- cases99.funnel$`ukb-d-30750_irnt.gFljFJ` + 
  ggtitle("Cases 99") + theme(plot.title = element_text(color = "blue"))

annotate_figure(ggarrange(p1,p2,p3,p4, nrow = 2, ncol = 2) ,
                top = paste0("Single SNP estimates of ", Pilling_harm_dat$exposure[1], " on Longevity"))

ggsave("images/HbA1c_funnel.png")
```




# Prostate Cancer 7
```{r}
betas <- sel_graph %>%
  filter(exposure == c( "Prostate cancer || id:ieu-b-85"))%>%
  dplyr::select( b.pilling, b.timmers, b.90, b.99) %>%
  unlist()
se <- sel_graph %>%
  filter(exposure == c( "Prostate cancer || id:ieu-b-85"))%>%
  dplyr::select(se.pilling, se.timmers, se.90, se.99)%>%
  unlist()
nsnp <- sel_graph %>%
  filter(exposure == c( "Prostate cancer || id:ieu-b-85"))%>%
  dplyr::select(nsnp.pilling, nsnp.timmers, nsnp.90, nsnp.99)%>%
  unlist()

Outcome <- c( "Pilling", "Timmers", "Deleen 90", "Deleen 99")
exposure <- c( "Prostate cancer || id:ieu-b-85")
graph.data <- as.data.frame(cbind(Outcome,exposure, betas, se, nsnp))

graph.data$betas <- as.numeric(graph.data$betas)
graph.data$se <- as.numeric(graph.data$se)

graph.data7 <- graph.data %>%
  mutate(lower.CI = betas - (1.96*se),
         upper.CI = betas + (1.96*se),
         OR = exp(betas),
         lower.OR.CI = exp(lower.CI),
         upper.OR.CI = exp(upper.CI))
graph.data7
ggplot(graph.data7) +
  geom_segment(aes(x = lower.CI, xend = upper.CI, y= 1:4, yend = 1:4, col = Outcome)) +
  xlab("") +
  ylab("") +
  ggtitle("Confidence Intervals for Four Longevity Outcomes")
```




## Sensivity Analysis

```{r include=FALSE}



Pilling_harm_dat <- Pilling %>%
  filter(exposure ==  "Prostate cancer || id:ieu-b-85")

Timmers_harm_dat <- Timmers %>%
  filter(exposure ==  "Prostate cancer || id:ieu-b-85")

Cases90_harm_dat <- Cases90 %>%
  filter(exposure ==  "Prostate cancer || id:ieu-b-85")

Cases99_harm_dat <- Cases99 %>%
  filter(exposure ==  "Prostate cancer || id:ieu-b-85")
```

### Heterogeneity Test

```{r include=FALSE}
het.pilling <- mr_heterogeneity(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
het.timmers <- mr_heterogeneity(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
het.cases90 <- mr_heterogeneity(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
het.cases99 <- mr_heterogeneity(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

het_test7 <- rbind(het.pilling,het.timmers,het.cases90,het.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, method, Q, Q_df, Q_pval) %>%
  rename(`Longevity Definition` = Outcome.dat)
```

```{r}
kbl(het_test7, booktabs = T, caption = "Test for Heterogeneity using Cochran's Q-test")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Evidence of Heterogeneity in the Timmers data.

### Check for Pleiotropy

```{r include=FALSE}
plei.pilling <- mr_pleiotropy_test(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
plei.timmers <- mr_pleiotropy_test(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
plei.cases90 <- mr_pleiotropy_test(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
plei.cases99 <- mr_pleiotropy_test(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

pli_test7 <- rbind(plei.pilling, plei.timmers,plei.cases90,plei.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, egger_intercept, se, pval) %>%
  rename(`Longevity Definition` = Outcome.dat,
         `Egger Intercept` = egger_intercept)

```

```{r}
kbl(pli_test7, booktabs = T, caption = "Test for Pleiotropy ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

no indication from this test that there is the presence of horizontal pleiotropy


## Leave on out Analysis - Check to see if there is one SNP influencing the results
```{r include=FALSE}
loo.pilling <- mr_leaveoneout(Pilling_harm_dat)%>%
  mutate(outcome.data = "Pilling")
loo.timmers <-mr_leaveoneout(Timmers_harm_dat)%>%
  mutate(outcome.data = "Timmers")
loo.cases90 <-mr_leaveoneout(Cases90_harm_dat)%>%
  mutate(outcome.data = "Cases90")
loo.cases99 <- mr_leaveoneout(Cases99_harm_dat)%>%
  mutate(outcome.data = "Cases99")

loo_test7 <- rbind(loo.pilling, loo.timmers,loo.cases90,loo.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(outcome.data, exposure, samplesize, SNP, b,se, p) %>%
  rename(`Longevity Definition` = outcome.data,
         `Sample Size` = samplesize,
         Beta = b,
         SE = se,
         `p-value` = p)

p1 <- mr_leaveoneout_plot(loo.pilling)
p2 <- mr_leaveoneout_plot(loo.timmers)
p3 <- mr_leaveoneout_plot(loo.cases90)
p4 <- mr_leaveoneout_plot(loo.cases99)

p1 <- grid.arrange(grobs = p1)
p2 <- grid.arrange(grobs = p2)
p3 <- grid.arrange(grobs = p3)
p4 <- grid.arrange(grobs = p4)
```


```{r}
plot_grid(p1, p2, labels = c("Pilling", "Timmers"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -1, label_colour = "blue")

plot_grid(p3, p4, labels = c("Deleen 90", "Deleen 99"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")

kbl(loo_test7, booktabs = T, caption = "Leave one out Analysis ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Although there are some SNPs which significantly affect the point estimate, all SNPs overlap.


## Scatter plot of Three Methods
```{r cache= TRUE, include=FALSE}
mr.pilling <- mr(Pilling_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.timmers <- mr(Timmers_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases90 <- mr(Cases90_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases99 <- mr(Cases99_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))

sc.pilling <- mrscatter_nolegend(mr.pilling, Pilling_harm_dat, title = "(a) Pilling")
sc.timmers <- mrscatter_nolegend(mr.timmers, Timmers_harm_dat, title ="(b) Timmers")
sc.cases90 <- mrscatter_nolegend(mr.cases90, Cases90_harm_dat, title ="(c) Deleen 90th Percentile")
sc.cases99 <- mrscatter_legend(mr.cases99, Cases99_harm_dat, title ="(d) Deleen 99th Percentile")

```

```{r}
exposure <- Pilling_harm_dat$exposure[1]

scatter <- plot_grid(sc.pilling[[1]],sc.timmers[[1]], sc.cases90[[1]], sc.cases99[[1]], nrow = 2)

scatter7 <- add_sub(scatter, paste("SNP effect on ", exposure), y = 0.2, vjust = 0, size = 8 )
ggdraw(scatter7)

# to check if the graph is the same
#mr_scatter_plot(mr.timmers, Timmers_harm_dat)

ggsave("images/prostatecancer_scatter.png")
```

## Forest Plot of Single SNP estimates compared to estimates with all SNPs

```{r include=FALSE}
pilling.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
timmers.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases90.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases99.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
```

```{r}
plot_grid(pilling.forest, timmers.forest,labels =  c("Pilling", "Timmers"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
plot_grid(cases90.forest, cases99.forest,labels =  c("Deleen 90", "Deleen 99"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
```

## Funnel plot of Single SNP analysis compared to the estimates from all SNPs



```{r include=FALSE}
pilling.funnel <- mr_funnel_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
timmers.funnel <- mr_funnel_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases90.funnel <- mr_funnel_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases99.funnel <- mr_funnel_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
pilling.funnel$`ieu-b-85.UgsZxW`$theme$legend.direction <- 'none'
pilling.funnel$`ieu-b-85.UgsZxW`$theme$legend.position <- 'none'
timmers.funnel$`ieu-b-85.L7ZNH0`$theme$legend.direction <- 'none'
timmers.funnel$`ieu-b-85.L7ZNH0`$theme$legend.position <- 'none'
cases90.funnel$`ieu-b-85.UstNEj`$theme$legend.direction <- 'none'
cases90.funnel$`ieu-b-85.UstNEj`$theme$legend.position <- 'none'
cases99.funnel$`ieu-b-85.qEMtZG`$theme$legend.direction <- 'none'
cases99.funnel$`ieu-b-85.qEMtZG`$theme$legend.position <- 'none'

Pilling_harm_dat$exposure[1]

```

```{r}


p1 <- pilling.funnel$`ieu-b-85.UgsZxW` + 
  ggtitle("Parental 90") + theme(plot.title = element_text(color = "blue"))
p2 <- timmers.funnel$`ieu-b-85.L7ZNH0` + 
  ggtitle("Parental") + theme(plot.title = element_text(color = "blue"))
p3 <- cases90.funnel$`ieu-b-85.UstNEj` + 
  ggtitle("Cases 90" ) + theme(plot.title = element_text(color = "blue"))
p4 <- cases99.funnel$`ieu-b-85.qEMtZG` + 
  ggtitle("Cases 99") + theme(plot.title = element_text(color = "blue"))

annotate_figure(ggarrange(p1,p2,p3,p4, nrow = 2, ncol = 2) ,
                top = paste0("Single SNP estimates of ", Pilling_harm_dat$exposure[1], " on Longevity"))

ggsave("images/prostatecancer_funnel.png")
```





# Coronary Artery Disease 8

```{r}
betas <- sel_graph %>%
  filter(exposure == c("Coronary artery disease || id:ebi-a-GCST005194"))%>%
  dplyr::select( b.pilling, b.timmers, b.90, b.99) %>%
  unlist()
se <- sel_graph %>%
  filter(exposure == c("Coronary artery disease || id:ebi-a-GCST005194"))%>%
  dplyr::select(se.pilling, se.timmers, se.90, se.99)%>%
  unlist()
nsnp <- sel_graph %>%
  filter(exposure == c("Coronary artery disease || id:ebi-a-GCST005194"))%>%
  dplyr::select(nsnp.pilling, nsnp.timmers, nsnp.90, nsnp.99)%>%
  unlist()

Outcome <- c("Pilling", "Timmers", "Deleen 90", "Deleen 99")
exposure <- c("Coronary artery disease || id:ebi-a-GCST005194")
graph.data <- as.data.frame(cbind(Outcome,exposure, betas, se, nsnp))

graph.data$betas <- as.numeric(graph.data$betas)
graph.data$se <- as.numeric(graph.data$se)

graph.data8 <- graph.data %>%
  mutate(lower.CI = betas - (1.96*se),
         upper.CI = betas + (1.96*se),
         OR = exp(betas),
         lower.OR.CI = exp(lower.CI),
         upper.OR.CI = exp(upper.CI))
graph.data8
ggplot(graph.data8) +
  geom_segment(aes(x = lower.CI, xend = upper.CI, y= 1:4, yend = 1:4, col = Outcome)) +
  xlab("") +
  ylab("") +
  ggtitle("Confidence Intervals for Four Longevity Outcomes")
```


As expected, the most overlap is between the two Deleen definitions and the Pilling and Timmers datasets (UK Biobank).

## Sensivity Analysis

```{r include=FALSE}



Pilling_harm_dat <- Pilling %>%
  filter(exposure == "Coronary artery disease || id:ebi-a-GCST005194")

Timmers_harm_dat<- Timmers %>%
  filter(exposure == "Coronary artery disease || id:ebi-a-GCST005194")

Cases90_harm_dat <- Cases90 %>%
  filter(exposure == "Coronary artery disease || id:ebi-a-GCST005194")

Cases99_harm_dat <- Cases99 %>%
  filter(exposure == "Coronary artery disease || id:ebi-a-GCST005194")
```

### Heterogeneity Test

```{r include=FALSE}
het.pilling <- mr_heterogeneity(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
het.timmers <- mr_heterogeneity(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
het.cases90 <- mr_heterogeneity(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
het.cases99 <- mr_heterogeneity(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

het_test8 <- rbind(het.pilling,het.timmers,het.cases90,het.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, method, Q, Q_df, Q_pval) %>%
  rename(`Longevity Definition` = Outcome.dat)
```

```{r}
kbl(het_test8, booktabs = T, caption = "Test for Heterogeneity using Cochran's Q-test")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Evidence of Heterogeneity in the Timmers data.

### Check for Pleiotropy

```{r include=FALSE}
plei.pilling <- mr_pleiotropy_test(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
plei.timmers <- mr_pleiotropy_test(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
plei.cases90 <- mr_pleiotropy_test(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
plei.cases99 <- mr_pleiotropy_test(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

pli_test8 <- rbind(plei.pilling, plei.timmers,plei.cases90,plei.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, egger_intercept, se, pval) %>%
  rename(`Longevity Definition` = Outcome.dat,
         `Egger Intercept` = egger_intercept)

```

```{r}
kbl(pli_test8, booktabs = T, caption = "Test for Pleiotropy ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

no indication from this test that there is the presence of horizontal pleiotropy


## Leave on out Analysis - Check to see if there is one SNP influencing the results
```{r include=FALSE}
loo.pilling <- mr_leaveoneout(Pilling_harm_dat)%>%
  mutate(outcome.data = "Pilling")
loo.timmers <-mr_leaveoneout(Timmers_harm_dat)%>%
  mutate(outcome.data = "Timmers")
loo.cases90 <-mr_leaveoneout(Cases90_harm_dat)%>%
  mutate(outcome.data = "Cases90")
loo.cases99 <- mr_leaveoneout(Cases99_harm_dat)%>%
  mutate(outcome.data = "Cases99")

loo_test8 <- rbind(loo.pilling, loo.timmers,loo.cases90,loo.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(outcome.data, exposure, samplesize, SNP, b,se, p) %>%
  rename(`Longevity Definition` = outcome.data,
         `Sample Size` = samplesize,
         Beta = b,
         SE = se,
         `p-value` = p)

```


```{r}
plot_grid(p1, p2, labels = c("Pilling", "Timmers"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -1, label_colour = "blue")

plot_grid(p3, p4, labels = c("Deleen 90", "Deleen 99"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")

kbl(loo_test8, booktabs = T, caption = "Leave one out Analysis ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Although there are some SNPs which significantly affect the point estimate, most SNPs overlap.


## Scatter plot of Three Methods
```{r cache= TRUE, include=FALSE}
mr.pilling <- mr(Pilling_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.timmers <- mr(Timmers_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases90 <- mr(Cases90_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases99 <- mr(Cases99_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))

sc.pilling <- mrscatter_nolegend(mr.pilling, Pilling_harm_dat, title = "(a) Pilling")
sc.timmers <- mrscatter_nolegend(mr.timmers, Timmers_harm_dat, title ="(b) Timmers")
sc.cases90 <- mrscatter_nolegend(mr.cases90, Cases90_harm_dat, title ="(c) Deleen 90th Percentile")
sc.cases99 <- mrscatter_legend(mr.cases99, Cases99_harm_dat, title ="(d) Deleen 99th Percentile")

```

```{r}
exposure <- Pilling_harm_dat$exposure[1]

scatter <- plot_grid(sc.pilling[[1]],sc.timmers[[1]], sc.cases90[[1]], sc.cases99[[1]], nrow = 2)

scatter8 <- add_sub(scatter, paste("SNP effect on ", exposure), y = 0.2, vjust = 0, size = 8 )
ggdraw(scatter8)

# to check if the graph is the same
#mr_scatter_plot(mr.timmers, Timmers_harm_dat)

ggsave("images/coronarydisease_scatter.png")
```

## Forest Plot of Single SNP estimates compared to estimates with all SNPs

```{r include=FALSE}
pilling.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
timmers.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases90.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases99.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
```

```{r}
plot_grid(pilling.forest, timmers.forest,labels =  c("Pilling", "Timmers"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
plot_grid(cases90.forest, cases99.forest,labels =  c("Deleen 90", "Deleen 99"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
```

## Funnel plot of Single SNP analysis compared to the estimates from all SNPs



```{r include=FALSE}
pilling.funnel <- mr_funnel_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
timmers.funnel <- mr_funnel_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases90.funnel <- mr_funnel_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases99.funnel <- mr_funnel_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
pilling.funnel$`ebi-a-GCST005194.8WRRHB`$theme$legend.direction <- 'none'
pilling.funnel$`ebi-a-GCST005194.8WRRHB`$theme$legend.position <- 'none'
timmers.funnel$`ebi-a-GCST005194.UeZfVW`$theme$legend.direction <- 'none'
timmers.funnel$`ebi-a-GCST005194.UeZfVW`$theme$legend.position <- 'none'
cases90.funnel$`ebi-a-GCST005194.OHpR2s`$theme$legend.direction <- 'none'
cases90.funnel$`ebi-a-GCST005194.OHpR2s`$theme$legend.position <- 'none'
cases99.funnel$`ebi-a-GCST005194.hCW1Ye`$theme$legend.direction <- 'none'
cases99.funnel$`ebi-a-GCST005194.hCW1Ye`$theme$legend.position <- 'none'

Pilling_harm_dat$exposure[1]

```

```{r}


p1 <- pilling.funnel$`ebi-a-GCST005194.8WRRHB` + 
  ggtitle("Parental 90") + theme(plot.title = element_text(color = "blue"))
p2 <- timmers.funnel$`ebi-a-GCST005194.UeZfVW` + 
  ggtitle("Parental") + theme(plot.title = element_text(color = "blue"))
p3 <- cases90.funnel$`ebi-a-GCST005194.OHpR2s`+ 
  ggtitle("Cases 90" ) + theme(plot.title = element_text(color = "blue"))
p4 <- cases99.funnel$`ebi-a-GCST005194.hCW1Ye` + 
  ggtitle("Cases 99") + theme(plot.title = element_text(color = "blue"))

annotate_figure(ggarrange(p1,p2,p3,p4, nrow = 2, ncol = 2) ,
                top = paste0("Single SNP estimates of ", Pilling_harm_dat$exposure[1], " on Longevity"))

ggsave("images/coronarydisease_funnel.png")

```


# Type 2 Diabetes 9


```{r}
betas <- sel_graph %>%
  filter(exposure == c("Type 2 diabetes || id:ebi-a-GCST006867"))%>%
  dplyr::select( b.pilling, b.timmers, b.90, b.99) %>%
  unlist()
se <- sel_graph %>%
  filter(exposure == c("Type 2 diabetes || id:ebi-a-GCST006867"))%>%
  dplyr::select(se.pilling, se.timmers, se.90, se.99)%>%
  unlist()
nsnp <- sel_graph %>%
  filter(exposure == c("Type 2 diabetes || id:ebi-a-GCST006867"))%>%
  dplyr::select(nsnp.pilling, nsnp.timmers, nsnp.90, nsnp.99)%>%
  unlist()
Outcome <- c("Pilling", "Timmers", "Deleen 90", "Deleen 99")
exposure <- c("Type 2 diabetes || id:ebi-a-GCST006867")
graph.data <- as.data.frame(cbind(Outcome,exposure, betas, se,nsnp))

graph.data$betas <- as.numeric(graph.data$betas)
graph.data$se <- as.numeric(graph.data$se)

graph.data9 <- graph.data %>%
  mutate(lower.CI = betas - (1.96*se),
         upper.CI = betas + (1.96*se),
         OR = exp(betas),
         lower.OR.CI = exp(lower.CI),
         upper.OR.CI = exp(upper.CI))
graph.data9
ggplot(graph.data9) +
  geom_segment(aes(x = lower.CI, xend = upper.CI, y= 1:4, yend = 1:4, col = Outcome)) +
  xlab("") +
  ylab("") +
  ggtitle("Confidence Intervals for Four Longevity Outcomes")
```




## Sensivity Analysis

```{r include=FALSE}

#filter by Alzheimer

Pilling_harm_dat <- Pilling %>%
  filter(exposure == "Type 2 diabetes || id:ebi-a-GCST006867")

Timmers_harm_dat  <- Timmers %>%
  filter(exposure == "Type 2 diabetes || id:ebi-a-GCST006867")

Cases90_harm_dat  <- Cases90 %>%
  filter(exposure == "Type 2 diabetes || id:ebi-a-GCST006867")

Cases99_harm_dat  <- Cases99 %>%
  filter(exposure == "Type 2 diabetes || id:ebi-a-GCST006867")
```

### Heterogeneity Test

```{r include=FALSE}
het.pilling <- mr_heterogeneity(Pilling_harm_dat )%>%
  mutate(Outcome.dat = "Pilling")
het.timmers <- mr_heterogeneity(Timmers_harm_dat )%>%
  mutate(Outcome.dat = "Timmers")
het.cases90 <- mr_heterogeneity(Cases90_harm_dat )%>%
  mutate(Outcome.dat = "Cases90")
het.cases99 <- mr_heterogeneity(Cases99_harm_dat )%>%
  mutate(Outcome.dat = "Cases99")

het_test9 <- rbind(het.pilling,het.timmers,het.cases90,het.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, method, Q, Q_df, Q_pval) %>%
  rename(`Longevity Definition` = Outcome.dat)
```

```{r}
kbl(het_test9, booktabs = T, caption = "Test for Heterogeneity using Cochran's Q-test")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Evidence of Heterogeneity in the Timmers data.

### Check for Pleiotropy

```{r include=FALSE}
plei.pilling <- mr_pleiotropy_test(Pilling_harm_dat )%>%
  mutate(Outcome.dat = "Pilling")
plei.timmers <- mr_pleiotropy_test(Timmers_harm_dat )%>%
  mutate(Outcome.dat = "Timmers")
plei.cases90 <- mr_pleiotropy_test(Cases90_harm_dat )%>%
  mutate(Outcome.dat = "Cases90")
plei.cases99 <- mr_pleiotropy_test(Cases99_harm_dat )%>%
  mutate(Outcome.dat = "Cases99")

pli_test9 <- rbind(plei.pilling, plei.timmers,plei.cases90,plei.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, egger_intercept, se, pval) %>%
  rename(`Longevity Definition` = Outcome.dat,
         `Egger Intercept` = egger_intercept)

```

```{r}
kbl(pli_test9, booktabs = T, caption = "Test for Pleiotropy ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

no indication from this test that there is the presence of horizontal pleiotropy


## Leave on out Analysis - Check to see if there is one SNP influencing the results
```{r include=FALSE}
loo.pilling <- mr_leaveoneout(Pilling_harm_dat)%>%
  mutate(outcome.data = "Pilling")
loo.timmers <-mr_leaveoneout(Timmers_harm_dat)%>%
  mutate(outcome.data = "Timmers")
loo.cases90 <-mr_leaveoneout(Cases90_harm_dat)%>%
  mutate(outcome.data = "Cases90")
loo.cases99 <- mr_leaveoneout(Cases99_harm_dat)%>%
  mutate(outcome.data = "Cases99")

loo_test9 <- rbind(loo.pilling, loo.timmers,loo.cases90,loo.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(outcome.data, exposure, samplesize, SNP, b,se, p) %>%
  rename(`Longevity Definition` = outcome.data,
         `Sample Size` = samplesize,
         Beta = b,
         SE = se,
         `p-value` = p)


```


```{r}
plot_grid(p1, p2, labels = c("Pilling", "Timmers"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -1, label_colour = "blue")

plot_grid(p3, p4, labels = c("Deleen 90", "Deleen 99"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")

kbl(loo_test9, booktabs = T, caption = "Leave one out Analysis ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Although there are some SNPs which significantly affect the point estimate, most SNPs overlap.


## Scatter plot of Three Methods
```{r cache= TRUE, include=FALSE}
mr.pilling <- mr(Pilling_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.timmers <- mr(Timmers_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases90 <- mr(Cases90_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases99 <- mr(Cases99_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))

sc.pilling <- mrscatter_nolegend(mr.pilling, Pilling_harm_dat, title = "(a) Pilling")
sc.timmers <- mrscatter_nolegend(mr.timmers, Timmers_harm_dat, title ="(b) Timmers")
sc.cases90 <- mrscatter_nolegend(mr.cases90, Cases90_harm_dat, title ="(c) Deleen 90th Percentile")
sc.cases99 <- mrscatter_legend(mr.cases99, Cases99_harm_dat, title ="(d) Deleen 99th Percentile")

```

```{r}
exposure <- Pilling_harm_dat$exposure[1]

scatter <- plot_grid(sc.pilling[[1]],sc.timmers[[1]], sc.cases90[[1]], sc.cases99[[1]], nrow = 2)

scatter9 <- add_sub(scatter, paste("SNP effect on ", exposure), y = 0.2, vjust = 0, size = 8 )
ggdraw(scatter9)

# to check if the graph is the same
#mr_scatter_plot(mr.timmers, Timmers_harm_dat)


ggsave("images/type2diabetes_scatter.png")
```

## Forest Plot of Single SNP estimates compared to estimates with all SNPs

```{r include=FALSE}
pilling.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
timmers.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases90.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases99.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
```

```{r}
plot_grid(pilling.forest, timmers.forest,labels =  c("Pilling", "Timmers"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
plot_grid(cases90.forest, cases99.forest,labels =  c("Deleen 90", "Deleen 99"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
```

## Funnel plot of Single SNP analysis compared to the estimates from all SNPs



```{r include=FALSE}
pilling.funnel <- mr_funnel_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
timmers.funnel <- mr_funnel_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases90.funnel <- mr_funnel_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases99.funnel <- mr_funnel_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
pilling.funnel$`ebi-a-GCST006867.P3u1a1`$theme$legend.direction <- 'none'
pilling.funnel$`ebi-a-GCST006867.P3u1a1`$theme$legend.position <- 'none'
timmers.funnel$`ebi-a-GCST006867.FXT4Yt`$theme$legend.direction <- 'none'
timmers.funnel$`ebi-a-GCST006867.FXT4Yt`$theme$legend.position <- 'none'
cases90.funnel$`ebi-a-GCST006867.LXTpae`$theme$legend.direction <- 'none'
cases90.funnel$`ebi-a-GCST006867.LXTpae`$theme$legend.position <- 'none'
cases99.funnel$`ebi-a-GCST006867.lDDA6L`$theme$legend.direction <- 'none'
cases99.funnel$`ebi-a-GCST006867.lDDA6L`$theme$legend.position <- 'none'

Pilling_harm_dat$exposure[1]

```

```{r}


p1 <- pilling.funnel$`ebi-a-GCST006867.P3u1a1` + 
  ggtitle("Parental 90") + theme(plot.title = element_text(color = "blue"))
p2 <- timmers.funnel$`ebi-a-GCST006867.FXT4Yt` + 
  ggtitle("Parental") + theme(plot.title = element_text(color = "blue"))
p3 <- cases90.funnel$`ebi-a-GCST006867.LXTpae`+ 
  ggtitle("Cases 90" ) + theme(plot.title = element_text(color = "blue"))
p4 <- cases99.funnel$`ebi-a-GCST006867.lDDA6L` + 
  ggtitle("Cases 99") + theme(plot.title = element_text(color = "blue"))

annotate_figure(ggarrange(p1,p2,p3,p4, nrow = 2, ncol = 2) ,
                top = paste0("Single SNP estimates of ", Pilling_harm_dat$exposure[1], " on Longevity"))

ggsave("images/type2diabetes_funnel.png")

```

# Years of Education 10

```{r}
betas <- sel_graph %>%
  filter(exposure == c( "Years of schooling || id:ieu-a-1239"))%>%
  dplyr::select( b.pilling, b.timmers, b.90, b.99) %>%
  unlist()
se <- sel_graph %>%
  filter(exposure == c( "Years of schooling || id:ieu-a-1239"))%>%
  dplyr::select(se.pilling, se.timmers, se.90, se.99)%>%
  unlist()
nsnp <- sel_graph %>%
  filter(exposure == c( "Years of schooling || id:ieu-a-1239"))%>%
  dplyr::select(nsnp.pilling, nsnp.timmers, nsnp.90, nsnp.99)%>%
  unlist()

Outcome <- c("Pilling", "Timmers", "Deleen 90", "Deleen 99")
exposure <- c( "Years of schooling || id:ieu-a-1239")
graph.data <- as.data.frame(cbind(Outcome,exposure, betas, se, nsnp))

graph.data$betas <- as.numeric(graph.data$betas)
graph.data$se <- as.numeric(graph.data$se)

graph.data10 <- graph.data %>%
  mutate(lower.CI = betas - (1.96*se),
         upper.CI = betas + (1.96*se),
         OR = exp(betas),
         lower.OR.CI = exp(lower.CI),
         upper.OR.CI = exp(upper.CI))
graph.data10
ggplot(graph.data10) +
  geom_segment(aes(x = lower.CI, xend = upper.CI, y= 1:4, yend = 1:4, col = Outcome)) +
  xlab("") +
  ylab("") +
  ggtitle("Confidence Intervals for Four Longevity Outcomes")
```


As expected, the most overlap is between the two Deleen definitions and the Pilling and Timmers datasets (UK Biobank).

## Sensivity Analysis

```{r include=FALSE}

#filter by Alzheimer

Pilling_harm_dat <- Pilling %>%
  filter(exposure ==  "Years of schooling || id:ieu-a-1239")

Timmers_harm_dat <- Timmers %>%
  filter(exposure ==  "Years of schooling || id:ieu-a-1239")

Cases90_harm_dat <- Cases90 %>%
  filter(exposure ==  "Years of schooling || id:ieu-a-1239")

Cases99_harm_dat <- Cases99 %>%
  filter(exposure ==  "Years of schooling || id:ieu-a-1239")
```

### Heterogeneity Test

```{r include=FALSE}
het.pilling <- mr_heterogeneity(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
het.timmers <- mr_heterogeneity(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
het.cases90 <- mr_heterogeneity(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
het.cases99 <- mr_heterogeneity(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

het_test10 <- rbind(het.pilling,het.timmers,het.cases90,het.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, method, Q, Q_df, Q_pval) %>%
  rename(`Longevity Definition` = Outcome.dat)
```

```{r}
kbl(het_test10, booktabs = T, caption = "Test for Heterogeneity using Cochran's Q-test")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Evidence of Heterogeneity in the Timmers data.

### Check for Pleiotropy

```{r include=FALSE}
plei.pilling <- mr_pleiotropy_test(Pilling_harm_dat)%>%
  mutate(Outcome.dat = "Pilling")
plei.timmers <- mr_pleiotropy_test(Timmers_harm_dat)%>%
  mutate(Outcome.dat = "Timmers")
plei.cases90 <- mr_pleiotropy_test(Cases90_harm_dat)%>%
  mutate(Outcome.dat = "Cases90")
plei.cases99 <- mr_pleiotropy_test(Cases99_harm_dat)%>%
  mutate(Outcome.dat = "Cases99")

pli_test10 <- rbind(plei.pilling, plei.timmers,plei.cases90,plei.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(Outcome.dat, exposure, egger_intercept, se, pval) %>%
  rename(`Longevity Definition` = Outcome.dat,
         `Egger Intercept` = egger_intercept)

```

```{r}
kbl(pli_test10, booktabs = T, caption = "Test for Pleiotropy ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

no indication from this test that there is the presence of horizontal pleiotropy


## Leave on out Analysis - Check to see if there is one SNP influencing the results
```{r include=FALSE}
loo.pilling <- mr_leaveoneout(Pilling_harm_dat)%>%
  mutate(outcome.data = "Pilling")
loo.timmers <-mr_leaveoneout(Timmers_harm_dat)%>%
  mutate(outcome.data = "Timmers")
loo.cases90 <-mr_leaveoneout(Cases90_harm_dat)%>%
  mutate(outcome.data = "Cases90")
loo.cases99 <- mr_leaveoneout(Cases99_harm_dat)%>%
  mutate(outcome.data = "Cases99")

loo_test10 <- rbind(loo.pilling, loo.timmers,loo.cases90,loo.cases99) %>%
  select(-id.exposure, -id.outcome, -outcome) %>%
  select(outcome.data, exposure, samplesize, SNP, b,se, p) %>%
  rename(`Longevity Definition` = outcome.data,
         `Sample Size` = samplesize,
         Beta = b,
         SE = se,
         `p-value` = p)



```


```{r}
plot_grid(p1, p2, labels = c("Pilling", "Timmers"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -1, label_colour = "blue")

plot_grid(p3, p4, labels = c("Deleen 90", "Deleen 99"),label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")

kbl(loo_test10, booktabs = T, caption = "Leave one out Analysis ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```

Although there are some SNPs which significantly affect the point estimate, most SNPs overlap.


## Scatter plot of Three Methods
```{r cache= TRUE, include=FALSE}
mr.pilling <- mr(Pilling_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.timmers <- mr(Timmers_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases90 <- mr(Cases90_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))
mr.cases99 <- mr(Cases99_harm_dat, method_list = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))

sc.pilling <- mrscatter_nolegend(mr.pilling, Pilling_harm_dat, title = "(a) Pilling")
sc.timmers <- mrscatter_nolegend(mr.timmers, Timmers_harm_dat, title ="(b) Timmers")
sc.cases90 <- mrscatter_nolegend(mr.cases90, Cases90_harm_dat, title ="(c) Deleen 90th Percentile")
sc.cases99 <- mrscatter_legend(mr.cases99, Cases99_harm_dat, title ="(d) Deleen 99th Percentile")

```

```{r}
exposure <- Pilling_harm_dat$exposure[1]

scatter <- plot_grid(sc.pilling[[1]],sc.timmers[[1]], sc.cases90[[1]], sc.cases99[[1]], nrow = 2)

scatter10 <- add_sub(scatter, paste("SNP effect on ", exposure), y = 0.2, vjust = 0, size = 8 )
ggdraw(scatter10)

# to check if the graph is the same
#mr_scatter_plot(mr.timmers, Timmers_harm_dat)


ggsave("images/education_scatter.png")
```

## Forest Plot of Single SNP estimates compared to estimates with all SNPs

```{r include=FALSE}
pilling.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
timmers.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases90.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
cases99.forest <- grid.arrange(grobs = mr_forest_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median"))))
```

```{r}
plot_grid(pilling.forest, timmers.forest,labels =  c("Pilling", "Timmers"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
plot_grid(cases90.forest, cases99.forest,labels =  c("Deleen 90", "Deleen 99"), label_x = 0, label_y = 0.05,
  hjust = 0, vjust = -0.5, label_colour = "blue")
```

## Funnel plot of Single SNP analysis compared to the estimates from all SNPs



```{r include=FALSE}
pilling.funnel <-  mr_funnel_plot(mr_singlesnp(Pilling_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
timmers.funnel <- mr_funnel_plot(mr_singlesnp(Timmers_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases90.funnel <- mr_funnel_plot(mr_singlesnp(Cases90_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
cases99.funnel <- mr_funnel_plot(mr_singlesnp(Cases99_harm_dat, all_method = c("mr_egger_regression", "mr_ivw", "mr_weighted_median")))
pilling.funnel$`ieu-a-1239.SdFUtY`$theme$legend.direction <- 'none'
pilling.funnel$`ieu-a-1239.SdFUtY`$theme$legend.position <- 'none'
timmers.funnel$`ieu-a-1239.RI5rdD`$theme$legend.direction <- 'none'
timmers.funnel$`ieu-a-1239.RI5rdD`$theme$legend.position <- 'none'
cases90.funnel$`ieu-a-1239.j2IkkU`$theme$legend.direction <- 'none'
cases90.funnel$`ieu-a-1239.j2IkkU`$theme$legend.position <- 'none'
cases99.funnel$`ieu-a-1239.cytrdU`$theme$legend.direction <- 'none'
cases99.funnel$`ieu-a-1239.cytrdU`$theme$legend.position <- 'none'

Pilling_harm_dat$exposure[1]

```

```{r}


p1 <- pilling.funnel$`ieu-a-1239.SdFUtY` + 
  ggtitle("Parental 90") + theme(plot.title = element_text(color = "blue"))
p2 <- timmers.funnel$`ieu-a-1239.RI5rdD` + 
  ggtitle("Parental") + theme(plot.title = element_text(color = "blue"))
p3 <- cases90.funnel$`ieu-a-1239.j2IkkU`+ 
  ggtitle("Cases 90" ) + theme(plot.title = element_text(color = "blue"))
p4 <- cases99.funnel$`ieu-a-1239.cytrdU` + 
  ggtitle("Cases 99") + theme(plot.title = element_text(color = "blue"))

annotate_figure(ggarrange(p1,p2,p3,p4, nrow = 2, ncol = 2) ,
                top = paste0("Single SNP estimates of ", Pilling_harm_dat$exposure[1], " on Longevity"))

ggsave("images/education_funnel.png")

```



# summary of selected Exposures

## Selected Exposures

```{r}
mr_cases90 <- readRDS("data/MR/MR_cases_90_gwas_ieu.rds") %>% 
  filter(exposure %in% c("Basal metabolic rate || id:ukb-b-16446", 
  "Age first had sexual intercourse || id:ukb-b-6591", "Overall health rating || id:ukb-b-6306",  
  "Non-cancer illness code, self-reported: hypertension || id:ukb-b-14057", "Arm fat mass (left) || id:ukb-a-287",
  "Glycated haemoglobin || id:ukb-d-30750_irnt","Prostate cancer || id:ieu-b-85", 
  "Coronary artery disease || id:ebi-a-GCST005194", "Type 2 diabetes || id:ebi-a-GCST006867", 
  "Years of schooling || id:ieu-a-1239"))%>% 
  mutate(Outcome = "Cases 90")
mr_cases99 <- readRDS("data/MR/MR_cases_99_gwas_ieu.rds")%>% 
  filter(exposure %in% c("Basal metabolic rate || id:ukb-b-16446", 
  "Age first had sexual intercourse || id:ukb-b-6591", "Overall health rating || id:ukb-b-6306",  
  "Non-cancer illness code, self-reported: hypertension || id:ukb-b-14057", "Arm fat mass (left) || id:ukb-a-287",
  "Glycated haemoglobin || id:ukb-d-30750_irnt","Prostate cancer || id:ieu-b-85", 
  "Coronary artery disease || id:ebi-a-GCST005194", "Type 2 diabetes || id:ebi-a-GCST006867", 
  "Years of schooling || id:ieu-a-1239"))%>%
  mutate(Outcome = "Cases 99")
mr_pilling <- readRDS("data/MR/MR_pilling_gwas_ieu.rds")%>% 
  filter(exposure %in% c("Basal metabolic rate || id:ukb-b-16446", 
  "Age first had sexual intercourse || id:ukb-b-6591", "Overall health rating || id:ukb-b-6306",  
  "Non-cancer illness code, self-reported: hypertension || id:ukb-b-14057", "Arm fat mass (left) || id:ukb-a-287",
  "Glycated haemoglobin || id:ukb-d-30750_irnt","Prostate cancer || id:ieu-b-85", 
  "Coronary artery disease || id:ebi-a-GCST005194", "Type 2 diabetes || id:ebi-a-GCST006867", 
  "Years of schooling || id:ieu-a-1239"))%>%
  mutate(Outcome = "Pilling")
mr_timmers <- readRDS("data/MR/MR_timmers_gwas_ieu.rds")%>% 
  filter(exposure %in% c("Basal metabolic rate || id:ukb-b-16446", 
  "Age first had sexual intercourse || id:ukb-b-6591", "Overall health rating || id:ukb-b-6306",  
  "Non-cancer illness code, self-reported: hypertension || id:ukb-b-14057", "Arm fat mass (left) || id:ukb-a-287",
  "Glycated haemoglobin || id:ukb-d-30750_irnt","Prostate cancer || id:ieu-b-85", 
  "Coronary artery disease || id:ebi-a-GCST005194", "Type 2 diabetes || id:ebi-a-GCST006867", 
  "Years of schooling || id:ieu-a-1239"))%>%
  mutate(Outcome = "Timmers")


table2 <- rbind(mr_cases90, mr_cases99, mr_pilling, mr_timmers)%>%
  mutate(lower.CI = b - (1.96*se),
         upper.CI = b + (1.96*se),
         OR = signif(exp(b),3),
         lower.OR.CI = signif(exp(lower.CI), 3),
         upper.OR.CI = signif(exp(upper.CI),3))
 OR_exp <- table2 %>%
   mutate(CI.95 = paste("(", lower.OR.CI, ", ", upper.OR.CI, ")"))%>%
   select(Outcome, method, exposure, OR, CI.95, nsnp) %>% 
   arrange(method) %>%
   filter(method %in% c("Inverse variance weighted", "MR Egger", "Weighted median"))
colnames(OR_exp) <- c("Outcome", "Estimation Method", "Exposure", "OR", "95% CI", "No. of SNPs")


kbl(OR_exp, booktabs = T, caption = "Table 2. Selected Exposure for Further Analysis: Basal Metabolic Rate.  ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```


## Heterogeneity

```{r}
het_test <- rbind(het_test1,het_test2, het_test3, het_test4, het_test5, het_test6, het_test7,het_test8, het_test9, het_test10)

kbl(het_test, booktabs = T, caption = "Test for Heterogeneity using Cochran's Q-test")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```


## Pleiotropy

```{r}
pli_test <- rbind(pli_test1, pli_test2, pli_test3, pli_test4, pli_test5, pli_test6, pli_test7,pli_test8,  pli_test9, pli_test10)
kbl(pli_test, booktabs = T, caption = "Test for Pleiotropy ")%>%
  kable_classic()%>%
  kable_styling(bootstrap_options = "striped")
```


#leveoneout
# ```{r}
# loo_test <- rbind(loo_test1, loo_test2, loo_test3, loo_test4, loo_test5, loo_test6, loo_test7,  loo_test9, loo_test10, loo_test11, loo_test12, loo_test13, loo_test14)
# kbl(loo_test, booktabs = T, caption = "Leave one out Analysis ")%>%
#   kable_classic()%>%
#   kable_styling(bootstrap_options = "striped")
# ```

# Cleaned up LOO graphs


```{r }

#basal metabolic rate
loo_graphs(loo_test1, exposure = "Basal metabolic rate || id:ukb-b-16446", outcome = "Longevity" )
ggsave( "images/Basal_met_rate_loo.png")

#Age at first sexual intercourse
loo_graphs(loo_test2, exposure = "Age first had sexual intercourse || id:ukb-b-6591", outcome = "Longevity" )
ggsave( "images/first_sex_loo.png")

# Overal health
loo_graphs(loo_test3, exposure = "Overall health rating || id:ukb-b-6306", outcome = "Longevity" )
ggsave( "images/overall_health_loo.png")

#hypertension
loo_graphs(loo_test4, exposure = " hypertension || id:ukb-b-14057", outcome = "Longevity" )
ggsave( "images/hypertension_loo.png")

#arm fat
loo_graphs(loo_test5, exposure ="Arm fat mass (left) || id:ukb-a-287", outcome = "Longevity" )
ggsave( "images/arm_fat_loo.png")

#HbA1c
loo_graphs(loo_test6, exposure = "Glycated haemoglobin || id:ukb-d-30750_irnt", outcome = "Longevity" )
ggsave( "images/HbA1c_loo.png")

#Prostate Cancer
loo_graphs(loo_test7, exposure = "Prostate cancer || id:ieu-b-85", outcome = "Longevity" )
ggsave( "images/prostatecancer_loo.png")

#Coronary disease
loo_graphs(loo_test8, exposure = "Coronary artery disease || id:ebi-a-GCST005194", outcome = "Longevity" )
ggsave( "images/coronarydisease_loo.png")

#Type 2 diabetes
loo_graphs(loo_test9, exposure = "Type 2 diabetes || id:ebi-a-GCST006867", outcome = "Longevity" )
ggsave( "images/type2diabetes_loo.png")

#Cognitive scores : highest math class taken
loo_graphs(loo_test10, exposure = "Years of schooling || id:ieu-a-1239", outcome = "Longevity" )
ggsave( "images/education.png")





```



# Flowchart

```{r}
library(DiagrammeR)

grViz("digraph flowchart {
      # node definitions with substituted label text
      node [fontname = Helvetica, shape = rectangle]
    
      
      tab1 [label = '@@1']
      tab2 [label = '@@2']
      tab3 [label = '@@3']
      tab4 [label = '@@4']
      tab5 [label = '@@5']
      tab6 [label = '@@6']
      tab7 [label = '@@7']
      tab8 [label = '@@8']
      tab9 [label = '@@9']
      tab10 [label = '@@10']
      tab11 [label = '@@11']
      tab12 [label = '@@12']
      tab13 [label = '@@13']
      tab14 [label = '@@14']
      tab15 [label = '@@15']
      tab16 [label = '@@16']
      tab17 [label = '@@17' , color = red]
      tab18 [label = '@@18']
      tab19 [label = '@@19']
      tab20 [label = '@@20']

      # edge definitions with the node IDs
      tab19 -> tab1 -> tab2 -> tab6 -> tab10;
      tab20 -> tab1 -> tab3 -> tab7 -> tab11;
      tab1 -> tab4 ->tab8 -> tab12;
      tab1 -> tab5 -> tab9 -> tab13;
      tab10 -> tab14;
      tab11 -> tab14;
      tab12 -> tab14;
      tab13 -> tab14;
      tab14 -> tab15 -> tab17;
      tab16 -> tab15;
      tab18 -> tab17
      
 
      
      }
      
      
      [1]: 'GWAS Catalog 4112 Unique traits'
      [2]: 'Outcome 1: Parental Longevity Top 10%'
      [3]: 'Outcome 2: Parental Longevity'
      [4]: 'Outcome 3: Individual Longevity Top 10%'
      [5]: 'Outcome 4: Individual Logevity Top 1%'
      [6]: '3990 Harmonized traits'
      [7]: '4007 Harmonized exposures'
      [8]: '3841 Harmonized exposures'
      [9]: '3811 Harmonized exposures'
      [10]: '353 significant exposures'
      [11]: '593 significant exposures'
      [12]: '235 significant exposures'
      [13]: '139 significant exposures'
      [14]: ' 685 Common exposures'
      [15]: '141 exposures meet criteria'
      [16]: 'Have at least 100 SNPs and significant associations with at least of three of the four longevity definitions'
      [17]: '10 exposures for sensitivity analyses'
      [18]: 'Chose one exposure from each category'
      [19]: '9357 GWAS associated traits'
      [20]: 'Has all information needed to perform 2SMR'
      ")
```



